{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "derived-folks",
   "metadata": {},
   "source": [
    "**Author:** Elias Rafael Ruiz-Morales\n",
    "\n",
    "**Institution:** Wellcome Sanger institute\n",
    "\n",
    "**July, 2023**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "valid-wilson",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "accepting-relative",
   "metadata": {},
   "source": [
    "# Logistic regression for cell annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "middle-relay",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format = 'retina'\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import os\n",
    "import sklearn\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "import matplotlib as mpl\n",
    "import scipy\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bronze-mustang",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "sublime-milwaukee",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: If you miss a compact list, please try `print_header`!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.7.6\n",
      "scanpy      1.7.0\n",
      "sinfo       0.3.1\n",
      "-----\n",
      "PIL                 8.3.0\n",
      "anndata             0.7.6\n",
      "anyio               NA\n",
      "attr                21.2.0\n",
      "babel               2.9.1\n",
      "backcall            0.2.0\n",
      "beta_ufunc          NA\n",
      "binom_ufunc         NA\n",
      "brotli              NA\n",
      "cairo               1.20.1\n",
      "certifi             2021.05.30\n",
      "cffi                1.14.4\n",
      "chardet             4.0.0\n",
      "colorama            0.4.4\n",
      "cycler              0.10.0\n",
      "cython_runtime      NA\n",
      "dateutil            2.8.1\n",
      "debugpy             1.3.0\n",
      "decorator           5.0.9\n",
      "defusedxml          0.7.1\n",
      "dunamai             1.5.5\n",
      "get_version         3.2\n",
      "google              NA\n",
      "h5py                3.3.0\n",
      "idna                2.10\n",
      "igraph              0.8.3\n",
      "ipykernel           6.0.0\n",
      "ipython_genutils    0.2.0\n",
      "ipywidgets          7.6.3\n",
      "jedi                0.18.0\n",
      "jinja2              3.0.1\n",
      "joblib              1.0.1\n",
      "json5               NA\n",
      "jsonschema          3.2.0\n",
      "jupyter_server      1.9.0\n",
      "jupyterlab_server   2.6.0\n",
      "kiwisolver          1.3.1\n",
      "legacy_api_wrap     0.0.0\n",
      "leidenalg           0.8.3\n",
      "llvmlite            0.34.0\n",
      "louvain             0.7.0\n",
      "markupsafe          2.0.1\n",
      "matplotlib          3.4.2\n",
      "matplotlib_inline   NA\n",
      "mpl_toolkits        NA\n",
      "natsort             7.1.1\n",
      "nbclassic           NA\n",
      "nbformat            5.1.3\n",
      "nbinom_ufunc        NA\n",
      "numba               0.51.2\n",
      "numexpr             2.7.3\n",
      "numpy               1.19.5\n",
      "packaging           20.9\n",
      "pandas              1.2.4\n",
      "parso               0.8.2\n",
      "pexpect             4.8.0\n",
      "pickleshare         0.7.5\n",
      "pkg_resources       NA\n",
      "prometheus_client   NA\n",
      "prompt_toolkit      3.0.19\n",
      "ptyprocess          0.7.0\n",
      "pvectorc            NA\n",
      "pygments            2.9.0\n",
      "pyparsing           2.4.7\n",
      "pyrsistent          NA\n",
      "pytz                2021.1\n",
      "requests            2.25.1\n",
      "scanpy              1.7.0\n",
      "scipy               1.7.0\n",
      "seaborn             0.11.1\n",
      "send2trash          NA\n",
      "setuptools_scm      NA\n",
      "sinfo               0.3.1\n",
      "six                 1.16.0\n",
      "sklearn             0.24.2\n",
      "sniffio             1.2.0\n",
      "socks               1.7.1\n",
      "sphinxcontrib       NA\n",
      "statsmodels         0.12.2\n",
      "storemagic          NA\n",
      "tables              3.6.1\n",
      "terminado           0.10.1\n",
      "texttable           1.6.3\n",
      "tornado             6.1\n",
      "traitlets           5.0.5\n",
      "typing_extensions   NA\n",
      "urllib3             1.26.6\n",
      "wcwidth             0.2.5\n",
      "websocket           0.57.0\n",
      "yaml                5.4.1\n",
      "zmq                 22.1.0\n",
      "-----\n",
      "IPython             7.25.0\n",
      "jupyter_client      6.1.12\n",
      "jupyter_core        4.7.1\n",
      "jupyterlab          3.0.16\n",
      "notebook            6.4.0\n",
      "-----\n",
      "Python 3.8.0 | packaged by conda-forge | (default, Nov 22 2019, 19:11:38) [GCC 7.3.0]\n",
      "Linux-4.15.0-219-generic-x86_64-with-glibc2.10\n",
      "64 logical CPU cores, x86_64\n",
      "-----\n",
      "Session information updated at 2023-11-27 19:01\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_versions()\n",
    "sc.settings.set_figure_params(dpi=80)  # low dpi (dots per inch) yields small inline figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "chronic-dividend",
   "metadata": {},
   "outputs": [],
   "source": [
    "for folder in ['../results/log_regression/']:\n",
    "    if not os.path.exists(folder):\n",
    "        os.makedirs(folder)\n",
    "        #loop over the subfolders of the rawdata folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "front-visiting",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_path= '../results/log_regression/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "heard-combat",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_MFIRef = sc.read('../../24h/data/adata_raw_MFIdata_annotated.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dressed-renaissance",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>sample</th>\n",
       "      <th>technology</th>\n",
       "      <th>tissue</th>\n",
       "      <th>dev_age</th>\n",
       "      <th>donor</th>\n",
       "      <th>dataset</th>\n",
       "      <th>run</th>\n",
       "      <th>number_of_individuals_multiplexed</th>\n",
       "      <th>batch</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>scrublet_score</th>\n",
       "      <th>scrublet_cluster_score</th>\n",
       "      <th>bh_pval</th>\n",
       "      <th>is_doublet</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode_sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACGGGCATTGGCGC-1_FCA7167219</th>\n",
       "      <td>1941</td>\n",
       "      <td>FCA7167219</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>decidua_immune</td>\n",
       "      <td>13_PCW</td>\n",
       "      <td>6</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.090287</td>\n",
       "      <td>5992.0</td>\n",
       "      <td>0.164948</td>\n",
       "      <td>0.112619</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>False</td>\n",
       "      <td>dNK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGTCGCGATCG-1_FCA7167219</th>\n",
       "      <td>1246</td>\n",
       "      <td>FCA7167219</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>decidua_immune</td>\n",
       "      <td>13_PCW</td>\n",
       "      <td>6</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.010213</td>\n",
       "      <td>3525.0</td>\n",
       "      <td>0.050251</td>\n",
       "      <td>0.069505</td>\n",
       "      <td>0.905171</td>\n",
       "      <td>False</td>\n",
       "      <td>dT_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGATGAGCAATATG-1_FCA7167219</th>\n",
       "      <td>1516</td>\n",
       "      <td>FCA7167219</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>decidua_immune</td>\n",
       "      <td>13_PCW</td>\n",
       "      <td>6</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.020889</td>\n",
       "      <td>4452.0</td>\n",
       "      <td>0.096774</td>\n",
       "      <td>0.087838</td>\n",
       "      <td>0.905171</td>\n",
       "      <td>False</td>\n",
       "      <td>T_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGATGAGTTCGCGC-1_FCA7167219</th>\n",
       "      <td>2099</td>\n",
       "      <td>FCA7167219</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>decidua_immune</td>\n",
       "      <td>13_PCW</td>\n",
       "      <td>6</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.047764</td>\n",
       "      <td>6888.0</td>\n",
       "      <td>0.096774</td>\n",
       "      <td>0.112619</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>False</td>\n",
       "      <td>dNK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGATGCATGTCGAT-1_FCA7167219</th>\n",
       "      <td>1636</td>\n",
       "      <td>FCA7167219</td>\n",
       "      <td>10X_3'_scRNA-seq</td>\n",
       "      <td>decidua_immune</td>\n",
       "      <td>13_PCW</td>\n",
       "      <td>6</td>\n",
       "      <td>Vento_Nature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.024991</td>\n",
       "      <td>5882.0</td>\n",
       "      <td>0.087838</td>\n",
       "      <td>0.072727</td>\n",
       "      <td>0.905171</td>\n",
       "      <td>False</td>\n",
       "      <td>T_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCGTCAAGT-1_Pla_Camb10714920</th>\n",
       "      <td>2409</td>\n",
       "      <td>Pla_Camb10714920</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>decidua_placenta_myometrium</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>P13</td>\n",
       "      <td>Ashley_collection_sn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>0.007270</td>\n",
       "      <td>5227.0</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>0.014381</td>\n",
       "      <td>0.999631</td>\n",
       "      <td>False</td>\n",
       "      <td>uSMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCTTAGTCT-1_Pla_Camb10714920</th>\n",
       "      <td>4129</td>\n",
       "      <td>Pla_Camb10714920</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>decidua_placenta_myometrium</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>P13</td>\n",
       "      <td>Ashley_collection_sn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>17379.0</td>\n",
       "      <td>0.102190</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.886380</td>\n",
       "      <td>False</td>\n",
       "      <td>SCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTCACAGCG-1_Pla_Camb10714920</th>\n",
       "      <td>3116</td>\n",
       "      <td>Pla_Camb10714920</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>decidua_placenta_myometrium</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>P13</td>\n",
       "      <td>Ashley_collection_sn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>0.004499</td>\n",
       "      <td>7779.0</td>\n",
       "      <td>0.036876</td>\n",
       "      <td>0.038067</td>\n",
       "      <td>0.999631</td>\n",
       "      <td>False</td>\n",
       "      <td>HOFB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTACTTG-1_Pla_Camb10714920</th>\n",
       "      <td>3696</td>\n",
       "      <td>Pla_Camb10714920</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>decidua_placenta_myometrium</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>P13</td>\n",
       "      <td>Ashley_collection_sn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>0.001992</td>\n",
       "      <td>20078.0</td>\n",
       "      <td>0.109137</td>\n",
       "      <td>0.116711</td>\n",
       "      <td>0.886380</td>\n",
       "      <td>False</td>\n",
       "      <td>SCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGAGCA-1_Pla_Camb10714920</th>\n",
       "      <td>2130</td>\n",
       "      <td>Pla_Camb10714920</td>\n",
       "      <td>10X_multiome</td>\n",
       "      <td>decidua_placenta_myometrium</td>\n",
       "      <td>8-9_PCW</td>\n",
       "      <td>P13</td>\n",
       "      <td>Ashley_collection_sn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>0.012930</td>\n",
       "      <td>5259.0</td>\n",
       "      <td>0.109137</td>\n",
       "      <td>0.114776</td>\n",
       "      <td>0.886380</td>\n",
       "      <td>False</td>\n",
       "      <td>SCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>325665 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     n_genes            sample  \\\n",
       "barcode_sample                                                   \n",
       "AAACGGGCATTGGCGC-1_FCA7167219           1941        FCA7167219   \n",
       "AAACGGGTCGCGATCG-1_FCA7167219           1246        FCA7167219   \n",
       "AAAGATGAGCAATATG-1_FCA7167219           1516        FCA7167219   \n",
       "AAAGATGAGTTCGCGC-1_FCA7167219           2099        FCA7167219   \n",
       "AAAGATGCATGTCGAT-1_FCA7167219           1636        FCA7167219   \n",
       "...                                      ...               ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920     2409  Pla_Camb10714920   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920     4129  Pla_Camb10714920   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920     3116  Pla_Camb10714920   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920     3696  Pla_Camb10714920   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920     2130  Pla_Camb10714920   \n",
       "\n",
       "                                           technology  \\\n",
       "barcode_sample                                          \n",
       "AAACGGGCATTGGCGC-1_FCA7167219        10X_3'_scRNA-seq   \n",
       "AAACGGGTCGCGATCG-1_FCA7167219        10X_3'_scRNA-seq   \n",
       "AAAGATGAGCAATATG-1_FCA7167219        10X_3'_scRNA-seq   \n",
       "AAAGATGAGTTCGCGC-1_FCA7167219        10X_3'_scRNA-seq   \n",
       "AAAGATGCATGTCGAT-1_FCA7167219        10X_3'_scRNA-seq   \n",
       "...                                               ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920      10X_multiome   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920      10X_multiome   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920      10X_multiome   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920      10X_multiome   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920      10X_multiome   \n",
       "\n",
       "                                                          tissue  dev_age  \\\n",
       "barcode_sample                                                              \n",
       "AAACGGGCATTGGCGC-1_FCA7167219                     decidua_immune   13_PCW   \n",
       "AAACGGGTCGCGATCG-1_FCA7167219                     decidua_immune   13_PCW   \n",
       "AAAGATGAGCAATATG-1_FCA7167219                     decidua_immune   13_PCW   \n",
       "AAAGATGAGTTCGCGC-1_FCA7167219                     decidua_immune   13_PCW   \n",
       "AAAGATGCATGTCGAT-1_FCA7167219                     decidua_immune   13_PCW   \n",
       "...                                                          ...      ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920  decidua_placenta_myometrium  8-9_PCW   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920  decidua_placenta_myometrium  8-9_PCW   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920  decidua_placenta_myometrium  8-9_PCW   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920  decidua_placenta_myometrium  8-9_PCW   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920  decidua_placenta_myometrium  8-9_PCW   \n",
       "\n",
       "                                    donor               dataset  run  \\\n",
       "barcode_sample                                                         \n",
       "AAACGGGCATTGGCGC-1_FCA7167219           6          Vento_Nature  NaN   \n",
       "AAACGGGTCGCGATCG-1_FCA7167219           6          Vento_Nature  NaN   \n",
       "AAAGATGAGCAATATG-1_FCA7167219           6          Vento_Nature  NaN   \n",
       "AAAGATGAGTTCGCGC-1_FCA7167219           6          Vento_Nature  NaN   \n",
       "AAAGATGCATGTCGAT-1_FCA7167219           6          Vento_Nature  NaN   \n",
       "...                                   ...                   ...  ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920   P13  Ashley_collection_sn  NaN   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920   P13  Ashley_collection_sn  NaN   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920   P13  Ashley_collection_sn  NaN   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920   P13  Ashley_collection_sn  NaN   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920   P13  Ashley_collection_sn  NaN   \n",
       "\n",
       "                                     number_of_individuals_multiplexed batch  \\\n",
       "barcode_sample                                                                 \n",
       "AAACGGGCATTGGCGC-1_FCA7167219                                        1     0   \n",
       "AAACGGGTCGCGATCG-1_FCA7167219                                        1     0   \n",
       "AAAGATGAGCAATATG-1_FCA7167219                                        1     0   \n",
       "AAAGATGAGTTCGCGC-1_FCA7167219                                        1     0   \n",
       "AAAGATGCATGTCGAT-1_FCA7167219                                        1     0   \n",
       "...                                                                ...   ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920                                  1    61   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920                                  1    61   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920                                  1    61   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920                                  1    61   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920                                  1    61   \n",
       "\n",
       "                                     percent_mito  n_counts  scrublet_score  \\\n",
       "barcode_sample                                                                \n",
       "AAACGGGCATTGGCGC-1_FCA7167219            0.090287    5992.0        0.164948   \n",
       "AAACGGGTCGCGATCG-1_FCA7167219            0.010213    3525.0        0.050251   \n",
       "AAAGATGAGCAATATG-1_FCA7167219            0.020889    4452.0        0.096774   \n",
       "AAAGATGAGTTCGCGC-1_FCA7167219            0.047764    6888.0        0.096774   \n",
       "AAAGATGCATGTCGAT-1_FCA7167219            0.024991    5882.0        0.087838   \n",
       "...                                           ...       ...             ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920      0.007270    5227.0        0.013699   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920      0.002532   17379.0        0.102190   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920      0.004499    7779.0        0.036876   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920      0.001992   20078.0        0.109137   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920      0.012930    5259.0        0.109137   \n",
       "\n",
       "                                     scrublet_cluster_score   bh_pval  \\\n",
       "barcode_sample                                                          \n",
       "AAACGGGCATTGGCGC-1_FCA7167219                      0.112619  0.857143   \n",
       "AAACGGGTCGCGATCG-1_FCA7167219                      0.069505  0.905171   \n",
       "AAAGATGAGCAATATG-1_FCA7167219                      0.087838  0.905171   \n",
       "AAAGATGAGTTCGCGC-1_FCA7167219                      0.112619  0.857143   \n",
       "AAAGATGCATGTCGAT-1_FCA7167219                      0.072727  0.905171   \n",
       "...                                                     ...       ...   \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920                0.014381  0.999631   \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920                0.125000  0.886380   \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920                0.038067  0.999631   \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920                0.116711  0.886380   \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920                0.114776  0.886380   \n",
       "\n",
       "                                     is_doublet cell_type  \n",
       "barcode_sample                                             \n",
       "AAACGGGCATTGGCGC-1_FCA7167219             False      dNK1  \n",
       "AAACGGGTCGCGATCG-1_FCA7167219             False  dT_cells  \n",
       "AAAGATGAGCAATATG-1_FCA7167219             False   T_cells  \n",
       "AAAGATGAGTTCGCGC-1_FCA7167219             False      dNK1  \n",
       "AAAGATGCATGTCGAT-1_FCA7167219             False   T_cells  \n",
       "...                                         ...       ...  \n",
       "TTTGTGTTCGTCAAGT-1_Pla_Camb10714920       False      uSMC  \n",
       "TTTGTGTTCTTAGTCT-1_Pla_Camb10714920       False       SCT  \n",
       "TTTGTTGGTCACAGCG-1_Pla_Camb10714920       False      HOFB  \n",
       "TTTGTTGGTTTACTTG-1_Pla_Camb10714920       False       SCT  \n",
       "TTTGTTGGTTTGAGCA-1_Pla_Camb10714920       False       SCT  \n",
       "\n",
       "[325665 rows x 17 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#reading original object with raw counts\n",
    "#adata_MFIRef=sc.read('../../placental_explant/starsolo/results/scVI/S9_scVIintegrated_fetalSurgSubset_adata_latent20_Raw_20220523.h5ad')\n",
    "\n",
    "adata_MFIRef.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "usual-injury",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<anndata._core.raw.Raw at 0x147a9d180880>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_MFIRef.raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "heard-childhood",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my test data\n",
    "adata_explants = sc.read('../results/scVI/rna3_scVIintegrated_All_latent20.h5ad')\n",
    "adata_explants.X = adata_explants.layers['raw_counts'].copy() #Setting the raw_counts in .X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "broken-pepper",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 119146 × 36601\n",
       "    obs: 'sample', 'stage', 'donor', 'hpi', 'infection', 'percent_mito', 'n_counts', 'batch', 'sample_barcode', 'assignment_SoC', 'donor_id', 'scrublet_score', 'scrublet_cluster_score', 'zscore', 'bh_pval', 'bonf_pval', 'is_doublet', 'infection_stage', 'S_score', 'G2M_score', 'phase', 'n_genes_by_counts', 'total_counts', 'total_counts_hs', 'pct_counts_hs', 'total_counts_tg', 'pct_counts_tg', 'Tg_infected', 'n_genes', '_scvi_batch', '_scvi_labels', '_scvi_local_l_mean', '_scvi_local_l_var', 'leiden_scvi'\n",
       "    var: 'gene_ids', 'feature_types', 'mean-0', 'std-0', 'mean-1', 'std-1', 'mean-2', 'std-2', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm'\n",
       "    uns: '_scvi', 'donor_id_colors', 'hpi_colors', 'hvg', 'infection_colors', 'leiden', 'leiden_scvi_colors', 'neighbors', 'sample_colors', 'umap'\n",
       "    obsm: 'X_scVI_n_latent_20', 'X_scVI_n_latent_30', 'X_umap', '_scvi_extra_categoricals'\n",
       "    layers: 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_explants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "compact-trigger",
   "metadata": {},
   "source": [
    "# LOGISTIC REGRESSION\n",
    "\n",
    "### Subsetting both datasets' copies to common genes only\n",
    "\n",
    "Can be either overlap of all genes that passed QC or HVG overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "affected-stockholm",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids-0</th>\n",
       "      <th>feature_types-0</th>\n",
       "      <th>genome-0</th>\n",
       "      <th>n_cells-0</th>\n",
       "      <th>gene_ids-1</th>\n",
       "      <th>feature_types-1</th>\n",
       "      <th>genome-1</th>\n",
       "      <th>n_cells-1</th>\n",
       "      <th>gene_ids-10</th>\n",
       "      <th>feature_types-10</th>\n",
       "      <th>...</th>\n",
       "      <th>genome-7</th>\n",
       "      <th>n_cells-7</th>\n",
       "      <th>gene_ids-8</th>\n",
       "      <th>feature_types-8</th>\n",
       "      <th>genome-8</th>\n",
       "      <th>n_cells-8</th>\n",
       "      <th>gene_ids-9</th>\n",
       "      <th>feature_types-9</th>\n",
       "      <th>genome-9</th>\n",
       "      <th>n_cells-9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>58.0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>118.0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>140.0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>185.0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>608.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>21.0</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>14.0</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>20.0</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>91.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>66.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>171.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>2117.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>59.0</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>61.0</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>11.0</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>21.0</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>26.0</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>32.0</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>122.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>15.0</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>41.0</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>22.0</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>47.0</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>183.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>134.0</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>284.0</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>265.0</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>319.0</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>2351.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>41.0</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>88.0</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>58.0</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>46.0</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>299.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30800 rows × 248 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_ids-0  feature_types-0 genome-0  n_cells-0  \\\n",
       "A1BG      ENSG00000121410  Gene Expression   GRCh38       58.0   \n",
       "A1BG-AS1  ENSG00000268895  Gene Expression   GRCh38        7.0   \n",
       "A1CF                  nan              nan      nan        NaN   \n",
       "A2M       ENSG00000175899  Gene Expression   GRCh38       66.0   \n",
       "A2M-AS1   ENSG00000245105  Gene Expression   GRCh38        6.0   \n",
       "...                   ...              ...      ...        ...   \n",
       "ZXDC      ENSG00000070476  Gene Expression   GRCh38       11.0   \n",
       "ZYG11A                nan              nan      nan        NaN   \n",
       "ZYG11B    ENSG00000162378  Gene Expression   GRCh38       15.0   \n",
       "ZYX       ENSG00000159840  Gene Expression   GRCh38      134.0   \n",
       "ZZEF1     ENSG00000074755  Gene Expression   GRCh38       41.0   \n",
       "\n",
       "               gene_ids-1  feature_types-1 genome-1  n_cells-1  \\\n",
       "A1BG      ENSG00000121410  Gene Expression   GRCh38      118.0   \n",
       "A1BG-AS1  ENSG00000268895  Gene Expression   GRCh38       21.0   \n",
       "A1CF                  nan              nan      nan        NaN   \n",
       "A2M       ENSG00000175899  Gene Expression   GRCh38      171.0   \n",
       "A2M-AS1   ENSG00000245105  Gene Expression   GRCh38        5.0   \n",
       "...                   ...              ...      ...        ...   \n",
       "ZXDC      ENSG00000070476  Gene Expression   GRCh38       21.0   \n",
       "ZYG11A                nan              nan      nan        NaN   \n",
       "ZYG11B    ENSG00000162378  Gene Expression   GRCh38       41.0   \n",
       "ZYX       ENSG00000159840  Gene Expression   GRCh38      284.0   \n",
       "ZZEF1     ENSG00000074755  Gene Expression   GRCh38       88.0   \n",
       "\n",
       "              gene_ids-10 feature_types-10  ... genome-7  n_cells-7  \\\n",
       "A1BG      ENSG00000121410  Gene Expression  ...   GRCh38      140.0   \n",
       "A1BG-AS1  ENSG00000268895  Gene Expression  ...   GRCh38       14.0   \n",
       "A1CF                  nan              nan  ...      nan        NaN   \n",
       "A2M       ENSG00000175899  Gene Expression  ...      nan        NaN   \n",
       "A2M-AS1   ENSG00000245105  Gene Expression  ...   GRCh38       59.0   \n",
       "...                   ...              ...  ...      ...        ...   \n",
       "ZXDC      ENSG00000070476  Gene Expression  ...   GRCh38       26.0   \n",
       "ZYG11A                nan              nan  ...      nan        NaN   \n",
       "ZYG11B    ENSG00000162378  Gene Expression  ...   GRCh38       22.0   \n",
       "ZYX       ENSG00000159840  Gene Expression  ...   GRCh38      265.0   \n",
       "ZZEF1     ENSG00000074755  Gene Expression  ...   GRCh38       58.0   \n",
       "\n",
       "               gene_ids-8  feature_types-8 genome-8  n_cells-8  \\\n",
       "A1BG      ENSG00000121410  Gene Expression   GRCh38      185.0   \n",
       "A1BG-AS1  ENSG00000268895  Gene Expression   GRCh38       20.0   \n",
       "A1CF                  nan              nan      nan        NaN   \n",
       "A2M       ENSG00000175899  Gene Expression   GRCh38        5.0   \n",
       "A2M-AS1   ENSG00000245105  Gene Expression   GRCh38       61.0   \n",
       "...                   ...              ...      ...        ...   \n",
       "ZXDC      ENSG00000070476  Gene Expression   GRCh38       32.0   \n",
       "ZYG11A                nan              nan      nan        NaN   \n",
       "ZYG11B    ENSG00000162378  Gene Expression   GRCh38       47.0   \n",
       "ZYX       ENSG00000159840  Gene Expression   GRCh38      319.0   \n",
       "ZZEF1     ENSG00000074755  Gene Expression   GRCh38       46.0   \n",
       "\n",
       "               gene_ids-9  feature_types-9 genome-9  n_cells-9  \n",
       "A1BG      ENSG00000121410  Gene Expression   GRCh38      608.0  \n",
       "A1BG-AS1  ENSG00000268895  Gene Expression   GRCh38       91.0  \n",
       "A1CF                  nan              nan      nan        NaN  \n",
       "A2M       ENSG00000175899  Gene Expression   GRCh38     2117.0  \n",
       "A2M-AS1   ENSG00000245105  Gene Expression   GRCh38       13.0  \n",
       "...                   ...              ...      ...        ...  \n",
       "ZXDC      ENSG00000070476  Gene Expression   GRCh38      122.0  \n",
       "ZYG11A                nan              nan      nan        NaN  \n",
       "ZYG11B    ENSG00000162378  Gene Expression   GRCh38      183.0  \n",
       "ZYX       ENSG00000159840  Gene Expression   GRCh38     2351.0  \n",
       "ZZEF1     ENSG00000074755  Gene Expression   GRCh38      299.0  \n",
       "\n",
       "[30800 rows x 248 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_MFIRef.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "associate-tradition",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "If you pass `n_top_genes`, all cutoffs are ignored.\n",
      "extracting highly variable genes\n"
     ]
    }
   ],
   "source": [
    "sc.pp.highly_variable_genes(adata_MFIRef, n_top_genes=3000, flavor='seurat_v3')\n",
    "sc.pl.highly_variable_genes(adata_MFIRef)\n",
    "\n",
    "#saving the highl variable genes\n",
    "highly_variable_genes = adata_MFIRef.var[\"highly_variable\"]\n",
    "adata_MFIRef_downsampled = adata_MFIRef[:, highly_variable_genes]\n",
    "\n",
    "print('Total number of cells: {:d}'.format(adata_MFIRef_downsampled.n_obs))\n",
    "print('Total number of genes: {:d}'.format(adata_MFIRef_downsampled.n_vars))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "analyzed-summit",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reading the HVG annotation done in S1\n",
    "#highly_variable_genes= pd.read_csv('../results/exp_infection/tables/highlyVariable_genes.csv', index_col=0)\n",
    "\n",
    "#highly_variable_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "august-locator",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(adata_explants, n_top_genes=3000, flavor='seurat_v3')\n",
    "sc.pl.highly_variable_genes(adata_explants)\n",
    "\n",
    "#saving the highl variable genes\n",
    "highly_variable_genes = adata_explants.var[\"highly_variable\"]\n",
    "adata_hvg = adata_explants[:, highly_variable_genes]\n",
    "\n",
    "print('Total number of cells: {:d}'.format(adata_hvg.n_obs))\n",
    "print('Total number of genes: {:d}'.format(adata_hvg.n_vars))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "taken-palmer",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hidden-report",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_hvg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "objective-negotiation",
   "metadata": {},
   "outputs": [],
   "source": [
    "### here using only HVG overlap\n",
    "###\n",
    "\n",
    "# normlog, scaled and HVG subsetted \n",
    "adata_train = adata_MFIRef_downsampled.copy()\n",
    "\n",
    "# normlog, scaled and HVG subsetted \n",
    "adata_test = adata_hvg.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "expanded-amplifier",
   "metadata": {},
   "outputs": [],
   "source": [
    "del(adata_hvg)\n",
    "del(adata_MFIRef_downsampled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aware-bread",
   "metadata": {},
   "outputs": [],
   "source": [
    "#normlog and scaling Training and Test sets\n",
    "sc.pp.normalize_total(adata_train, target_sum=1e4)\n",
    "sc.pp.log1p(adata_train)\n",
    "sc.pp.scale(adata_train, max_value=10)\n",
    "\n",
    "\n",
    "sc.pp.normalize_total(adata_test, target_sum=1e4)\n",
    "sc.pp.log1p(adata_test)\n",
    "sc.pp.scale(adata_test, max_value=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "foreign-spell",
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of genes shared between blood and bone marrow data\n",
    "# doing sorted on the overall list is VERY important for consistent ordering of the genes\n",
    "common_genes = sorted(list(set(adata_train.var_names) & set(adata_test.var_names)))\n",
    "print(len(common_genes),\"genes are shared between annotated train and our joint data \\n\",\n",
    "      len(set(adata_train.var_names)), 'genes in training data \\n',\n",
    "      len((set(adata_test.var_names))), 'genes in test joint data \\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "outer-precipitation",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Subseting the datasets\n",
    "adata_train_copy = adata_train[:,common_genes]\n",
    "adata_train_copy.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "accessory-thought",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Subseting the datasets\n",
    "adata_test_copy = adata_test[:,common_genes]\n",
    "adata_test_copy.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "taken-investor",
   "metadata": {},
   "outputs": [],
   "source": [
    "# checking gene correspondence - important!\n",
    "list(adata_test_copy.var_names) == list(adata_train_copy.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "suburban-scanning",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_train_copy.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "instrumental-announcement",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.unique(adata_train_copy.obs['cell_type'],return_counts=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "natural-river",
   "metadata": {},
   "source": [
    "# Training a logistic regression model on previously annotated 10X data Arutyunyan, et al. 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "personal-hudson",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_train_copy.obs['cell_type'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sensitive-oriental",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltypes_train = adata_train_copy.obs['cell_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lovely-uniform",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_train_copy.X.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mathematical-membership",
   "metadata": {},
   "source": [
    "## The following cell actually trains the model and may require quite a bit of time if you are using a big training dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecological-tactics",
   "metadata": {},
   "outputs": [],
   "source": [
    "#11:23\n",
    "%time\n",
    "# training a logistic regression model on scRNA-seq data\n",
    "# takes about 1.5hrs for HVGs intersection\n",
    "logisticRegr_sc = LogisticRegression(max_iter = 5000, n_jobs = -1, random_state = 0, C=0.2)\n",
    "logisticRegr_sc.fit(adata_train_copy.X, celltypes_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "northern-concentration",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Saving model\n",
    "\n",
    "#pkl_filename = \"./logreg_from_P13_downsampled_no_soupy_genes_coarse_troph_to_all_troph_organoids_premrna_HVGs.pkl\"\n",
    "pkl_filename = \"../results/log_regression/logreg_MFIref_InfectedExplants_HVGs_20230707_48h.pkl\"\n",
    "\n",
    "\n",
    "\n",
    "with open(pkl_filename, 'wb') as file:  \n",
    "    pickle.dump(logisticRegr_sc, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "vertical-charles",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('YAY')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "confidential-howard",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load model (scRNA-seq) from file\n",
    "\n",
    "# which one?\n",
    "pkl_filename = \"../results/log_regression/logreg_MFIref_InfectedExplants_HVGs_20230707_48h.pkl\"\n",
    "\n",
    "with open(pkl_filename, 'rb') as file:  \n",
    "    logisticRegr_sc = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "classified-neighbor",
   "metadata": {},
   "outputs": [],
   "source": [
    "logisticRegr_sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "naked-rings",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "predictions = logisticRegr_sc.predict(adata_test_copy.X)\n",
    "adata_test_copy.obs['celltype_predictions'] = predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "excess-egypt",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# probabilities\n",
    "probabilities = logisticRegr_sc.predict_proba(adata_test_copy.X)\n",
    "probs_adata = []\n",
    "for i,clus in enumerate(adata_test_copy.obs['celltype_predictions']):\n",
    "    probs_adata.append(probabilities[i,logisticRegr_sc.classes_==clus][0])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adult-serum",
   "metadata": {},
   "outputs": [],
   "source": [
    "#list(adata_test_copy.obs.index) == list(adata_hvg.obs.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "accredited-lesbian",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_test_copy.obs['celltype_predictions']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "present-sentence",
   "metadata": {},
   "outputs": [],
   "source": [
    "# now you can return to your object transferring celltype predictions and their probabilities from training\n",
    "\n",
    "# my test object\n",
    "adata_explants.obs['celltype_predictions'] = adata_test_copy.obs.loc[adata_explants.obs_names,'celltype_predictions']\n",
    "adata_explants.obs['probabilities'] = probs_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "broad-muslim",
   "metadata": {},
   "outputs": [],
   "source": [
    "# you may want to save results at this point\n",
    "adata_explants.obs.loc[:,['celltype_predictions','probabilities']].to_csv('../results/log_regression/logreg_MFIref_InfectedExplants_HVGs_20230707_48h.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "filled-congress",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
