{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "underlying-drove",
   "metadata": {},
   "source": [
    "**Author:** Elias Rafael Ruiz-Morales\n",
    "\n",
    "**Institution:** Wellcome Sanger institute\n",
    "\n",
    "**July, 2023**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "altered-asbestos",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "behind-chapter",
   "metadata": {},
   "source": [
    "## Generate cellphoneDB input files (1) _meta.tsv and (2) _degs.tsv"
   ]
  },
  {
   "cell_type": "raw",
   "id": "completed-burner",
   "metadata": {},
   "source": [
    "reference: https://github.com/ventolab/CellphoneDB/blob/master/notebooks/0_prepare_your_data_from_anndata.ipynb\n",
    "reference: https://github.com/ventolab/HGDA/blob/main/human/scRNAseq/cellphoneDB/cellphoneDB1_granulosa_preparefiles.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "forbidden-strike",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/opt/conda/bin/python'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import os\n",
    "import sys\n",
    "sys.executable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "contained-hartford",
   "metadata": {},
   "outputs": [],
   "source": [
    "def recoverDEGs(DEGdf, infection, celltype, direction):\n",
    "\n",
    "    \n",
    "    ### Up reg\n",
    "    ###\n",
    "    if direction=='up':\n",
    "\n",
    "        #selecting only up regulated genes\n",
    "        tempUP = DEGdf[DEGdf['logFC'] > 0].copy()\n",
    "        \n",
    "        #making new columns\n",
    "        tempUP['cell_type'] = celltype\n",
    "        tempUP['cluster'] = celltype + '.' + infection\n",
    "        tempUP['gene']=list(tempUP.index.values)\n",
    "\n",
    "        # select significant less restrictive\n",
    "        tempUP =tempUP[ tempUP.p_val_adj < 0.05]\n",
    "        tempUP = tempUP[ tempUP.logFC > 0.1]\n",
    "\n",
    "        \n",
    "        return(tempUP)\n",
    "\n",
    "\n",
    "    ### Down reg\n",
    "    ###\n",
    "    if direction=='down':\n",
    "        #selecting only down regulated genes\n",
    "        tempUP = DEGdf[DEGdf['logFC'] < 0].copy()\n",
    "        \n",
    "        #making new columns\n",
    "        tempUP['cell_type'] = celltype\n",
    "        tempUP['cluster'] = celltype + '.' + infection\n",
    "        tempUP['gene']=list(tempUP.index.values)\n",
    "\n",
    "\n",
    "        # select significant less restrictive\n",
    "        tempUP =tempUP[ tempUP.p_val_adj < 0.05]\n",
    "        tempUP = tempUP[ tempUP.logFC < -0.1]\n",
    "        \n",
    "        return(tempDwn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "portable-tiger",
   "metadata": {},
   "source": [
    "## Prepare INPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "maritime-track",
   "metadata": {},
   "source": [
    "### (1) Load annadata_ single-cell explants from malaria infection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "increasing-trailer",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read('../../data_integration/results/scVI/malaria_singleCell_24h.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "minute-carroll",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>freqs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>categories</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EVT_1</th>\n",
       "      <td>301</td>\n",
       "      <td>0.006171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EVT_2</th>\n",
       "      <td>701</td>\n",
       "      <td>0.014372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Endo_f</th>\n",
       "      <td>5788</td>\n",
       "      <td>0.118667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>9059</td>\n",
       "      <td>0.185730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F_p</th>\n",
       "      <td>566</td>\n",
       "      <td>0.011604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F_sm</th>\n",
       "      <td>483</td>\n",
       "      <td>0.009903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HBC</th>\n",
       "      <td>9000</td>\n",
       "      <td>0.184521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HBC_p</th>\n",
       "      <td>11</td>\n",
       "      <td>0.000226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PAMM1</th>\n",
       "      <td>2217</td>\n",
       "      <td>0.045454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PV</th>\n",
       "      <td>5606</td>\n",
       "      <td>0.114936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT</th>\n",
       "      <td>1597</td>\n",
       "      <td>0.032742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_CCC</th>\n",
       "      <td>1411</td>\n",
       "      <td>0.028929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_fusing</th>\n",
       "      <td>9172</td>\n",
       "      <td>0.188047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_p</th>\n",
       "      <td>2147</td>\n",
       "      <td>0.044018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iEVT</th>\n",
       "      <td>716</td>\n",
       "      <td>0.014680</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            counts     freqs\n",
       "categories                  \n",
       "EVT_1          301  0.006171\n",
       "EVT_2          701  0.014372\n",
       "Endo_f        5788  0.118667\n",
       "F             9059  0.185730\n",
       "F_p            566  0.011604\n",
       "F_sm           483  0.009903\n",
       "HBC           9000  0.184521\n",
       "HBC_p           11  0.000226\n",
       "PAMM1         2217  0.045454\n",
       "PV            5606  0.114936\n",
       "VCT           1597  0.032742\n",
       "VCT_CCC       1411  0.028929\n",
       "VCT_fusing    9172  0.188047\n",
       "VCT_p         2147  0.044018\n",
       "iEVT           716  0.014680"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['cell_type'].values.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "respective-singapore",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>freqs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>categories</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Pf_24h</th>\n",
       "      <td>27100</td>\n",
       "      <td>0.555613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UI_Pf_24h</th>\n",
       "      <td>21675</td>\n",
       "      <td>0.444387</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            counts     freqs\n",
       "categories                  \n",
       "Pf_24h       27100  0.555613\n",
       "UI_Pf_24h    21675  0.444387"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['stage_perInfection'].values.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "finished-listing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 48775 × 36601\n",
       "    obs: 'sample', 'stage', 'hpi', 'infection', 'percent_mito', 'n_counts', 'sample_barcode', 'assignment_SoC', 'donor_id', 'scrublet_score', 'scrublet_cluster_score', 'zscore', 'bh_pval', 'bonf_pval', 'S_score', 'G2M_score', 'phase', 'n_genes_by_counts', 'total_counts', 'total_counts_hs', 'pct_counts_hs', 'total_counts_tg', 'pct_counts_tg', 'Tg_infected', 'n_genes', '_scvi_batch', '_scvi_labels', '_scvi_local_l_mean', '_scvi_local_l_var', 'leiden_scvi', 'celltype_predictions', 'probabilities', 'scrublet_doublet', 'cell_type_2022', 'cell_type', 'souporcell_MFgenotype', 'MFgenotype', 'cell_type_broad', 'umap_density_Tg_infected', 'stage_perInfection', 'celltype-Stage', 'Tg_intracellular', 'celltype-Intracellular', 'Dev_Stage'\n",
       "    var: 'gene_ids', 'feature_types', 'mean-0', 'std-0', 'mean-1', 'std-1', 'mean-2', 'std-2', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm'\n",
       "    uns: 'Dev_Stage_colors', 'MFgenotype_colors', '_scvi', 'cell_type_2022_colors', 'cell_type_broad_colors', 'cell_type_colors', 'celltype_predictions_colors', 'donor_id_colors', 'hpi_colors', 'hvg', 'infection_colors', 'leiden', 'neighbors', 'sample_colors', 'scrublet_doublet_colors', 'umap', 'umap_density_Tg_infected_params'\n",
       "    obsm: 'X_scVI_n_latent_20', 'X_scVI_n_latent_30', 'X_scVI_n_latent_40', 'X_umap', '_scvi_extra_categoricals'\n",
       "    layers: 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "green-values",
   "metadata": {},
   "source": [
    "### DO NOT Subset dataset, use full matrix counts from infected manifod!! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "least-ivory",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "#Take only the INFECTED counts matrix\n",
    "cellphoneDB_adata = adata[adata.obs['stage_perInfection']=='Pf_24h']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "confident-colony",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Take only the UNINFECTED counts matrix\n",
    "#cellphoneDB_adataUI = adata[adata.obs['stage_perInfection']=='UI_Pf_24h']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "polyphonic-footwear",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Pla_HDBR11345034_AAACCCAAGGAGGTTC    EVT_1\n",
       "Pla_HDBR11345034_AAACCCAAGGCCCAAA      HBC\n",
       "Pla_HDBR11345034_AAACCCAAGTGGTCAG       PV\n",
       "Pla_HDBR11345034_AAACCCACATACACCA        F\n",
       "Pla_HDBR11345034_AAACCCATCAAATGAG      HBC\n",
       "                                     ...  \n",
       "Pla_HDBR13661570_TTTGTTGGTCTTGCGG     F_sm\n",
       "Pla_HDBR13661570_TTTGTTGGTTCGGCGT      HBC\n",
       "Pla_HDBR13661570_TTTGTTGGTTCTGAGT        F\n",
       "Pla_HDBR13661570_TTTGTTGTCAGGAGAC       PV\n",
       "Pla_HDBR13661570_TTTGTTGTCGCACGGT      HBC\n",
       "Name: cell_type, Length: 27100, dtype: category\n",
       "Categories (15, object): ['EVT_1', 'EVT_2', 'Endo_f', 'F', ..., 'VCT_CCC', 'VCT_fusing', 'VCT_p', 'iEVT']"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['cell_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "accepting-concert",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>freqs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>categories</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Hrv124</th>\n",
       "      <td>8212</td>\n",
       "      <td>0.303026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hrv168</th>\n",
       "      <td>7112</td>\n",
       "      <td>0.262435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hrv232</th>\n",
       "      <td>2376</td>\n",
       "      <td>0.087675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hrv236</th>\n",
       "      <td>9400</td>\n",
       "      <td>0.346863</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            counts     freqs\n",
       "categories                  \n",
       "Hrv124        8212  0.303026\n",
       "Hrv168        7112  0.262435\n",
       "Hrv232        2376  0.087675\n",
       "Hrv236        9400  0.346863"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['donor_id'].values.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "color-ranking",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>freqs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>categories</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Pf_24h</th>\n",
       "      <td>27100</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            counts  freqs\n",
       "categories               \n",
       "Pf_24h       27100    1.0"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['stage_perInfection'].values.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "functional-netherlands",
   "metadata": {},
   "source": [
    "### Load *new* cell clusters annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "asian-miniature",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    }
   ],
   "source": [
    "### Add cluster subname such that this new name will match with DEG cluster name\n",
    "\n",
    "#INFECTED\n",
    "cellphoneDB_adata.obs['cell_type']= cellphoneDB_adata.obs.cell_type.astype('string')+'.'+cellphoneDB_adata.obs.infection.astype('string')\n",
    "\n",
    "\n",
    "#UI\n",
    "#cellphoneDB_adataUI.obs['cell_type']= cellphoneDB_adataUI.obs.cell_type.astype('string')+'.'+cellphoneDB_adataUI.obs.stage_perInfection.astype('string')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "assigned-testament",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Pla_HDBR11345034_AAACCCAAGGAGGTTC    EVT_1.Pf\n",
       "Pla_HDBR11345034_AAACCCAAGGCCCAAA      HBC.Pf\n",
       "Pla_HDBR11345034_AAACCCAAGTGGTCAG       PV.Pf\n",
       "Pla_HDBR11345034_AAACCCACATACACCA        F.Pf\n",
       "Pla_HDBR11345034_AAACCCATCAAATGAG      HBC.Pf\n",
       "                                       ...   \n",
       "Pla_HDBR13661570_TTTGTTGGTCTTGCGG     F_sm.Pf\n",
       "Pla_HDBR13661570_TTTGTTGGTTCGGCGT      HBC.Pf\n",
       "Pla_HDBR13661570_TTTGTTGGTTCTGAGT        F.Pf\n",
       "Pla_HDBR13661570_TTTGTTGTCAGGAGAC       PV.Pf\n",
       "Pla_HDBR13661570_TTTGTTGTCGCACGGT      HBC.Pf\n",
       "Name: cell_type, Length: 27100, dtype: string"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['cell_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "expired-quest",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "owned-pixel",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 27100 × 36601\n",
       "    obs: 'sample', 'stage', 'hpi', 'infection', 'percent_mito', 'n_counts', 'sample_barcode', 'assignment_SoC', 'donor_id', 'scrublet_score', 'scrublet_cluster_score', 'zscore', 'bh_pval', 'bonf_pval', 'S_score', 'G2M_score', 'phase', 'n_genes_by_counts', 'total_counts', 'total_counts_hs', 'pct_counts_hs', 'total_counts_tg', 'pct_counts_tg', 'Tg_infected', 'n_genes', '_scvi_batch', '_scvi_labels', '_scvi_local_l_mean', '_scvi_local_l_var', 'leiden_scvi', 'celltype_predictions', 'probabilities', 'scrublet_doublet', 'cell_type_2022', 'cell_type', 'souporcell_MFgenotype', 'MFgenotype', 'cell_type_broad', 'umap_density_Tg_infected', 'stage_perInfection', 'celltype-Stage', 'Tg_intracellular', 'celltype-Intracellular', 'Dev_Stage'\n",
       "    var: 'gene_ids', 'feature_types', 'mean-0', 'std-0', 'mean-1', 'std-1', 'mean-2', 'std-2', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm'\n",
       "    uns: 'Dev_Stage_colors', 'MFgenotype_colors', '_scvi', 'cell_type_2022_colors', 'cell_type_broad_colors', 'cell_type_colors', 'celltype_predictions_colors', 'donor_id_colors', 'hpi_colors', 'hvg', 'infection_colors', 'leiden', 'neighbors', 'sample_colors', 'scrublet_doublet_colors', 'umap', 'umap_density_Tg_infected_params'\n",
       "    obsm: 'X_scVI_n_latent_20', 'X_scVI_n_latent_30', 'X_scVI_n_latent_40', 'X_umap', '_scvi_extra_categoricals'\n",
       "    layers: 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cardiac-console",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "engaging-grave",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "sc.pp.normalize_per_cell(cellphoneDB_adata)\n",
    "sc.pp.log1p(cellphoneDB_adata)                                                                                                            \n",
    "# sc.pp.scale(adata, max_value=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "incident-wheat",
   "metadata": {},
   "source": [
    "### The anndata object contains counts that have been normalized (per cell) and log-transformed, therefore save this as below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "accredited-drunk",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'cell_type' as categorical\n"
     ]
    }
   ],
   "source": [
    "# counts input files \n",
    "cellphoneDB_adata.write('../data/malaria_inf_normlogTransformed_24h.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "sunset-aging",
   "metadata": {},
   "outputs": [],
   "source": [
    "# meta file\n",
    "df_meta = pd.DataFrame(data={'Cell':list(cellphoneDB_adata.obs.index),\n",
    "                             'cell_type':[ str(i) for i in cellphoneDB_adata.obs['cell_type']] })\n",
    "df_meta.set_index('Cell', inplace=True)\n",
    "df_meta.to_csv('../data/malaria_inf_meta_24h.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "aware-spread",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR11345034_AAACCCAAGGAGGTTC</th>\n",
       "      <td>EVT_1.Pf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR11345034_AAACCCAAGGCCCAAA</th>\n",
       "      <td>HBC.Pf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR11345034_AAACCCAAGTGGTCAG</th>\n",
       "      <td>PV.Pf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR11345034_AAACCCACATACACCA</th>\n",
       "      <td>F.Pf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR11345034_AAACCCATCAAATGAG</th>\n",
       "      <td>HBC.Pf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13661570_TTTGTTGGTCTTGCGG</th>\n",
       "      <td>F_sm.Pf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13661570_TTTGTTGGTTCGGCGT</th>\n",
       "      <td>HBC.Pf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13661570_TTTGTTGGTTCTGAGT</th>\n",
       "      <td>F.Pf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13661570_TTTGTTGTCAGGAGAC</th>\n",
       "      <td>PV.Pf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13661570_TTTGTTGTCGCACGGT</th>\n",
       "      <td>HBC.Pf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27100 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  cell_type\n",
       "Cell                                       \n",
       "Pla_HDBR11345034_AAACCCAAGGAGGTTC  EVT_1.Pf\n",
       "Pla_HDBR11345034_AAACCCAAGGCCCAAA    HBC.Pf\n",
       "Pla_HDBR11345034_AAACCCAAGTGGTCAG     PV.Pf\n",
       "Pla_HDBR11345034_AAACCCACATACACCA      F.Pf\n",
       "Pla_HDBR11345034_AAACCCATCAAATGAG    HBC.Pf\n",
       "...                                     ...\n",
       "Pla_HDBR13661570_TTTGTTGGTCTTGCGG   F_sm.Pf\n",
       "Pla_HDBR13661570_TTTGTTGGTTCGGCGT    HBC.Pf\n",
       "Pla_HDBR13661570_TTTGTTGGTTCTGAGT      F.Pf\n",
       "Pla_HDBR13661570_TTTGTTGTCAGGAGAC     PV.Pf\n",
       "Pla_HDBR13661570_TTTGTTGTCGCACGGT    HBC.Pf\n",
       "\n",
       "[27100 rows x 1 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "fifth-integral",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Pla_HDBR11345034_AAACCCAAGGAGGTTC    EVT_1.Pf\n",
       "Pla_HDBR11345034_AAACCCAAGGCCCAAA      HBC.Pf\n",
       "Pla_HDBR11345034_AAACCCAAGTGGTCAG       PV.Pf\n",
       "Pla_HDBR11345034_AAACCCACATACACCA        F.Pf\n",
       "Pla_HDBR11345034_AAACCCATCAAATGAG      HBC.Pf\n",
       "                                       ...   \n",
       "Pla_HDBR13661570_TTTGTTGGTCTTGCGG     F_sm.Pf\n",
       "Pla_HDBR13661570_TTTGTTGGTTCGGCGT      HBC.Pf\n",
       "Pla_HDBR13661570_TTTGTTGGTTCTGAGT        F.Pf\n",
       "Pla_HDBR13661570_TTTGTTGTCAGGAGAC       PV.Pf\n",
       "Pla_HDBR13661570_TTTGTTGTCGCACGGT      HBC.Pf\n",
       "Name: cell_type, Length: 27100, dtype: category\n",
       "Categories (15, object): ['EVT_1.Pf', 'EVT_2.Pf', 'Endo_f.Pf', 'F.Pf', ..., 'VCT_CCC.Pf', 'VCT_fusing.Pf', 'VCT_p.Pf', 'iEVT.Pf']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['cell_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "incorporated-essence",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "contemporary-speed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "structured-secretariat",
   "metadata": {},
   "source": [
    "### (2) Do DEGs individually from precalculated DEGs files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "organizational-clarity",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PATH to immune cells\n",
    "path='../../diffGeneExpression/macrophages/results/SC_limma_Immune_MAL_24h/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "utility-discussion",
   "metadata": {},
   "source": [
    "### HB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "incomplete-evening",
   "metadata": {},
   "outputs": [],
   "source": [
    "HBDEG = pd.read_csv(path+'1_DEGs_SC_MAL_24h_HBC.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "sexual-tender",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ZNF341-AS1</th>\n",
       "      <td>0.940907</td>\n",
       "      <td>3.927395</td>\n",
       "      <td>199.622904</td>\n",
       "      <td>2.783162e-13</td>\n",
       "      <td>2.283585e-09</td>\n",
       "      <td>2.232571e-09</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Pf</td>\n",
       "      <td>ZNF341-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z82244.2</th>\n",
       "      <td>2.443107</td>\n",
       "      <td>3.379129</td>\n",
       "      <td>178.314946</td>\n",
       "      <td>9.526965e-13</td>\n",
       "      <td>7.816875e-09</td>\n",
       "      <td>2.605625e-09</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Pf</td>\n",
       "      <td>Z82244.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTGIR</th>\n",
       "      <td>1.725812</td>\n",
       "      <td>3.955854</td>\n",
       "      <td>168.512944</td>\n",
       "      <td>1.754124e-12</td>\n",
       "      <td>1.439258e-08</td>\n",
       "      <td>3.598146e-09</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Pf</td>\n",
       "      <td>PTGIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MCRIP2</th>\n",
       "      <td>0.964274</td>\n",
       "      <td>5.891893</td>\n",
       "      <td>165.032206</td>\n",
       "      <td>2.195188e-12</td>\n",
       "      <td>1.801151e-08</td>\n",
       "      <td>3.602303e-09</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Pf</td>\n",
       "      <td>MCRIP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL27A1</th>\n",
       "      <td>1.343494</td>\n",
       "      <td>3.982235</td>\n",
       "      <td>161.456229</td>\n",
       "      <td>2.776139e-12</td>\n",
       "      <td>2.277822e-08</td>\n",
       "      <td>3.796370e-09</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Pf</td>\n",
       "      <td>COL27A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               logFC    logCPM           F        PValue    bonferroni  \\\n",
       "ZNF341-AS1  0.940907  3.927395  199.622904  2.783162e-13  2.283585e-09   \n",
       "Z82244.2    2.443107  3.379129  178.314946  9.526965e-13  7.816875e-09   \n",
       "PTGIR       1.725812  3.955854  168.512944  1.754124e-12  1.439258e-08   \n",
       "MCRIP2      0.964274  5.891893  165.032206  2.195188e-12  1.801151e-08   \n",
       "COL27A1     1.343494  3.982235  161.456229  2.776139e-12  2.277822e-08   \n",
       "\n",
       "               p_val_adj cell_type cluster        gene  \n",
       "ZNF341-AS1  2.232571e-09       HBC  HBC.Pf  ZNF341-AS1  \n",
       "Z82244.2    2.605625e-09       HBC  HBC.Pf    Z82244.2  \n",
       "PTGIR       3.598146e-09       HBC  HBC.Pf       PTGIR  \n",
       "MCRIP2      3.602303e-09       HBC  HBC.Pf      MCRIP2  \n",
       "COL27A1     3.796370e-09       HBC  HBC.Pf     COL27A1  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HBDEG_up= recoverDEGs(HBDEG, infection='Pf', celltype='HBC', direction='up')\n",
    "HBDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "guided-filename",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in HBDEG_up.cluster])\n",
    "HBDEG_up = HBDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "alpine-borough",
   "metadata": {},
   "outputs": [],
   "source": [
    "##  For cellphoneDB DEG mode, the 1st column = cluster; 2nd column = gene.\n",
    "##  All other columns are ignored.\n",
    "cncol = HBDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "HBDEG_up= HBDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(HBDEG_up).to_csv('../data/malaria_inf_HBC_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "accredited-cartoon",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VEGFA</th>\n",
       "      <td>HBC.Pf</td>\n",
       "      <td>VEGFA</td>\n",
       "      <td>0.614822</td>\n",
       "      <td>6.749733</td>\n",
       "      <td>35.556577</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.028104</td>\n",
       "      <td>0.00008</td>\n",
       "      <td>HBC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cluster   gene     logFC    logCPM          F    PValue  bonferroni  \\\n",
       "VEGFA  HBC.Pf  VEGFA  0.614822  6.749733  35.556577  0.000003    0.028104   \n",
       "\n",
       "       p_val_adj cell_type  \n",
       "VEGFA    0.00008       HBC  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#HBDEG_up=pd.read_table('../data/malaria_inf_HBC_DEGsup.tsv')\n",
    "#HBDEG_up.loc[['VEGFA']]\n",
    "HBDEG_up.loc[HBDEG_up['gene'] == 'VEGFA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "horizontal-things",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "surrounded-phrase",
   "metadata": {},
   "source": [
    "### PAMM1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "dense-butter",
   "metadata": {},
   "outputs": [],
   "source": [
    "PAMM1DEG = pd.read_csv(path+'2_DEGs_SC_MAL_24h_PAMM1.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "invalid-identification",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C4orf3</th>\n",
       "      <td>0.616426</td>\n",
       "      <td>7.769232</td>\n",
       "      <td>63.068401</td>\n",
       "      <td>2.753980e-08</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Pf</td>\n",
       "      <td>C4orf3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM162A</th>\n",
       "      <td>1.298828</td>\n",
       "      <td>6.355774</td>\n",
       "      <td>63.024569</td>\n",
       "      <td>2.771374e-08</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Pf</td>\n",
       "      <td>FAM162A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYO1B</th>\n",
       "      <td>1.025988</td>\n",
       "      <td>5.912743</td>\n",
       "      <td>63.319447</td>\n",
       "      <td>3.561647e-08</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Pf</td>\n",
       "      <td>MYO1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNF</th>\n",
       "      <td>0.910981</td>\n",
       "      <td>6.178153</td>\n",
       "      <td>60.599354</td>\n",
       "      <td>5.230316e-08</td>\n",
       "      <td>0.000301</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Pf</td>\n",
       "      <td>TNF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HILPDA</th>\n",
       "      <td>1.421535</td>\n",
       "      <td>7.880796</td>\n",
       "      <td>48.940179</td>\n",
       "      <td>3.177228e-07</td>\n",
       "      <td>0.001829</td>\n",
       "      <td>0.000305</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Pf</td>\n",
       "      <td>HILPDA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            logFC    logCPM          F        PValue  bonferroni  p_val_adj  \\\n",
       "C4orf3   0.616426  7.769232  63.068401  2.753980e-08    0.000159   0.000068   \n",
       "FAM162A  1.298828  6.355774  63.024569  2.771374e-08    0.000160   0.000068   \n",
       "MYO1B    1.025988  5.912743  63.319447  3.561647e-08    0.000205   0.000068   \n",
       "TNF      0.910981  6.178153  60.599354  5.230316e-08    0.000301   0.000075   \n",
       "HILPDA   1.421535  7.880796  48.940179  3.177228e-07    0.001829   0.000305   \n",
       "\n",
       "        cell_type   cluster     gene  \n",
       "C4orf3      PAMM1  PAMM1.Pf   C4orf3  \n",
       "FAM162A     PAMM1  PAMM1.Pf  FAM162A  \n",
       "MYO1B       PAMM1  PAMM1.Pf    MYO1B  \n",
       "TNF         PAMM1  PAMM1.Pf      TNF  \n",
       "HILPDA      PAMM1  PAMM1.Pf   HILPDA  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PAMM1DEG_up= recoverDEGs(PAMM1DEG, infection='Pf', celltype='PAMM1', direction='up')\n",
    "PAMM1DEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "sought-lemon",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in PAMM1DEG_up.cluster])\n",
    "PAMM1DEG_up = PAMM1DEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "moving-belfast",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = PAMM1DEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "PAMM1DEG_up= PAMM1DEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(PAMM1DEG_up).to_csv('../data/malaria_inf_PAMM1_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "continental-sentence",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C4orf3</th>\n",
       "      <td>PAMM1.Pf</td>\n",
       "      <td>C4orf3</td>\n",
       "      <td>0.616426</td>\n",
       "      <td>7.769232</td>\n",
       "      <td>63.068401</td>\n",
       "      <td>2.753980e-08</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM162A</th>\n",
       "      <td>PAMM1.Pf</td>\n",
       "      <td>FAM162A</td>\n",
       "      <td>1.298828</td>\n",
       "      <td>6.355774</td>\n",
       "      <td>63.024569</td>\n",
       "      <td>2.771374e-08</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYO1B</th>\n",
       "      <td>PAMM1.Pf</td>\n",
       "      <td>MYO1B</td>\n",
       "      <td>1.025988</td>\n",
       "      <td>5.912743</td>\n",
       "      <td>63.319447</td>\n",
       "      <td>3.561647e-08</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNF</th>\n",
       "      <td>PAMM1.Pf</td>\n",
       "      <td>TNF</td>\n",
       "      <td>0.910981</td>\n",
       "      <td>6.178153</td>\n",
       "      <td>60.599354</td>\n",
       "      <td>5.230316e-08</td>\n",
       "      <td>0.000301</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HILPDA</th>\n",
       "      <td>PAMM1.Pf</td>\n",
       "      <td>HILPDA</td>\n",
       "      <td>1.421535</td>\n",
       "      <td>7.880796</td>\n",
       "      <td>48.940179</td>\n",
       "      <td>3.177228e-07</td>\n",
       "      <td>0.001829</td>\n",
       "      <td>0.000305</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          cluster     gene     logFC    logCPM          F        PValue  \\\n",
       "C4orf3   PAMM1.Pf   C4orf3  0.616426  7.769232  63.068401  2.753980e-08   \n",
       "FAM162A  PAMM1.Pf  FAM162A  1.298828  6.355774  63.024569  2.771374e-08   \n",
       "MYO1B    PAMM1.Pf    MYO1B  1.025988  5.912743  63.319447  3.561647e-08   \n",
       "TNF      PAMM1.Pf      TNF  0.910981  6.178153  60.599354  5.230316e-08   \n",
       "HILPDA   PAMM1.Pf   HILPDA  1.421535  7.880796  48.940179  3.177228e-07   \n",
       "\n",
       "         bonferroni  p_val_adj cell_type  \n",
       "C4orf3     0.000159   0.000068     PAMM1  \n",
       "FAM162A    0.000160   0.000068     PAMM1  \n",
       "MYO1B      0.000205   0.000068     PAMM1  \n",
       "TNF        0.000301   0.000075     PAMM1  \n",
       "HILPDA     0.001829   0.000305     PAMM1  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PAMM1DEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "electronic-access",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "expensive-regular",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "simple-granny",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PATH to FibroEndo cells DGE tables\n",
    "path='../../diffGeneExpression/fibroEndo/results/SC_limma_fibroEndo_MAL_24h/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "liked-spiritual",
   "metadata": {},
   "source": [
    "### F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "eight-backup",
   "metadata": {},
   "outputs": [],
   "source": [
    "FDEG = pd.read_csv(path+'/1_DEGs_SC_MAL_24h_F.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "serious-clearing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HLA-DQB1</th>\n",
       "      <td>1.740492</td>\n",
       "      <td>4.800328</td>\n",
       "      <td>268.175627</td>\n",
       "      <td>6.820775e-13</td>\n",
       "      <td>4.961432e-09</td>\n",
       "      <td>4.961432e-09</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Pf</td>\n",
       "      <td>HLA-DQB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL1A</th>\n",
       "      <td>1.410089</td>\n",
       "      <td>6.560718</td>\n",
       "      <td>169.960655</td>\n",
       "      <td>1.531624e-12</td>\n",
       "      <td>1.114104e-08</td>\n",
       "      <td>5.570518e-09</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Pf</td>\n",
       "      <td>IL1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC43A3</th>\n",
       "      <td>0.685064</td>\n",
       "      <td>5.872149</td>\n",
       "      <td>159.081183</td>\n",
       "      <td>3.119786e-12</td>\n",
       "      <td>2.269333e-08</td>\n",
       "      <td>7.162107e-09</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Pf</td>\n",
       "      <td>SLC43A3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C15orf48</th>\n",
       "      <td>0.995682</td>\n",
       "      <td>7.925145</td>\n",
       "      <td>155.652466</td>\n",
       "      <td>3.938470e-12</td>\n",
       "      <td>2.864843e-08</td>\n",
       "      <td>7.162107e-09</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Pf</td>\n",
       "      <td>C15orf48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC2A1</th>\n",
       "      <td>1.027442</td>\n",
       "      <td>6.796381</td>\n",
       "      <td>151.772000</td>\n",
       "      <td>5.155001e-12</td>\n",
       "      <td>3.749748e-08</td>\n",
       "      <td>7.499496e-09</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Pf</td>\n",
       "      <td>SLC2A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             logFC    logCPM           F        PValue    bonferroni  \\\n",
       "HLA-DQB1  1.740492  4.800328  268.175627  6.820775e-13  4.961432e-09   \n",
       "IL1A      1.410089  6.560718  169.960655  1.531624e-12  1.114104e-08   \n",
       "SLC43A3   0.685064  5.872149  159.081183  3.119786e-12  2.269333e-08   \n",
       "C15orf48  0.995682  7.925145  155.652466  3.938470e-12  2.864843e-08   \n",
       "SLC2A1    1.027442  6.796381  151.772000  5.155001e-12  3.749748e-08   \n",
       "\n",
       "             p_val_adj cell_type cluster      gene  \n",
       "HLA-DQB1  4.961432e-09         F    F.Pf  HLA-DQB1  \n",
       "IL1A      5.570518e-09         F    F.Pf      IL1A  \n",
       "SLC43A3   7.162107e-09         F    F.Pf   SLC43A3  \n",
       "C15orf48  7.162107e-09         F    F.Pf  C15orf48  \n",
       "SLC2A1    7.499496e-09         F    F.Pf    SLC2A1  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FDEG_up= recoverDEGs(FDEG, infection='Pf', celltype='F', direction='up')\n",
    "FDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "sporting-recognition",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in FDEG_up.cluster])\n",
    "FDEG_up = FDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "foster-suspect",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = FDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "FDEG_up= FDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(FDEG_up).to_csv('../data/malaria_inf_F_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "capable-enterprise",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HLA-DQB1</th>\n",
       "      <td>F.Pf</td>\n",
       "      <td>HLA-DQB1</td>\n",
       "      <td>1.740492</td>\n",
       "      <td>4.800328</td>\n",
       "      <td>268.175627</td>\n",
       "      <td>6.820775e-13</td>\n",
       "      <td>4.961432e-09</td>\n",
       "      <td>4.961432e-09</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL1A</th>\n",
       "      <td>F.Pf</td>\n",
       "      <td>IL1A</td>\n",
       "      <td>1.410089</td>\n",
       "      <td>6.560718</td>\n",
       "      <td>169.960655</td>\n",
       "      <td>1.531624e-12</td>\n",
       "      <td>1.114104e-08</td>\n",
       "      <td>5.570518e-09</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC43A3</th>\n",
       "      <td>F.Pf</td>\n",
       "      <td>SLC43A3</td>\n",
       "      <td>0.685064</td>\n",
       "      <td>5.872149</td>\n",
       "      <td>159.081183</td>\n",
       "      <td>3.119786e-12</td>\n",
       "      <td>2.269333e-08</td>\n",
       "      <td>7.162107e-09</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C15orf48</th>\n",
       "      <td>F.Pf</td>\n",
       "      <td>C15orf48</td>\n",
       "      <td>0.995682</td>\n",
       "      <td>7.925145</td>\n",
       "      <td>155.652466</td>\n",
       "      <td>3.938470e-12</td>\n",
       "      <td>2.864843e-08</td>\n",
       "      <td>7.162107e-09</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC2A1</th>\n",
       "      <td>F.Pf</td>\n",
       "      <td>SLC2A1</td>\n",
       "      <td>1.027442</td>\n",
       "      <td>6.796381</td>\n",
       "      <td>151.772000</td>\n",
       "      <td>5.155001e-12</td>\n",
       "      <td>3.749748e-08</td>\n",
       "      <td>7.499496e-09</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         cluster      gene     logFC    logCPM           F        PValue  \\\n",
       "HLA-DQB1    F.Pf  HLA-DQB1  1.740492  4.800328  268.175627  6.820775e-13   \n",
       "IL1A        F.Pf      IL1A  1.410089  6.560718  169.960655  1.531624e-12   \n",
       "SLC43A3     F.Pf   SLC43A3  0.685064  5.872149  159.081183  3.119786e-12   \n",
       "C15orf48    F.Pf  C15orf48  0.995682  7.925145  155.652466  3.938470e-12   \n",
       "SLC2A1      F.Pf    SLC2A1  1.027442  6.796381  151.772000  5.155001e-12   \n",
       "\n",
       "            bonferroni     p_val_adj cell_type  \n",
       "HLA-DQB1  4.961432e-09  4.961432e-09         F  \n",
       "IL1A      1.114104e-08  5.570518e-09         F  \n",
       "SLC43A3   2.269333e-08  7.162107e-09         F  \n",
       "C15orf48  2.864843e-08  7.162107e-09         F  \n",
       "SLC2A1    3.749748e-08  7.499496e-09         F  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "analyzed-routine",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD99</th>\n",
       "      <td>F.Pf</td>\n",
       "      <td>CD99</td>\n",
       "      <td>0.204357</td>\n",
       "      <td>7.025717</td>\n",
       "      <td>6.955509</td>\n",
       "      <td>0.01427</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.043196</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     cluster  gene     logFC    logCPM         F   PValue  bonferroni  \\\n",
       "CD99    F.Pf  CD99  0.204357  7.025717  6.955509  0.01427         1.0   \n",
       "\n",
       "      p_val_adj cell_type  \n",
       "CD99   0.043196         F  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FDEG_up.loc[['CD99']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "conceptual-nicholas",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "placed-convertible",
   "metadata": {},
   "source": [
    "### PV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "designed-spectrum",
   "metadata": {},
   "outputs": [],
   "source": [
    "PVDEG = pd.read_csv(path+'/2_DEGs_SC_MAL_24h_PV.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "ceramic-gasoline",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C15orf48</th>\n",
       "      <td>1.387066</td>\n",
       "      <td>6.225848</td>\n",
       "      <td>172.385507</td>\n",
       "      <td>6.192700e-13</td>\n",
       "      <td>5.426044e-09</td>\n",
       "      <td>5.426044e-09</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Pf</td>\n",
       "      <td>C15orf48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMSN1</th>\n",
       "      <td>1.233680</td>\n",
       "      <td>4.497978</td>\n",
       "      <td>119.561670</td>\n",
       "      <td>3.474524e-11</td>\n",
       "      <td>3.044378e-07</td>\n",
       "      <td>9.219040e-08</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Pf</td>\n",
       "      <td>SAMSN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNX10</th>\n",
       "      <td>1.564362</td>\n",
       "      <td>3.189589</td>\n",
       "      <td>124.008690</td>\n",
       "      <td>2.347921e-11</td>\n",
       "      <td>2.057248e-07</td>\n",
       "      <td>9.219040e-08</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Pf</td>\n",
       "      <td>SNX10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ICAM1</th>\n",
       "      <td>1.099168</td>\n",
       "      <td>7.059811</td>\n",
       "      <td>117.435461</td>\n",
       "      <td>4.208646e-11</td>\n",
       "      <td>3.687616e-07</td>\n",
       "      <td>9.219040e-08</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Pf</td>\n",
       "      <td>ICAM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALOX5AP</th>\n",
       "      <td>1.159447</td>\n",
       "      <td>3.605170</td>\n",
       "      <td>146.496630</td>\n",
       "      <td>6.942756e-11</td>\n",
       "      <td>6.083243e-07</td>\n",
       "      <td>1.216649e-07</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Pf</td>\n",
       "      <td>ALOX5AP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             logFC    logCPM           F        PValue    bonferroni  \\\n",
       "C15orf48  1.387066  6.225848  172.385507  6.192700e-13  5.426044e-09   \n",
       "SAMSN1    1.233680  4.497978  119.561670  3.474524e-11  3.044378e-07   \n",
       "SNX10     1.564362  3.189589  124.008690  2.347921e-11  2.057248e-07   \n",
       "ICAM1     1.099168  7.059811  117.435461  4.208646e-11  3.687616e-07   \n",
       "ALOX5AP   1.159447  3.605170  146.496630  6.942756e-11  6.083243e-07   \n",
       "\n",
       "             p_val_adj cell_type cluster      gene  \n",
       "C15orf48  5.426044e-09        PV   PV.Pf  C15orf48  \n",
       "SAMSN1    9.219040e-08        PV   PV.Pf    SAMSN1  \n",
       "SNX10     9.219040e-08        PV   PV.Pf     SNX10  \n",
       "ICAM1     9.219040e-08        PV   PV.Pf     ICAM1  \n",
       "ALOX5AP   1.216649e-07        PV   PV.Pf   ALOX5AP  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PVDEG_up= recoverDEGs(PVDEG, infection='Pf', celltype='PV', direction='up')\n",
    "PVDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "outer-charles",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in PVDEG_up.cluster])\n",
    "PVDEG_up = PVDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "accepted-departure",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = PVDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "PVDEG_up= PVDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(PVDEG_up).to_csv('../data/malaria_inf_PV_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "senior-parks",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C15orf48</th>\n",
       "      <td>PV.Pf</td>\n",
       "      <td>C15orf48</td>\n",
       "      <td>1.387066</td>\n",
       "      <td>6.225848</td>\n",
       "      <td>172.385507</td>\n",
       "      <td>6.192700e-13</td>\n",
       "      <td>5.426044e-09</td>\n",
       "      <td>5.426044e-09</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMSN1</th>\n",
       "      <td>PV.Pf</td>\n",
       "      <td>SAMSN1</td>\n",
       "      <td>1.233680</td>\n",
       "      <td>4.497978</td>\n",
       "      <td>119.561670</td>\n",
       "      <td>3.474524e-11</td>\n",
       "      <td>3.044378e-07</td>\n",
       "      <td>9.219040e-08</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNX10</th>\n",
       "      <td>PV.Pf</td>\n",
       "      <td>SNX10</td>\n",
       "      <td>1.564362</td>\n",
       "      <td>3.189589</td>\n",
       "      <td>124.008690</td>\n",
       "      <td>2.347921e-11</td>\n",
       "      <td>2.057248e-07</td>\n",
       "      <td>9.219040e-08</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ICAM1</th>\n",
       "      <td>PV.Pf</td>\n",
       "      <td>ICAM1</td>\n",
       "      <td>1.099168</td>\n",
       "      <td>7.059811</td>\n",
       "      <td>117.435461</td>\n",
       "      <td>4.208646e-11</td>\n",
       "      <td>3.687616e-07</td>\n",
       "      <td>9.219040e-08</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALOX5AP</th>\n",
       "      <td>PV.Pf</td>\n",
       "      <td>ALOX5AP</td>\n",
       "      <td>1.159447</td>\n",
       "      <td>3.605170</td>\n",
       "      <td>146.496630</td>\n",
       "      <td>6.942756e-11</td>\n",
       "      <td>6.083243e-07</td>\n",
       "      <td>1.216649e-07</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         cluster      gene     logFC    logCPM           F        PValue  \\\n",
       "C15orf48   PV.Pf  C15orf48  1.387066  6.225848  172.385507  6.192700e-13   \n",
       "SAMSN1     PV.Pf    SAMSN1  1.233680  4.497978  119.561670  3.474524e-11   \n",
       "SNX10      PV.Pf     SNX10  1.564362  3.189589  124.008690  2.347921e-11   \n",
       "ICAM1      PV.Pf     ICAM1  1.099168  7.059811  117.435461  4.208646e-11   \n",
       "ALOX5AP    PV.Pf   ALOX5AP  1.159447  3.605170  146.496630  6.942756e-11   \n",
       "\n",
       "            bonferroni     p_val_adj cell_type  \n",
       "C15orf48  5.426044e-09  5.426044e-09        PV  \n",
       "SAMSN1    3.044378e-07  9.219040e-08        PV  \n",
       "SNX10     2.057248e-07  9.219040e-08        PV  \n",
       "ICAM1     3.687616e-07  9.219040e-08        PV  \n",
       "ALOX5AP   6.083243e-07  1.216649e-07        PV  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PVDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "pretty-retail",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bottom-mistress",
   "metadata": {},
   "source": [
    "### Endo_f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "valid-alabama",
   "metadata": {},
   "outputs": [],
   "source": [
    "EndoDEG = pd.read_csv(path+'/3_DEGs_SC_MAL_24h_Endof.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "requested-tuning",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IL1RN</th>\n",
       "      <td>2.190200</td>\n",
       "      <td>3.914339</td>\n",
       "      <td>184.166411</td>\n",
       "      <td>2.181283e-13</td>\n",
       "      <td>1.816572e-09</td>\n",
       "      <td>1.816572e-09</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Pf</td>\n",
       "      <td>IL1RN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCL3L1</th>\n",
       "      <td>2.244097</td>\n",
       "      <td>4.111346</td>\n",
       "      <td>146.678153</td>\n",
       "      <td>2.917805e-12</td>\n",
       "      <td>2.429948e-08</td>\n",
       "      <td>1.214974e-08</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Pf</td>\n",
       "      <td>CCL3L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL1A</th>\n",
       "      <td>1.178834</td>\n",
       "      <td>5.366196</td>\n",
       "      <td>123.323318</td>\n",
       "      <td>1.990271e-11</td>\n",
       "      <td>1.657498e-07</td>\n",
       "      <td>5.057966e-08</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Pf</td>\n",
       "      <td>IL1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FTH1</th>\n",
       "      <td>0.839001</td>\n",
       "      <td>12.135905</td>\n",
       "      <td>121.088272</td>\n",
       "      <td>2.429378e-11</td>\n",
       "      <td>2.023186e-07</td>\n",
       "      <td>5.057966e-08</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Pf</td>\n",
       "      <td>FTH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCL5</th>\n",
       "      <td>2.153293</td>\n",
       "      <td>4.829508</td>\n",
       "      <td>111.313957</td>\n",
       "      <td>6.031100e-11</td>\n",
       "      <td>5.022700e-07</td>\n",
       "      <td>1.004540e-07</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Pf</td>\n",
       "      <td>CCL5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           logFC     logCPM           F        PValue    bonferroni  \\\n",
       "IL1RN   2.190200   3.914339  184.166411  2.181283e-13  1.816572e-09   \n",
       "CCL3L1  2.244097   4.111346  146.678153  2.917805e-12  2.429948e-08   \n",
       "IL1A    1.178834   5.366196  123.323318  1.990271e-11  1.657498e-07   \n",
       "FTH1    0.839001  12.135905  121.088272  2.429378e-11  2.023186e-07   \n",
       "CCL5    2.153293   4.829508  111.313957  6.031100e-11  5.022700e-07   \n",
       "\n",
       "           p_val_adj cell_type    cluster    gene  \n",
       "IL1RN   1.816572e-09    Endo_f  Endo_f.Pf   IL1RN  \n",
       "CCL3L1  1.214974e-08    Endo_f  Endo_f.Pf  CCL3L1  \n",
       "IL1A    5.057966e-08    Endo_f  Endo_f.Pf    IL1A  \n",
       "FTH1    5.057966e-08    Endo_f  Endo_f.Pf    FTH1  \n",
       "CCL5    1.004540e-07    Endo_f  Endo_f.Pf    CCL5  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EndoDEG_up= recoverDEGs(EndoDEG, infection='Pf', celltype='Endo_f', direction='up')\n",
    "EndoDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "threaded-ireland",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in EndoDEG_up.cluster])\n",
    "EndoDEG_up = EndoDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "textile-cherry",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = EndoDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "EndoDEG_up= EndoDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(EndoDEG_up).to_csv('../data/malaria_inf_Endof_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "palestinian-excellence",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IL1RN</th>\n",
       "      <td>Endo_f.Pf</td>\n",
       "      <td>IL1RN</td>\n",
       "      <td>2.190200</td>\n",
       "      <td>3.914339</td>\n",
       "      <td>184.166411</td>\n",
       "      <td>2.181283e-13</td>\n",
       "      <td>1.816572e-09</td>\n",
       "      <td>1.816572e-09</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCL3L1</th>\n",
       "      <td>Endo_f.Pf</td>\n",
       "      <td>CCL3L1</td>\n",
       "      <td>2.244097</td>\n",
       "      <td>4.111346</td>\n",
       "      <td>146.678153</td>\n",
       "      <td>2.917805e-12</td>\n",
       "      <td>2.429948e-08</td>\n",
       "      <td>1.214974e-08</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL1A</th>\n",
       "      <td>Endo_f.Pf</td>\n",
       "      <td>IL1A</td>\n",
       "      <td>1.178834</td>\n",
       "      <td>5.366196</td>\n",
       "      <td>123.323318</td>\n",
       "      <td>1.990271e-11</td>\n",
       "      <td>1.657498e-07</td>\n",
       "      <td>5.057966e-08</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FTH1</th>\n",
       "      <td>Endo_f.Pf</td>\n",
       "      <td>FTH1</td>\n",
       "      <td>0.839001</td>\n",
       "      <td>12.135905</td>\n",
       "      <td>121.088272</td>\n",
       "      <td>2.429378e-11</td>\n",
       "      <td>2.023186e-07</td>\n",
       "      <td>5.057966e-08</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCL5</th>\n",
       "      <td>Endo_f.Pf</td>\n",
       "      <td>CCL5</td>\n",
       "      <td>2.153293</td>\n",
       "      <td>4.829508</td>\n",
       "      <td>111.313957</td>\n",
       "      <td>6.031100e-11</td>\n",
       "      <td>5.022700e-07</td>\n",
       "      <td>1.004540e-07</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          cluster    gene     logFC     logCPM           F        PValue  \\\n",
       "IL1RN   Endo_f.Pf   IL1RN  2.190200   3.914339  184.166411  2.181283e-13   \n",
       "CCL3L1  Endo_f.Pf  CCL3L1  2.244097   4.111346  146.678153  2.917805e-12   \n",
       "IL1A    Endo_f.Pf    IL1A  1.178834   5.366196  123.323318  1.990271e-11   \n",
       "FTH1    Endo_f.Pf    FTH1  0.839001  12.135905  121.088272  2.429378e-11   \n",
       "CCL5    Endo_f.Pf    CCL5  2.153293   4.829508  111.313957  6.031100e-11   \n",
       "\n",
       "          bonferroni     p_val_adj cell_type  \n",
       "IL1RN   1.816572e-09  1.816572e-09    Endo_f  \n",
       "CCL3L1  2.429948e-08  1.214974e-08    Endo_f  \n",
       "IL1A    1.657498e-07  5.057966e-08    Endo_f  \n",
       "FTH1    2.023186e-07  5.057966e-08    Endo_f  \n",
       "CCL5    5.022700e-07  1.004540e-07    Endo_f  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EndoDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bronze-miracle",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fancy-globe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "recorded-language",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PATH to Trophoblast cells DGE tables\n",
    "path='../../diffGeneExpression/trophoblast/results/SC_limma_Trophoblast_MAL_24h'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "plastic-dragon",
   "metadata": {},
   "source": [
    "### VCT_fusing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "forty-olympus",
   "metadata": {},
   "outputs": [],
   "source": [
    "VCTfDEG = pd.read_csv(path+'/1_DEGs_SC_MAL_24h_VCT_fusing.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "surface-candidate",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MMP24OS</th>\n",
       "      <td>0.374704</td>\n",
       "      <td>5.726263</td>\n",
       "      <td>90.631744</td>\n",
       "      <td>1.225836e-09</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Pf</td>\n",
       "      <td>MMP24OS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GUK1</th>\n",
       "      <td>0.230347</td>\n",
       "      <td>7.289740</td>\n",
       "      <td>72.022147</td>\n",
       "      <td>1.058798e-08</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Pf</td>\n",
       "      <td>GUK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERRFI1</th>\n",
       "      <td>0.519455</td>\n",
       "      <td>8.211099</td>\n",
       "      <td>58.491610</td>\n",
       "      <td>6.755355e-08</td>\n",
       "      <td>0.000535</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Pf</td>\n",
       "      <td>ERRFI1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NFKBIZ</th>\n",
       "      <td>0.328492</td>\n",
       "      <td>5.681678</td>\n",
       "      <td>58.968408</td>\n",
       "      <td>6.296046e-08</td>\n",
       "      <td>0.000499</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Pf</td>\n",
       "      <td>NFKBIZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATOX1</th>\n",
       "      <td>0.257473</td>\n",
       "      <td>6.586556</td>\n",
       "      <td>57.898700</td>\n",
       "      <td>7.377820e-08</td>\n",
       "      <td>0.000584</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Pf</td>\n",
       "      <td>ATOX1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            logFC    logCPM          F        PValue  bonferroni  p_val_adj  \\\n",
       "MMP24OS  0.374704  5.726263  90.631744  1.225836e-09    0.000010   0.000010   \n",
       "GUK1     0.230347  7.289740  72.022147  1.058798e-08    0.000084   0.000017   \n",
       "ERRFI1   0.519455  8.211099  58.491610  6.755355e-08    0.000535   0.000049   \n",
       "NFKBIZ   0.328492  5.681678  58.968408  6.296046e-08    0.000499   0.000049   \n",
       "ATOX1    0.257473  6.586556  57.898700  7.377820e-08    0.000584   0.000049   \n",
       "\n",
       "          cell_type        cluster     gene  \n",
       "MMP24OS  VCT_fusing  VCT_fusing.Pf  MMP24OS  \n",
       "GUK1     VCT_fusing  VCT_fusing.Pf     GUK1  \n",
       "ERRFI1   VCT_fusing  VCT_fusing.Pf   ERRFI1  \n",
       "NFKBIZ   VCT_fusing  VCT_fusing.Pf   NFKBIZ  \n",
       "ATOX1    VCT_fusing  VCT_fusing.Pf    ATOX1  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTfDEG_up= recoverDEGs(VCTfDEG, infection='Pf', celltype='VCT_fusing', direction='up')\n",
    "VCTfDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "strange-clinton",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in VCTfDEG_up.cluster])\n",
    "VCTfVCTfDEG_upDEG = VCTfDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "powered-guest",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = VCTfDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "VCTfDEG_up= VCTfDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(VCTfDEG_up).to_csv('../data/malaria_inf_VCTfusing_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "proprietary-identification",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MMP24OS</th>\n",
       "      <td>VCT_fusing.Pf</td>\n",
       "      <td>MMP24OS</td>\n",
       "      <td>0.374704</td>\n",
       "      <td>5.726263</td>\n",
       "      <td>90.631744</td>\n",
       "      <td>1.225836e-09</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GUK1</th>\n",
       "      <td>VCT_fusing.Pf</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>0.230347</td>\n",
       "      <td>7.289740</td>\n",
       "      <td>72.022147</td>\n",
       "      <td>1.058798e-08</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERRFI1</th>\n",
       "      <td>VCT_fusing.Pf</td>\n",
       "      <td>ERRFI1</td>\n",
       "      <td>0.519455</td>\n",
       "      <td>8.211099</td>\n",
       "      <td>58.491610</td>\n",
       "      <td>6.755355e-08</td>\n",
       "      <td>0.000535</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NFKBIZ</th>\n",
       "      <td>VCT_fusing.Pf</td>\n",
       "      <td>NFKBIZ</td>\n",
       "      <td>0.328492</td>\n",
       "      <td>5.681678</td>\n",
       "      <td>58.968408</td>\n",
       "      <td>6.296046e-08</td>\n",
       "      <td>0.000499</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATOX1</th>\n",
       "      <td>VCT_fusing.Pf</td>\n",
       "      <td>ATOX1</td>\n",
       "      <td>0.257473</td>\n",
       "      <td>6.586556</td>\n",
       "      <td>57.898700</td>\n",
       "      <td>7.377820e-08</td>\n",
       "      <td>0.000584</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               cluster     gene     logFC    logCPM          F        PValue  \\\n",
       "MMP24OS  VCT_fusing.Pf  MMP24OS  0.374704  5.726263  90.631744  1.225836e-09   \n",
       "GUK1     VCT_fusing.Pf     GUK1  0.230347  7.289740  72.022147  1.058798e-08   \n",
       "ERRFI1   VCT_fusing.Pf   ERRFI1  0.519455  8.211099  58.491610  6.755355e-08   \n",
       "NFKBIZ   VCT_fusing.Pf   NFKBIZ  0.328492  5.681678  58.968408  6.296046e-08   \n",
       "ATOX1    VCT_fusing.Pf    ATOX1  0.257473  6.586556  57.898700  7.377820e-08   \n",
       "\n",
       "         bonferroni  p_val_adj   cell_type  \n",
       "MMP24OS    0.000010   0.000010  VCT_fusing  \n",
       "GUK1       0.000084   0.000017  VCT_fusing  \n",
       "ERRFI1     0.000535   0.000049  VCT_fusing  \n",
       "NFKBIZ     0.000499   0.000049  VCT_fusing  \n",
       "ATOX1      0.000584   0.000049  VCT_fusing  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTfDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "cooperative-desire",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SPP1</th>\n",
       "      <td>VCT_fusing.Pf</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>1.152154</td>\n",
       "      <td>6.005685</td>\n",
       "      <td>18.427151</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.004494</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            cluster  gene     logFC    logCPM          F    PValue  \\\n",
       "SPP1  VCT_fusing.Pf  SPP1  1.152154  6.005685  18.427151  0.000249   \n",
       "\n",
       "      bonferroni  p_val_adj   cell_type  \n",
       "SPP1         1.0   0.004494  VCT_fusing  "
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTfDEG_up.loc[['SPP1']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "material-cornwall",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "legislative-pontiac",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "dietary-package",
   "metadata": {},
   "source": [
    "### VCT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "mighty-chest",
   "metadata": {},
   "outputs": [],
   "source": [
    "VCTDEG = pd.read_csv(path+'/2_DEGs_SC_MAL_24h_VCT.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "romantic-penetration",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CCL3</th>\n",
       "      <td>1.994071</td>\n",
       "      <td>5.772547</td>\n",
       "      <td>155.362130</td>\n",
       "      <td>9.110064e-13</td>\n",
       "      <td>9.107330e-09</td>\n",
       "      <td>9.107330e-09</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Pf</td>\n",
       "      <td>CCL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPP1</th>\n",
       "      <td>1.706711</td>\n",
       "      <td>5.689939</td>\n",
       "      <td>67.544276</td>\n",
       "      <td>7.463222e-09</td>\n",
       "      <td>7.460983e-05</td>\n",
       "      <td>2.486994e-05</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Pf</td>\n",
       "      <td>SPP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HMGN2</th>\n",
       "      <td>0.439966</td>\n",
       "      <td>9.238759</td>\n",
       "      <td>60.852914</td>\n",
       "      <td>2.051519e-08</td>\n",
       "      <td>2.050904e-04</td>\n",
       "      <td>3.257878e-05</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Pf</td>\n",
       "      <td>HMGN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM265</th>\n",
       "      <td>0.727250</td>\n",
       "      <td>5.965756</td>\n",
       "      <td>60.179202</td>\n",
       "      <td>2.281199e-08</td>\n",
       "      <td>2.280514e-04</td>\n",
       "      <td>3.257878e-05</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Pf</td>\n",
       "      <td>TMEM265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCL20</th>\n",
       "      <td>1.452804</td>\n",
       "      <td>5.077363</td>\n",
       "      <td>54.852381</td>\n",
       "      <td>5.443452e-08</td>\n",
       "      <td>5.441819e-04</td>\n",
       "      <td>6.802274e-05</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Pf</td>\n",
       "      <td>CCL20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            logFC    logCPM           F        PValue    bonferroni  \\\n",
       "CCL3     1.994071  5.772547  155.362130  9.110064e-13  9.107330e-09   \n",
       "SPP1     1.706711  5.689939   67.544276  7.463222e-09  7.460983e-05   \n",
       "HMGN2    0.439966  9.238759   60.852914  2.051519e-08  2.050904e-04   \n",
       "TMEM265  0.727250  5.965756   60.179202  2.281199e-08  2.280514e-04   \n",
       "CCL20    1.452804  5.077363   54.852381  5.443452e-08  5.441819e-04   \n",
       "\n",
       "            p_val_adj cell_type cluster     gene  \n",
       "CCL3     9.107330e-09       VCT  VCT.Pf     CCL3  \n",
       "SPP1     2.486994e-05       VCT  VCT.Pf     SPP1  \n",
       "HMGN2    3.257878e-05       VCT  VCT.Pf    HMGN2  \n",
       "TMEM265  3.257878e-05       VCT  VCT.Pf  TMEM265  \n",
       "CCL20    6.802274e-05       VCT  VCT.Pf    CCL20  "
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTDEG_up= recoverDEGs(VCTDEG, infection='Pf', celltype='VCT', direction='up')\n",
    "VCTDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "decreased-forest",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in VCTDEG_up.cluster])\n",
    "VCTDEG_up = VCTDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "developmental-extraction",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = VCTDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "VCTDEG_up= VCTDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(VCTDEG_up).to_csv('../data/malaria_inf_VCT_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "cheap-daily",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CCL3</th>\n",
       "      <td>VCT.Pf</td>\n",
       "      <td>CCL3</td>\n",
       "      <td>1.994071</td>\n",
       "      <td>5.772547</td>\n",
       "      <td>155.362130</td>\n",
       "      <td>9.110064e-13</td>\n",
       "      <td>9.107330e-09</td>\n",
       "      <td>9.107330e-09</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPP1</th>\n",
       "      <td>VCT.Pf</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>1.706711</td>\n",
       "      <td>5.689939</td>\n",
       "      <td>67.544276</td>\n",
       "      <td>7.463222e-09</td>\n",
       "      <td>7.460983e-05</td>\n",
       "      <td>2.486994e-05</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HMGN2</th>\n",
       "      <td>VCT.Pf</td>\n",
       "      <td>HMGN2</td>\n",
       "      <td>0.439966</td>\n",
       "      <td>9.238759</td>\n",
       "      <td>60.852914</td>\n",
       "      <td>2.051519e-08</td>\n",
       "      <td>2.050904e-04</td>\n",
       "      <td>3.257878e-05</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM265</th>\n",
       "      <td>VCT.Pf</td>\n",
       "      <td>TMEM265</td>\n",
       "      <td>0.727250</td>\n",
       "      <td>5.965756</td>\n",
       "      <td>60.179202</td>\n",
       "      <td>2.281199e-08</td>\n",
       "      <td>2.280514e-04</td>\n",
       "      <td>3.257878e-05</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCL20</th>\n",
       "      <td>VCT.Pf</td>\n",
       "      <td>CCL20</td>\n",
       "      <td>1.452804</td>\n",
       "      <td>5.077363</td>\n",
       "      <td>54.852381</td>\n",
       "      <td>5.443452e-08</td>\n",
       "      <td>5.441819e-04</td>\n",
       "      <td>6.802274e-05</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        cluster     gene     logFC    logCPM           F        PValue  \\\n",
       "CCL3     VCT.Pf     CCL3  1.994071  5.772547  155.362130  9.110064e-13   \n",
       "SPP1     VCT.Pf     SPP1  1.706711  5.689939   67.544276  7.463222e-09   \n",
       "HMGN2    VCT.Pf    HMGN2  0.439966  9.238759   60.852914  2.051519e-08   \n",
       "TMEM265  VCT.Pf  TMEM265  0.727250  5.965756   60.179202  2.281199e-08   \n",
       "CCL20    VCT.Pf    CCL20  1.452804  5.077363   54.852381  5.443452e-08   \n",
       "\n",
       "           bonferroni     p_val_adj cell_type  \n",
       "CCL3     9.107330e-09  9.107330e-09       VCT  \n",
       "SPP1     7.460983e-05  2.486994e-05       VCT  \n",
       "HMGN2    2.050904e-04  3.257878e-05       VCT  \n",
       "TMEM265  2.280514e-04  3.257878e-05       VCT  \n",
       "CCL20    5.441819e-04  6.802274e-05       VCT  "
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "seasonal-closing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SPP1</th>\n",
       "      <td>VCT.Pf</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>1.706711</td>\n",
       "      <td>5.689939</td>\n",
       "      <td>67.544276</td>\n",
       "      <td>7.463222e-09</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     cluster  gene     logFC    logCPM          F        PValue  bonferroni  \\\n",
       "SPP1  VCT.Pf  SPP1  1.706711  5.689939  67.544276  7.463222e-09    0.000075   \n",
       "\n",
       "      p_val_adj cell_type  \n",
       "SPP1   0.000025       VCT  "
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTDEG_up.loc[['SPP1']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "durable-annual",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "raised-choir",
   "metadata": {},
   "source": [
    "### (3) Run cellphoneDB"
   ]
  },
  {
   "cell_type": "raw",
   "id": "biblical-mississippi",
   "metadata": {},
   "source": [
    "source activate cpdb\n",
    "\n",
    "bash run_cellphoneDB.sh "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "stainless-beatles",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
