{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "related-theta",
   "metadata": {},
   "source": [
    "**Author:** Elias Rafael Ruiz-Morales\n",
    "\n",
    "**Institution:** Wellcome Sanger institute\n",
    "\n",
    "**July, 2023**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "musical-posting",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "comfortable-interference",
   "metadata": {},
   "source": [
    "## Generate cellphoneDB input files (1) _meta.tsv and (2) _degs.tsv"
   ]
  },
  {
   "cell_type": "raw",
   "id": "confirmed-fiction",
   "metadata": {},
   "source": [
    "reference: https://github.com/ventolab/CellphoneDB/blob/master/notebooks/0_prepare_your_data_from_anndata.ipynb\n",
    "reference: https://github.com/ventolab/HGDA/blob/main/human/scRNAseq/cellphoneDB/cellphoneDB1_granulosa_preparefiles.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "organizational-cooking",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/opt/conda/bin/python'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import os\n",
    "import sys\n",
    "sys.executable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "about-guyana",
   "metadata": {},
   "outputs": [],
   "source": [
    "def recoverDEGs(DEGdf, infection, celltype, direction):\n",
    "\n",
    "    \n",
    "    ### Up reg\n",
    "    ###\n",
    "    if direction=='up':\n",
    "\n",
    "        #selecting only up regulated genes\n",
    "        tempUP = DEGdf[DEGdf['logFC'] > 0].copy()\n",
    "        \n",
    "        #making new columns\n",
    "        tempUP['cell_type'] = celltype\n",
    "        tempUP['cluster'] = celltype + '.' + infection\n",
    "        tempUP['gene']=list(tempUP.index.values)\n",
    "\n",
    "        # select significant less restrictive\n",
    "        tempUP =tempUP[ tempUP.p_val_adj < 0.05]\n",
    "        tempUP = tempUP[ tempUP.logFC > 0.1]\n",
    "\n",
    "        \n",
    "        return(tempUP)\n",
    "\n",
    "\n",
    "    ### Down reg\n",
    "    ###\n",
    "    if direction=='down':\n",
    "        #selecting only down regulated genes\n",
    "        tempUP = DEGdf[DEGdf['logFC'] < 0].copy()\n",
    "        \n",
    "        #making new columns\n",
    "        tempUP['cell_type'] = celltype\n",
    "        tempUP['cluster'] = celltype + '.' + infection\n",
    "        tempUP['gene']=list(tempUP.index.values)\n",
    "\n",
    "\n",
    "        # select significant less restrictive\n",
    "        tempUP =tempUP[ tempUP.p_val_adj < 0.05]\n",
    "        tempUP = tempUP[ tempUP.logFC < -0.1]\n",
    "        \n",
    "        return(tempDwn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "atmospheric-snake",
   "metadata": {},
   "source": [
    "## Prepare INPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "developed-sentence",
   "metadata": {},
   "source": [
    "### (1) Load annadata_ single-cell explants from toxoplasma infection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "close-advertiser",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read('../../data_integration/results/scVI/toxoplasma_singleCell_24h.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "critical-powell",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>freqs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>categories</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EVT_1</th>\n",
       "      <td>223</td>\n",
       "      <td>0.006673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EVT_2</th>\n",
       "      <td>606</td>\n",
       "      <td>0.018134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Endo_f</th>\n",
       "      <td>3520</td>\n",
       "      <td>0.105332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>4956</td>\n",
       "      <td>0.148303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F_p</th>\n",
       "      <td>420</td>\n",
       "      <td>0.012568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F_sm</th>\n",
       "      <td>309</td>\n",
       "      <td>0.009247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HBC</th>\n",
       "      <td>11481</td>\n",
       "      <td>0.343557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HBC_p</th>\n",
       "      <td>120</td>\n",
       "      <td>0.003591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PAMM1</th>\n",
       "      <td>1402</td>\n",
       "      <td>0.041953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PV</th>\n",
       "      <td>2635</td>\n",
       "      <td>0.078850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT</th>\n",
       "      <td>2796</td>\n",
       "      <td>0.083667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_CCC</th>\n",
       "      <td>773</td>\n",
       "      <td>0.023131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_fusing</th>\n",
       "      <td>2728</td>\n",
       "      <td>0.081633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_p</th>\n",
       "      <td>1094</td>\n",
       "      <td>0.032737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iEVT</th>\n",
       "      <td>355</td>\n",
       "      <td>0.010623</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            counts     freqs\n",
       "categories                  \n",
       "EVT_1          223  0.006673\n",
       "EVT_2          606  0.018134\n",
       "Endo_f        3520  0.105332\n",
       "F             4956  0.148303\n",
       "F_p            420  0.012568\n",
       "F_sm           309  0.009247\n",
       "HBC          11481  0.343557\n",
       "HBC_p          120  0.003591\n",
       "PAMM1         1402  0.041953\n",
       "PV            2635  0.078850\n",
       "VCT           2796  0.083667\n",
       "VCT_CCC        773  0.023131\n",
       "VCT_fusing    2728  0.081633\n",
       "VCT_p         1094  0.032737\n",
       "iEVT           355  0.010623"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['cell_type'].values.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "loved-railway",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>freqs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>categories</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tg_24h</th>\n",
       "      <td>16806</td>\n",
       "      <td>0.502903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UI_Tg_24h</th>\n",
       "      <td>16612</td>\n",
       "      <td>0.497097</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            counts     freqs\n",
       "categories                  \n",
       "Tg_24h       16806  0.502903\n",
       "UI_Tg_24h    16612  0.497097"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['stage_perInfection'].values.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "amateur-cradle",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 33418 × 36601\n",
       "    obs: 'sample', 'stage', 'hpi', 'infection', 'percent_mito', 'n_counts', 'sample_barcode', 'assignment_SoC', 'donor_id', 'scrublet_score', 'scrublet_cluster_score', 'zscore', 'bh_pval', 'bonf_pval', 'S_score', 'G2M_score', 'phase', 'n_genes_by_counts', 'total_counts', 'total_counts_hs', 'pct_counts_hs', 'total_counts_tg', 'pct_counts_tg', 'Tg_infected', 'n_genes', '_scvi_batch', '_scvi_labels', '_scvi_local_l_mean', '_scvi_local_l_var', 'leiden_scvi', 'celltype_predictions', 'probabilities', 'scrublet_doublet', 'cell_type_2022', 'cell_type', 'souporcell_MFgenotype', 'MFgenotype', 'cell_type_broad', 'umap_density_Tg_infected', 'stage_perInfection', 'celltype-Stage', 'Tg_intracellular', 'celltype-Intracellular', 'Dev_Stage'\n",
       "    var: 'gene_ids', 'feature_types', 'mean-0', 'std-0', 'mean-1', 'std-1', 'mean-2', 'std-2', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm'\n",
       "    uns: 'Dev_Stage_colors', 'MFgenotype_colors', '_scvi', 'cell_type_2022_colors', 'cell_type_broad_colors', 'cell_type_colors', 'celltype_predictions_colors', 'donor_id_colors', 'hpi_colors', 'hvg', 'infection_colors', 'leiden', 'neighbors', 'sample_colors', 'scrublet_doublet_colors', 'umap', 'umap_density_Tg_infected_params'\n",
       "    obsm: 'X_scVI_n_latent_20', 'X_scVI_n_latent_30', 'X_scVI_n_latent_40', 'X_umap', '_scvi_extra_categoricals'\n",
       "    layers: 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "floating-bandwidth",
   "metadata": {},
   "source": [
    "### DO NOT Subset dataset, use full matrix counts from infected manifod!! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "charged-operation",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "#Take only the INFECTED counts matrix\n",
    "cellphoneDB_adata = adata[adata.obs['stage_perInfection']=='Tg_24h']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "global-lindsay",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Take only the UNINFECTED counts matrix\n",
    "#cellphoneDB_adataUI = adata[adata.obs['stage_perInfection']=='UI_Tg_24h']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "widespread-property",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Pla_HDBR13007975_AAACCCAAGACACACG        HBC\n",
       "Pla_HDBR13007975_AAACCCACACTATCGA    VCT_CCC\n",
       "Pla_HDBR13007975_AAACCCACAGGGTTGA        HBC\n",
       "Pla_HDBR13007975_AAACCCACATCCGATA        HBC\n",
       "Pla_HDBR13007975_AAACCCAGTCCGAAGA          F\n",
       "                                      ...   \n",
       "Pla_HDBR13798224_TTTGTTGGTTTGCAGT        HBC\n",
       "Pla_HDBR13798224_TTTGTTGTCCCAAGCG     Endo_f\n",
       "Pla_HDBR13798224_TTTGTTGTCCGGCTTT        HBC\n",
       "Pla_HDBR13798224_TTTGTTGTCCTGATAG          F\n",
       "Pla_HDBR13798224_TTTGTTGTCTCATGGA      VCT_p\n",
       "Name: cell_type, Length: 16806, dtype: category\n",
       "Categories (15, object): ['EVT_1', 'EVT_2', 'Endo_f', 'F', ..., 'VCT_CCC', 'VCT_fusing', 'VCT_p', 'iEVT']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['cell_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "trying-vaccine",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>freqs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>categories</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>scDonor_Tg1</th>\n",
       "      <td>3722</td>\n",
       "      <td>0.221469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scDonor_Tg2</th>\n",
       "      <td>5144</td>\n",
       "      <td>0.306081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scDonor_Tg3</th>\n",
       "      <td>3549</td>\n",
       "      <td>0.211175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scDonor_Tg4</th>\n",
       "      <td>4391</td>\n",
       "      <td>0.261276</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             counts     freqs\n",
       "categories                   \n",
       "scDonor_Tg1    3722  0.221469\n",
       "scDonor_Tg2    5144  0.306081\n",
       "scDonor_Tg3    3549  0.211175\n",
       "scDonor_Tg4    4391  0.261276"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['donor_id'].values.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "close-classic",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>freqs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>categories</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tg_24h</th>\n",
       "      <td>16806</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            counts  freqs\n",
       "categories               \n",
       "Tg_24h       16806    1.0"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['stage_perInfection'].values.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "convinced-practice",
   "metadata": {},
   "source": [
    "### Load *new* cell clusters annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "twelve-stockholm",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    }
   ],
   "source": [
    "### Add cluster subname such that this new name will match with DEG cluster name\n",
    "\n",
    "#INFECTED\n",
    "cellphoneDB_adata.obs['cell_type']= cellphoneDB_adata.obs.cell_type.astype('string')+'.'+cellphoneDB_adata.obs.infection.astype('string')\n",
    "\n",
    "\n",
    "#UI\n",
    "#cellphoneDB_adataUI.obs['cell_type']= cellphoneDB_adataUI.obs.cell_type.astype('string')+'.'+cellphoneDB_adataUI.obs.stage_perInfection.astype('string')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "genuine-current",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Pla_HDBR13007975_AAACCCAAGACACACG        HBC.Tg\n",
       "Pla_HDBR13007975_AAACCCACACTATCGA    VCT_CCC.Tg\n",
       "Pla_HDBR13007975_AAACCCACAGGGTTGA        HBC.Tg\n",
       "Pla_HDBR13007975_AAACCCACATCCGATA        HBC.Tg\n",
       "Pla_HDBR13007975_AAACCCAGTCCGAAGA          F.Tg\n",
       "                                        ...    \n",
       "Pla_HDBR13798224_TTTGTTGGTTTGCAGT        HBC.Tg\n",
       "Pla_HDBR13798224_TTTGTTGTCCCAAGCG     Endo_f.Tg\n",
       "Pla_HDBR13798224_TTTGTTGTCCGGCTTT        HBC.Tg\n",
       "Pla_HDBR13798224_TTTGTTGTCCTGATAG          F.Tg\n",
       "Pla_HDBR13798224_TTTGTTGTCTCATGGA      VCT_p.Tg\n",
       "Name: cell_type, Length: 16806, dtype: string"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['cell_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "corresponding-trail",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16806 × 36601\n",
       "    obs: 'sample', 'stage', 'hpi', 'infection', 'percent_mito', 'n_counts', 'sample_barcode', 'assignment_SoC', 'donor_id', 'scrublet_score', 'scrublet_cluster_score', 'zscore', 'bh_pval', 'bonf_pval', 'S_score', 'G2M_score', 'phase', 'n_genes_by_counts', 'total_counts', 'total_counts_hs', 'pct_counts_hs', 'total_counts_tg', 'pct_counts_tg', 'Tg_infected', 'n_genes', '_scvi_batch', '_scvi_labels', '_scvi_local_l_mean', '_scvi_local_l_var', 'leiden_scvi', 'celltype_predictions', 'probabilities', 'scrublet_doublet', 'cell_type_2022', 'cell_type', 'souporcell_MFgenotype', 'MFgenotype', 'cell_type_broad', 'umap_density_Tg_infected', 'stage_perInfection', 'celltype-Stage', 'Tg_intracellular', 'celltype-Intracellular', 'Dev_Stage'\n",
       "    var: 'gene_ids', 'feature_types', 'mean-0', 'std-0', 'mean-1', 'std-1', 'mean-2', 'std-2', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm'\n",
       "    uns: 'Dev_Stage_colors', 'MFgenotype_colors', '_scvi', 'cell_type_2022_colors', 'cell_type_broad_colors', 'cell_type_colors', 'celltype_predictions_colors', 'donor_id_colors', 'hpi_colors', 'hvg', 'infection_colors', 'leiden', 'neighbors', 'sample_colors', 'scrublet_doublet_colors', 'umap', 'umap_density_Tg_infected_params'\n",
       "    obsm: 'X_scVI_n_latent_20', 'X_scVI_n_latent_30', 'X_scVI_n_latent_40', 'X_umap', '_scvi_extra_categoricals'\n",
       "    layers: 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "jewish-earthquake",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "varied-marshall",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "sc.pp.normalize_per_cell(cellphoneDB_adata)\n",
    "sc.pp.log1p(cellphoneDB_adata)                                                                                                            \n",
    "# sc.pp.scale(adata, max_value=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "quantitative-delta",
   "metadata": {},
   "source": [
    "### The anndata object contains counts that have been normalized (per cell) and log-transformed, therefore save this as below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "secondary-plenty",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'cell_type' as categorical\n"
     ]
    }
   ],
   "source": [
    "# counts input files \n",
    "cellphoneDB_adata.write('../data/toxoplasma_inf_normlogTransformed_24h.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "functioning-mystery",
   "metadata": {},
   "outputs": [],
   "source": [
    "# meta file\n",
    "df_meta = pd.DataFrame(data={'Cell':list(cellphoneDB_adata.obs.index),\n",
    "                             'cell_type':[ str(i) for i in cellphoneDB_adata.obs['cell_type']] })\n",
    "df_meta.set_index('Cell', inplace=True)\n",
    "df_meta.to_csv('../data/toxoplasma_inf_meta_24h.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "choice-brunei",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13007975_AAACCCAAGACACACG</th>\n",
       "      <td>HBC.Tg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13007975_AAACCCACACTATCGA</th>\n",
       "      <td>VCT_CCC.Tg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13007975_AAACCCACAGGGTTGA</th>\n",
       "      <td>HBC.Tg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13007975_AAACCCACATCCGATA</th>\n",
       "      <td>HBC.Tg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13007975_AAACCCAGTCCGAAGA</th>\n",
       "      <td>F.Tg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13798224_TTTGTTGGTTTGCAGT</th>\n",
       "      <td>HBC.Tg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13798224_TTTGTTGTCCCAAGCG</th>\n",
       "      <td>Endo_f.Tg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13798224_TTTGTTGTCCGGCTTT</th>\n",
       "      <td>HBC.Tg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13798224_TTTGTTGTCCTGATAG</th>\n",
       "      <td>F.Tg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13798224_TTTGTTGTCTCATGGA</th>\n",
       "      <td>VCT_p.Tg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16806 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    cell_type\n",
       "Cell                                         \n",
       "Pla_HDBR13007975_AAACCCAAGACACACG      HBC.Tg\n",
       "Pla_HDBR13007975_AAACCCACACTATCGA  VCT_CCC.Tg\n",
       "Pla_HDBR13007975_AAACCCACAGGGTTGA      HBC.Tg\n",
       "Pla_HDBR13007975_AAACCCACATCCGATA      HBC.Tg\n",
       "Pla_HDBR13007975_AAACCCAGTCCGAAGA        F.Tg\n",
       "...                                       ...\n",
       "Pla_HDBR13798224_TTTGTTGGTTTGCAGT      HBC.Tg\n",
       "Pla_HDBR13798224_TTTGTTGTCCCAAGCG   Endo_f.Tg\n",
       "Pla_HDBR13798224_TTTGTTGTCCGGCTTT      HBC.Tg\n",
       "Pla_HDBR13798224_TTTGTTGTCCTGATAG        F.Tg\n",
       "Pla_HDBR13798224_TTTGTTGTCTCATGGA    VCT_p.Tg\n",
       "\n",
       "[16806 rows x 1 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cutting-junior",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "blond-motion",
   "metadata": {},
   "source": [
    "### (2) Do DEGs individually from precalculated DEGs files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "suitable-bridge",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PATH to immune cells\n",
    "path='../../diffGeneExpression/macrophages/results/SC_limma_Immune_TOX_24h/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "common-joseph",
   "metadata": {},
   "source": [
    "### HB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "electoral-clear",
   "metadata": {},
   "outputs": [],
   "source": [
    "HBDEG = pd.read_csv(path+'1_DEGs_SC_TOX_24h_HBC.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ancient-switzerland",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>COL3A1</th>\n",
       "      <td>0.796033</td>\n",
       "      <td>6.282735</td>\n",
       "      <td>136.443574</td>\n",
       "      <td>2.289547e-11</td>\n",
       "      <td>1.689686e-07</td>\n",
       "      <td>1.689686e-07</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Tg</td>\n",
       "      <td>COL3A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL1A1</th>\n",
       "      <td>0.847071</td>\n",
       "      <td>4.494423</td>\n",
       "      <td>70.662051</td>\n",
       "      <td>1.338823e-08</td>\n",
       "      <td>9.880516e-05</td>\n",
       "      <td>4.940258e-05</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Tg</td>\n",
       "      <td>COL1A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DYNLRB1</th>\n",
       "      <td>0.332891</td>\n",
       "      <td>5.848545</td>\n",
       "      <td>57.359292</td>\n",
       "      <td>8.412154e-08</td>\n",
       "      <td>6.208170e-04</td>\n",
       "      <td>1.723061e-04</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Tg</td>\n",
       "      <td>DYNLRB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYO1G</th>\n",
       "      <td>0.608635</td>\n",
       "      <td>3.957769</td>\n",
       "      <td>53.631940</td>\n",
       "      <td>1.487998e-07</td>\n",
       "      <td>1.098143e-03</td>\n",
       "      <td>2.196286e-04</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Tg</td>\n",
       "      <td>MYO1G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DTNA</th>\n",
       "      <td>0.335151</td>\n",
       "      <td>7.616277</td>\n",
       "      <td>50.481931</td>\n",
       "      <td>2.463888e-07</td>\n",
       "      <td>1.818350e-03</td>\n",
       "      <td>2.597642e-04</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Tg</td>\n",
       "      <td>DTNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            logFC    logCPM           F        PValue    bonferroni  \\\n",
       "COL3A1   0.796033  6.282735  136.443574  2.289547e-11  1.689686e-07   \n",
       "COL1A1   0.847071  4.494423   70.662051  1.338823e-08  9.880516e-05   \n",
       "DYNLRB1  0.332891  5.848545   57.359292  8.412154e-08  6.208170e-04   \n",
       "MYO1G    0.608635  3.957769   53.631940  1.487998e-07  1.098143e-03   \n",
       "DTNA     0.335151  7.616277   50.481931  2.463888e-07  1.818350e-03   \n",
       "\n",
       "            p_val_adj cell_type cluster     gene  \n",
       "COL3A1   1.689686e-07       HBC  HBC.Tg   COL3A1  \n",
       "COL1A1   4.940258e-05       HBC  HBC.Tg   COL1A1  \n",
       "DYNLRB1  1.723061e-04       HBC  HBC.Tg  DYNLRB1  \n",
       "MYO1G    2.196286e-04       HBC  HBC.Tg    MYO1G  \n",
       "DTNA     2.597642e-04       HBC  HBC.Tg     DTNA  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HBDEG_up= recoverDEGs(HBDEG, infection='Tg', celltype='HBC', direction='up')\n",
    "HBDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "neutral-allah",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in HBDEG_up.cluster])\n",
    "HBDEG_up = HBDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "raising-fossil",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = HBDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "HBDEG_up= HBDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(HBDEG_up).to_csv('../data/toxoplasma_inf_HBC_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "czech-mediterranean",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>COL3A1</th>\n",
       "      <td>HBC.Tg</td>\n",
       "      <td>COL3A1</td>\n",
       "      <td>0.796033</td>\n",
       "      <td>6.282735</td>\n",
       "      <td>136.443574</td>\n",
       "      <td>2.289547e-11</td>\n",
       "      <td>1.689686e-07</td>\n",
       "      <td>1.689686e-07</td>\n",
       "      <td>HBC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL1A1</th>\n",
       "      <td>HBC.Tg</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>0.847071</td>\n",
       "      <td>4.494423</td>\n",
       "      <td>70.662051</td>\n",
       "      <td>1.338823e-08</td>\n",
       "      <td>9.880516e-05</td>\n",
       "      <td>4.940258e-05</td>\n",
       "      <td>HBC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DYNLRB1</th>\n",
       "      <td>HBC.Tg</td>\n",
       "      <td>DYNLRB1</td>\n",
       "      <td>0.332891</td>\n",
       "      <td>5.848545</td>\n",
       "      <td>57.359292</td>\n",
       "      <td>8.412154e-08</td>\n",
       "      <td>6.208170e-04</td>\n",
       "      <td>1.723061e-04</td>\n",
       "      <td>HBC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYO1G</th>\n",
       "      <td>HBC.Tg</td>\n",
       "      <td>MYO1G</td>\n",
       "      <td>0.608635</td>\n",
       "      <td>3.957769</td>\n",
       "      <td>53.631940</td>\n",
       "      <td>1.487998e-07</td>\n",
       "      <td>1.098143e-03</td>\n",
       "      <td>2.196286e-04</td>\n",
       "      <td>HBC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DTNA</th>\n",
       "      <td>HBC.Tg</td>\n",
       "      <td>DTNA</td>\n",
       "      <td>0.335151</td>\n",
       "      <td>7.616277</td>\n",
       "      <td>50.481931</td>\n",
       "      <td>2.463888e-07</td>\n",
       "      <td>1.818350e-03</td>\n",
       "      <td>2.597642e-04</td>\n",
       "      <td>HBC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        cluster     gene     logFC    logCPM           F        PValue  \\\n",
       "COL3A1   HBC.Tg   COL3A1  0.796033  6.282735  136.443574  2.289547e-11   \n",
       "COL1A1   HBC.Tg   COL1A1  0.847071  4.494423   70.662051  1.338823e-08   \n",
       "DYNLRB1  HBC.Tg  DYNLRB1  0.332891  5.848545   57.359292  8.412154e-08   \n",
       "MYO1G    HBC.Tg    MYO1G  0.608635  3.957769   53.631940  1.487998e-07   \n",
       "DTNA     HBC.Tg     DTNA  0.335151  7.616277   50.481931  2.463888e-07   \n",
       "\n",
       "           bonferroni     p_val_adj cell_type  \n",
       "COL3A1   1.689686e-07  1.689686e-07       HBC  \n",
       "COL1A1   9.880516e-05  4.940258e-05       HBC  \n",
       "DYNLRB1  6.208170e-04  1.723061e-04       HBC  \n",
       "MYO1G    1.098143e-03  2.196286e-04       HBC  \n",
       "DTNA     1.818350e-03  2.597642e-04       HBC  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HBDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "north-optimum",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cluster, gene, logFC, logCPM, F, PValue, bonferroni, p_val_adj, cell_type]\n",
       "Index: []"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#HBDEG_up=pd.read_table('../data/toxoplasma_inf_HBC_DEGsup.tsv')\n",
    "#HBDEG_up.loc[['VEGFA']]\n",
    "HBDEG_up.loc[HBDEG_up['gene'] == 'ISG15']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aquatic-insured",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "extraordinary-characteristic",
   "metadata": {},
   "source": [
    "### PAMM1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "regulated-ambassador",
   "metadata": {},
   "outputs": [],
   "source": [
    "PAMM1DEG = pd.read_csv(path+'2_DEGs_SC_TOX_24h_PAMM1.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "coupled-masters",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MET</th>\n",
       "      <td>1.003315</td>\n",
       "      <td>7.576231</td>\n",
       "      <td>30.267887</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.037725</td>\n",
       "      <td>0.003399</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Tg</td>\n",
       "      <td>MET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SERPINB9</th>\n",
       "      <td>0.606788</td>\n",
       "      <td>6.309922</td>\n",
       "      <td>31.040125</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.031063</td>\n",
       "      <td>0.003399</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Tg</td>\n",
       "      <td>SERPINB9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOCK8</th>\n",
       "      <td>0.589602</td>\n",
       "      <td>9.475681</td>\n",
       "      <td>29.358520</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.047588</td>\n",
       "      <td>0.003399</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Tg</td>\n",
       "      <td>DOCK8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDE3B</th>\n",
       "      <td>0.661068</td>\n",
       "      <td>7.912461</td>\n",
       "      <td>25.186082</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.145396</td>\n",
       "      <td>0.006322</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Tg</td>\n",
       "      <td>PDE3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CXCL5</th>\n",
       "      <td>0.649151</td>\n",
       "      <td>13.082072</td>\n",
       "      <td>25.270278</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.142030</td>\n",
       "      <td>0.006322</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Tg</td>\n",
       "      <td>CXCL5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             logFC     logCPM          F    PValue  bonferroni  p_val_adj  \\\n",
       "MET       1.003315   7.576231  30.267887  0.000006    0.037725   0.003399   \n",
       "SERPINB9  0.606788   6.309922  31.040125  0.000005    0.031063   0.003399   \n",
       "DOCK8     0.589602   9.475681  29.358520  0.000008    0.047588   0.003399   \n",
       "PDE3B     0.661068   7.912461  25.186082  0.000023    0.145396   0.006322   \n",
       "CXCL5     0.649151  13.082072  25.270278  0.000023    0.142030   0.006322   \n",
       "\n",
       "         cell_type   cluster      gene  \n",
       "MET          PAMM1  PAMM1.Tg       MET  \n",
       "SERPINB9     PAMM1  PAMM1.Tg  SERPINB9  \n",
       "DOCK8        PAMM1  PAMM1.Tg     DOCK8  \n",
       "PDE3B        PAMM1  PAMM1.Tg     PDE3B  \n",
       "CXCL5        PAMM1  PAMM1.Tg     CXCL5  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PAMM1DEG_up= recoverDEGs(PAMM1DEG, infection='Tg', celltype='PAMM1', direction='up')\n",
    "PAMM1DEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "selected-marina",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in PAMM1DEG_up.cluster])\n",
    "PAMM1DEG_up = PAMM1DEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "productive-response",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = PAMM1DEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "PAMM1DEG_up= PAMM1DEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(PAMM1DEG_up).to_csv('../data/toxoplasma_inf_PAMM1_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "accompanied-rochester",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MET</th>\n",
       "      <td>PAMM1.Tg</td>\n",
       "      <td>MET</td>\n",
       "      <td>1.003315</td>\n",
       "      <td>7.576231</td>\n",
       "      <td>30.267887</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.037725</td>\n",
       "      <td>0.003399</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SERPINB9</th>\n",
       "      <td>PAMM1.Tg</td>\n",
       "      <td>SERPINB9</td>\n",
       "      <td>0.606788</td>\n",
       "      <td>6.309922</td>\n",
       "      <td>31.040125</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.031063</td>\n",
       "      <td>0.003399</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOCK8</th>\n",
       "      <td>PAMM1.Tg</td>\n",
       "      <td>DOCK8</td>\n",
       "      <td>0.589602</td>\n",
       "      <td>9.475681</td>\n",
       "      <td>29.358520</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.047588</td>\n",
       "      <td>0.003399</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDE3B</th>\n",
       "      <td>PAMM1.Tg</td>\n",
       "      <td>PDE3B</td>\n",
       "      <td>0.661068</td>\n",
       "      <td>7.912461</td>\n",
       "      <td>25.186082</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.145396</td>\n",
       "      <td>0.006322</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CXCL5</th>\n",
       "      <td>PAMM1.Tg</td>\n",
       "      <td>CXCL5</td>\n",
       "      <td>0.649151</td>\n",
       "      <td>13.082072</td>\n",
       "      <td>25.270278</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.142030</td>\n",
       "      <td>0.006322</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           cluster      gene     logFC     logCPM          F    PValue  \\\n",
       "MET       PAMM1.Tg       MET  1.003315   7.576231  30.267887  0.000006   \n",
       "SERPINB9  PAMM1.Tg  SERPINB9  0.606788   6.309922  31.040125  0.000005   \n",
       "DOCK8     PAMM1.Tg     DOCK8  0.589602   9.475681  29.358520  0.000008   \n",
       "PDE3B     PAMM1.Tg     PDE3B  0.661068   7.912461  25.186082  0.000023   \n",
       "CXCL5     PAMM1.Tg     CXCL5  0.649151  13.082072  25.270278  0.000023   \n",
       "\n",
       "          bonferroni  p_val_adj cell_type  \n",
       "MET         0.037725   0.003399     PAMM1  \n",
       "SERPINB9    0.031063   0.003399     PAMM1  \n",
       "DOCK8       0.047588   0.003399     PAMM1  \n",
       "PDE3B       0.145396   0.006322     PAMM1  \n",
       "CXCL5       0.142030   0.006322     PAMM1  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PAMM1DEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bored-bedroom",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adapted-courage",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "express-kazakhstan",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "aggressive-funeral",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PATH to FibroEndo cells DGE tables\n",
    "path='../../diffGeneExpression/fibroEndo/results/SC_limma_fibroEndo_TOX_24h/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cooked-mainstream",
   "metadata": {},
   "source": [
    "### F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "spanish-classification",
   "metadata": {},
   "outputs": [],
   "source": [
    "FDEG = pd.read_csv(path+'/1_DEGs_SC_TOX_24h_F.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "crude-contract",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PLOD2</th>\n",
       "      <td>0.346437</td>\n",
       "      <td>9.710670</td>\n",
       "      <td>57.301917</td>\n",
       "      <td>4.569245e-08</td>\n",
       "      <td>0.000318</td>\n",
       "      <td>0.000318</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Tg</td>\n",
       "      <td>PLOD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGAM1</th>\n",
       "      <td>0.294528</td>\n",
       "      <td>7.414381</td>\n",
       "      <td>49.648052</td>\n",
       "      <td>1.649194e-07</td>\n",
       "      <td>0.001149</td>\n",
       "      <td>0.000574</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Tg</td>\n",
       "      <td>PGAM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNX9</th>\n",
       "      <td>0.359026</td>\n",
       "      <td>7.800909</td>\n",
       "      <td>45.928065</td>\n",
       "      <td>3.231338e-07</td>\n",
       "      <td>0.002251</td>\n",
       "      <td>0.000736</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Tg</td>\n",
       "      <td>SNX9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOTCH2</th>\n",
       "      <td>0.300958</td>\n",
       "      <td>6.890188</td>\n",
       "      <td>42.552030</td>\n",
       "      <td>6.140650e-07</td>\n",
       "      <td>0.004278</td>\n",
       "      <td>0.000856</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Tg</td>\n",
       "      <td>NOTCH2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYO10</th>\n",
       "      <td>0.288179</td>\n",
       "      <td>10.052584</td>\n",
       "      <td>39.956914</td>\n",
       "      <td>1.028769e-06</td>\n",
       "      <td>0.007166</td>\n",
       "      <td>0.000896</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Tg</td>\n",
       "      <td>MYO10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           logFC     logCPM          F        PValue  bonferroni  p_val_adj  \\\n",
       "PLOD2   0.346437   9.710670  57.301917  4.569245e-08    0.000318   0.000318   \n",
       "PGAM1   0.294528   7.414381  49.648052  1.649194e-07    0.001149   0.000574   \n",
       "SNX9    0.359026   7.800909  45.928065  3.231338e-07    0.002251   0.000736   \n",
       "NOTCH2  0.300958   6.890188  42.552030  6.140650e-07    0.004278   0.000856   \n",
       "MYO10   0.288179  10.052584  39.956914  1.028769e-06    0.007166   0.000896   \n",
       "\n",
       "       cell_type cluster    gene  \n",
       "PLOD2          F    F.Tg   PLOD2  \n",
       "PGAM1          F    F.Tg   PGAM1  \n",
       "SNX9           F    F.Tg    SNX9  \n",
       "NOTCH2         F    F.Tg  NOTCH2  \n",
       "MYO10          F    F.Tg   MYO10  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FDEG_up= recoverDEGs(FDEG, infection='Tg', celltype='F', direction='up')\n",
    "FDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "turkish-bundle",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in FDEG_up.cluster])\n",
    "FDEG_up = FDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "theoretical-village",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = FDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "FDEG_up= FDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(FDEG_up).to_csv('../data/toxoplasma_inf_F_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "athletic-senator",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PLOD2</th>\n",
       "      <td>F.Tg</td>\n",
       "      <td>PLOD2</td>\n",
       "      <td>0.346437</td>\n",
       "      <td>9.710670</td>\n",
       "      <td>57.301917</td>\n",
       "      <td>4.569245e-08</td>\n",
       "      <td>0.000318</td>\n",
       "      <td>0.000318</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGAM1</th>\n",
       "      <td>F.Tg</td>\n",
       "      <td>PGAM1</td>\n",
       "      <td>0.294528</td>\n",
       "      <td>7.414381</td>\n",
       "      <td>49.648052</td>\n",
       "      <td>1.649194e-07</td>\n",
       "      <td>0.001149</td>\n",
       "      <td>0.000574</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNX9</th>\n",
       "      <td>F.Tg</td>\n",
       "      <td>SNX9</td>\n",
       "      <td>0.359026</td>\n",
       "      <td>7.800909</td>\n",
       "      <td>45.928065</td>\n",
       "      <td>3.231338e-07</td>\n",
       "      <td>0.002251</td>\n",
       "      <td>0.000736</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOTCH2</th>\n",
       "      <td>F.Tg</td>\n",
       "      <td>NOTCH2</td>\n",
       "      <td>0.300958</td>\n",
       "      <td>6.890188</td>\n",
       "      <td>42.552030</td>\n",
       "      <td>6.140650e-07</td>\n",
       "      <td>0.004278</td>\n",
       "      <td>0.000856</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYO10</th>\n",
       "      <td>F.Tg</td>\n",
       "      <td>MYO10</td>\n",
       "      <td>0.288179</td>\n",
       "      <td>10.052584</td>\n",
       "      <td>39.956914</td>\n",
       "      <td>1.028769e-06</td>\n",
       "      <td>0.007166</td>\n",
       "      <td>0.000896</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster    gene     logFC     logCPM          F        PValue  \\\n",
       "PLOD2     F.Tg   PLOD2  0.346437   9.710670  57.301917  4.569245e-08   \n",
       "PGAM1     F.Tg   PGAM1  0.294528   7.414381  49.648052  1.649194e-07   \n",
       "SNX9      F.Tg    SNX9  0.359026   7.800909  45.928065  3.231338e-07   \n",
       "NOTCH2    F.Tg  NOTCH2  0.300958   6.890188  42.552030  6.140650e-07   \n",
       "MYO10     F.Tg   MYO10  0.288179  10.052584  39.956914  1.028769e-06   \n",
       "\n",
       "        bonferroni  p_val_adj cell_type  \n",
       "PLOD2     0.000318   0.000318         F  \n",
       "PGAM1     0.001149   0.000574         F  \n",
       "SNX9      0.002251   0.000736         F  \n",
       "NOTCH2    0.004278   0.000856         F  \n",
       "MYO10     0.007166   0.000896         F  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "desirable-eating",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PGAM1</th>\n",
       "      <td>F.Tg</td>\n",
       "      <td>PGAM1</td>\n",
       "      <td>0.294528</td>\n",
       "      <td>7.414381</td>\n",
       "      <td>49.648052</td>\n",
       "      <td>1.649194e-07</td>\n",
       "      <td>0.001149</td>\n",
       "      <td>0.000574</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cluster   gene     logFC    logCPM          F        PValue  bonferroni  \\\n",
       "PGAM1    F.Tg  PGAM1  0.294528  7.414381  49.648052  1.649194e-07    0.001149   \n",
       "\n",
       "       p_val_adj cell_type  \n",
       "PGAM1   0.000574         F  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FDEG_up.loc[['PGAM1']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "virtual-broadcasting",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "continent-kuwait",
   "metadata": {},
   "source": [
    "### PV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "automatic-tanzania",
   "metadata": {},
   "outputs": [],
   "source": [
    "PVDEG = pd.read_csv(path+'/2_DEGs_SC_TOX_24h_PV.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "instrumental-broadway",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD68</th>\n",
       "      <td>0.778783</td>\n",
       "      <td>4.457690</td>\n",
       "      <td>56.390919</td>\n",
       "      <td>3.322928e-08</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Tg</td>\n",
       "      <td>CD68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SERPINE1</th>\n",
       "      <td>0.596916</td>\n",
       "      <td>10.631820</td>\n",
       "      <td>42.412198</td>\n",
       "      <td>4.462031e-07</td>\n",
       "      <td>0.003306</td>\n",
       "      <td>0.001102</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Tg</td>\n",
       "      <td>SERPINE1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPP1</th>\n",
       "      <td>0.924016</td>\n",
       "      <td>7.838992</td>\n",
       "      <td>39.349764</td>\n",
       "      <td>8.459709e-07</td>\n",
       "      <td>0.006268</td>\n",
       "      <td>0.001254</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Tg</td>\n",
       "      <td>SPP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGAM1</th>\n",
       "      <td>0.393088</td>\n",
       "      <td>7.736512</td>\n",
       "      <td>37.133623</td>\n",
       "      <td>1.369823e-06</td>\n",
       "      <td>0.010149</td>\n",
       "      <td>0.001362</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Tg</td>\n",
       "      <td>PGAM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNASE1</th>\n",
       "      <td>0.628468</td>\n",
       "      <td>5.194213</td>\n",
       "      <td>37.819184</td>\n",
       "      <td>1.177979e-06</td>\n",
       "      <td>0.008728</td>\n",
       "      <td>0.001362</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Tg</td>\n",
       "      <td>RNASE1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             logFC     logCPM          F        PValue  bonferroni  p_val_adj  \\\n",
       "CD68      0.778783   4.457690  56.390919  3.322928e-08    0.000246   0.000246   \n",
       "SERPINE1  0.596916  10.631820  42.412198  4.462031e-07    0.003306   0.001102   \n",
       "SPP1      0.924016   7.838992  39.349764  8.459709e-07    0.006268   0.001254   \n",
       "PGAM1     0.393088   7.736512  37.133623  1.369823e-06    0.010149   0.001362   \n",
       "RNASE1    0.628468   5.194213  37.819184  1.177979e-06    0.008728   0.001362   \n",
       "\n",
       "         cell_type cluster      gene  \n",
       "CD68            PV   PV.Tg      CD68  \n",
       "SERPINE1        PV   PV.Tg  SERPINE1  \n",
       "SPP1            PV   PV.Tg      SPP1  \n",
       "PGAM1           PV   PV.Tg     PGAM1  \n",
       "RNASE1          PV   PV.Tg    RNASE1  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PVDEG_up= recoverDEGs(PVDEG, infection='Tg', celltype='PV', direction='up')\n",
    "PVDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "vanilla-occurrence",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in PVDEG_up.cluster])\n",
    "PVDEG_up = PVDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "worthy-society",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = PVDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "PVDEG_up= PVDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(PVDEG_up).to_csv('../data/toxoplasma_inf_PV_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "paperback-cover",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD68</th>\n",
       "      <td>PV.Tg</td>\n",
       "      <td>CD68</td>\n",
       "      <td>0.778783</td>\n",
       "      <td>4.457690</td>\n",
       "      <td>56.390919</td>\n",
       "      <td>3.322928e-08</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SERPINE1</th>\n",
       "      <td>PV.Tg</td>\n",
       "      <td>SERPINE1</td>\n",
       "      <td>0.596916</td>\n",
       "      <td>10.631820</td>\n",
       "      <td>42.412198</td>\n",
       "      <td>4.462031e-07</td>\n",
       "      <td>0.003306</td>\n",
       "      <td>0.001102</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPP1</th>\n",
       "      <td>PV.Tg</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>0.924016</td>\n",
       "      <td>7.838992</td>\n",
       "      <td>39.349764</td>\n",
       "      <td>8.459709e-07</td>\n",
       "      <td>0.006268</td>\n",
       "      <td>0.001254</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGAM1</th>\n",
       "      <td>PV.Tg</td>\n",
       "      <td>PGAM1</td>\n",
       "      <td>0.393088</td>\n",
       "      <td>7.736512</td>\n",
       "      <td>37.133623</td>\n",
       "      <td>1.369823e-06</td>\n",
       "      <td>0.010149</td>\n",
       "      <td>0.001362</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNASE1</th>\n",
       "      <td>PV.Tg</td>\n",
       "      <td>RNASE1</td>\n",
       "      <td>0.628468</td>\n",
       "      <td>5.194213</td>\n",
       "      <td>37.819184</td>\n",
       "      <td>1.177979e-06</td>\n",
       "      <td>0.008728</td>\n",
       "      <td>0.001362</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         cluster      gene     logFC     logCPM          F        PValue  \\\n",
       "CD68       PV.Tg      CD68  0.778783   4.457690  56.390919  3.322928e-08   \n",
       "SERPINE1   PV.Tg  SERPINE1  0.596916  10.631820  42.412198  4.462031e-07   \n",
       "SPP1       PV.Tg      SPP1  0.924016   7.838992  39.349764  8.459709e-07   \n",
       "PGAM1      PV.Tg     PGAM1  0.393088   7.736512  37.133623  1.369823e-06   \n",
       "RNASE1     PV.Tg    RNASE1  0.628468   5.194213  37.819184  1.177979e-06   \n",
       "\n",
       "          bonferroni  p_val_adj cell_type  \n",
       "CD68        0.000246   0.000246        PV  \n",
       "SERPINE1    0.003306   0.001102        PV  \n",
       "SPP1        0.006268   0.001254        PV  \n",
       "PGAM1       0.010149   0.001362        PV  \n",
       "RNASE1      0.008728   0.001362        PV  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PVDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "painful-pierce",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "conscious-consideration",
   "metadata": {},
   "source": [
    "### Endo_f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "surgical-plane",
   "metadata": {},
   "outputs": [],
   "source": [
    "EndoDEG = pd.read_csv(path+'/3_DEGs_SC_TOX_24h_Endof.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "satisfied-contents",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MT-ATP6</th>\n",
       "      <td>0.395582</td>\n",
       "      <td>12.310631</td>\n",
       "      <td>29.299554</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.060860</td>\n",
       "      <td>0.010143</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Tg</td>\n",
       "      <td>MT-ATP6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TM4SF1</th>\n",
       "      <td>0.261449</td>\n",
       "      <td>10.779658</td>\n",
       "      <td>25.942360</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.147604</td>\n",
       "      <td>0.018174</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Tg</td>\n",
       "      <td>TM4SF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHF5A</th>\n",
       "      <td>0.352516</td>\n",
       "      <td>4.803430</td>\n",
       "      <td>25.176129</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.182148</td>\n",
       "      <td>0.018174</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Tg</td>\n",
       "      <td>PHF5A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CO2</th>\n",
       "      <td>0.256410</td>\n",
       "      <td>12.635936</td>\n",
       "      <td>25.523726</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.165510</td>\n",
       "      <td>0.018174</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Tg</td>\n",
       "      <td>MT-CO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL1A2</th>\n",
       "      <td>0.590334</td>\n",
       "      <td>4.731228</td>\n",
       "      <td>24.484398</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>0.220835</td>\n",
       "      <td>0.018403</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Tg</td>\n",
       "      <td>COL1A2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            logFC     logCPM          F    PValue  bonferroni  p_val_adj  \\\n",
       "MT-ATP6  0.395582  12.310631  29.299554  0.000008    0.060860   0.010143   \n",
       "TM4SF1   0.261449  10.779658  25.942360  0.000020    0.147604   0.018174   \n",
       "PHF5A    0.352516   4.803430  25.176129  0.000024    0.182148   0.018174   \n",
       "MT-CO2   0.256410  12.635936  25.523726  0.000022    0.165510   0.018174   \n",
       "COL1A2   0.590334   4.731228  24.484398  0.000030    0.220835   0.018403   \n",
       "\n",
       "        cell_type    cluster     gene  \n",
       "MT-ATP6    Endo_f  Endo_f.Tg  MT-ATP6  \n",
       "TM4SF1     Endo_f  Endo_f.Tg   TM4SF1  \n",
       "PHF5A      Endo_f  Endo_f.Tg    PHF5A  \n",
       "MT-CO2     Endo_f  Endo_f.Tg   MT-CO2  \n",
       "COL1A2     Endo_f  Endo_f.Tg   COL1A2  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EndoDEG_up= recoverDEGs(EndoDEG, infection='Tg', celltype='Endo_f', direction='up')\n",
    "EndoDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "synthetic-charter",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in EndoDEG_up.cluster])\n",
    "EndoDEG_up = EndoDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "silver-grenada",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = EndoDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "EndoDEG_up= EndoDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(EndoDEG_up).to_csv('../data/toxoplasma_inf_Endof_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "religious-slave",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MT-ATP6</th>\n",
       "      <td>Endo_f.Tg</td>\n",
       "      <td>MT-ATP6</td>\n",
       "      <td>0.395582</td>\n",
       "      <td>12.310631</td>\n",
       "      <td>29.299554</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.060860</td>\n",
       "      <td>0.010143</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TM4SF1</th>\n",
       "      <td>Endo_f.Tg</td>\n",
       "      <td>TM4SF1</td>\n",
       "      <td>0.261449</td>\n",
       "      <td>10.779658</td>\n",
       "      <td>25.942360</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.147604</td>\n",
       "      <td>0.018174</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHF5A</th>\n",
       "      <td>Endo_f.Tg</td>\n",
       "      <td>PHF5A</td>\n",
       "      <td>0.352516</td>\n",
       "      <td>4.803430</td>\n",
       "      <td>25.176129</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.182148</td>\n",
       "      <td>0.018174</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CO2</th>\n",
       "      <td>Endo_f.Tg</td>\n",
       "      <td>MT-CO2</td>\n",
       "      <td>0.256410</td>\n",
       "      <td>12.635936</td>\n",
       "      <td>25.523726</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.165510</td>\n",
       "      <td>0.018174</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL1A2</th>\n",
       "      <td>Endo_f.Tg</td>\n",
       "      <td>COL1A2</td>\n",
       "      <td>0.590334</td>\n",
       "      <td>4.731228</td>\n",
       "      <td>24.484398</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>0.220835</td>\n",
       "      <td>0.018403</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           cluster     gene     logFC     logCPM          F    PValue  \\\n",
       "MT-ATP6  Endo_f.Tg  MT-ATP6  0.395582  12.310631  29.299554  0.000008   \n",
       "TM4SF1   Endo_f.Tg   TM4SF1  0.261449  10.779658  25.942360  0.000020   \n",
       "PHF5A    Endo_f.Tg    PHF5A  0.352516   4.803430  25.176129  0.000024   \n",
       "MT-CO2   Endo_f.Tg   MT-CO2  0.256410  12.635936  25.523726  0.000022   \n",
       "COL1A2   Endo_f.Tg   COL1A2  0.590334   4.731228  24.484398  0.000030   \n",
       "\n",
       "         bonferroni  p_val_adj cell_type  \n",
       "MT-ATP6    0.060860   0.010143    Endo_f  \n",
       "TM4SF1     0.147604   0.018174    Endo_f  \n",
       "PHF5A      0.182148   0.018174    Endo_f  \n",
       "MT-CO2     0.165510   0.018174    Endo_f  \n",
       "COL1A2     0.220835   0.018403    Endo_f  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EndoDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "intense-power",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PHF5A</th>\n",
       "      <td>Endo_f.Tg</td>\n",
       "      <td>PHF5A</td>\n",
       "      <td>0.352516</td>\n",
       "      <td>4.80343</td>\n",
       "      <td>25.176129</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.182148</td>\n",
       "      <td>0.018174</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         cluster   gene     logFC   logCPM          F    PValue  bonferroni  \\\n",
       "PHF5A  Endo_f.Tg  PHF5A  0.352516  4.80343  25.176129  0.000024    0.182148   \n",
       "\n",
       "       p_val_adj cell_type  \n",
       "PHF5A   0.018174    Endo_f  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EndoDEG_up.loc[['PHF5A']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "signed-falls",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "everyday-punch",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "moving-chinese",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "confused-florida",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "attended-supervision",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PATH to Trophoblast cells DGE tables\n",
    "path='../../diffGeneExpression/trophoblast/results/SC_limma_Trophoblast_TOX_24h/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "generous-cholesterol",
   "metadata": {},
   "source": [
    "### VCT_fusing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "stone-cause",
   "metadata": {},
   "outputs": [],
   "source": [
    "VCTfDEG = pd.read_csv(path+'/1_DEGs_SC_TOX_24h_VCT_fusing.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "lined-aluminum",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TIMP1</th>\n",
       "      <td>0.883189</td>\n",
       "      <td>6.423208</td>\n",
       "      <td>88.019460</td>\n",
       "      <td>1.126633e-09</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Tg</td>\n",
       "      <td>TIMP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IFI6</th>\n",
       "      <td>0.448185</td>\n",
       "      <td>10.421668</td>\n",
       "      <td>55.318923</td>\n",
       "      <td>8.556514e-08</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>0.000326</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Tg</td>\n",
       "      <td>IFI6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPP1</th>\n",
       "      <td>0.717596</td>\n",
       "      <td>6.877959</td>\n",
       "      <td>47.980641</td>\n",
       "      <td>2.897721e-07</td>\n",
       "      <td>0.002207</td>\n",
       "      <td>0.000736</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Tg</td>\n",
       "      <td>SPP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL3A1</th>\n",
       "      <td>0.814079</td>\n",
       "      <td>5.123309</td>\n",
       "      <td>43.569166</td>\n",
       "      <td>6.419772e-07</td>\n",
       "      <td>0.004890</td>\n",
       "      <td>0.001222</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Tg</td>\n",
       "      <td>COL3A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DYNLRB1</th>\n",
       "      <td>0.418769</td>\n",
       "      <td>5.728626</td>\n",
       "      <td>39.369841</td>\n",
       "      <td>1.439906e-06</td>\n",
       "      <td>0.010968</td>\n",
       "      <td>0.002194</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Tg</td>\n",
       "      <td>DYNLRB1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            logFC     logCPM          F        PValue  bonferroni  p_val_adj  \\\n",
       "TIMP1    0.883189   6.423208  88.019460  1.126633e-09    0.000009   0.000009   \n",
       "IFI6     0.448185  10.421668  55.318923  8.556514e-08    0.000652   0.000326   \n",
       "SPP1     0.717596   6.877959  47.980641  2.897721e-07    0.002207   0.000736   \n",
       "COL3A1   0.814079   5.123309  43.569166  6.419772e-07    0.004890   0.001222   \n",
       "DYNLRB1  0.418769   5.728626  39.369841  1.439906e-06    0.010968   0.002194   \n",
       "\n",
       "          cell_type        cluster     gene  \n",
       "TIMP1    VCT_fusing  VCT_fusing.Tg    TIMP1  \n",
       "IFI6     VCT_fusing  VCT_fusing.Tg     IFI6  \n",
       "SPP1     VCT_fusing  VCT_fusing.Tg     SPP1  \n",
       "COL3A1   VCT_fusing  VCT_fusing.Tg   COL3A1  \n",
       "DYNLRB1  VCT_fusing  VCT_fusing.Tg  DYNLRB1  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTfDEG_up= recoverDEGs(VCTfDEG, infection='Tg', celltype='VCT_fusing', direction='up')\n",
    "VCTfDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "complete-intervention",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in VCTfDEG_up.cluster])\n",
    "VCTfVCTfDEG_upDEG = VCTfDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "known-fundamental",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = VCTfDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "VCTfDEG_up= VCTfDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(VCTfDEG_up).to_csv('../data/toxoplasma_inf_VCTfusing_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "behind-settle",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TIMP1</th>\n",
       "      <td>VCT_fusing.Tg</td>\n",
       "      <td>TIMP1</td>\n",
       "      <td>0.883189</td>\n",
       "      <td>6.423208</td>\n",
       "      <td>88.019460</td>\n",
       "      <td>1.126633e-09</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IFI6</th>\n",
       "      <td>VCT_fusing.Tg</td>\n",
       "      <td>IFI6</td>\n",
       "      <td>0.448185</td>\n",
       "      <td>10.421668</td>\n",
       "      <td>55.318923</td>\n",
       "      <td>8.556514e-08</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>0.000326</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPP1</th>\n",
       "      <td>VCT_fusing.Tg</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>0.717596</td>\n",
       "      <td>6.877959</td>\n",
       "      <td>47.980641</td>\n",
       "      <td>2.897721e-07</td>\n",
       "      <td>0.002207</td>\n",
       "      <td>0.000736</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL3A1</th>\n",
       "      <td>VCT_fusing.Tg</td>\n",
       "      <td>COL3A1</td>\n",
       "      <td>0.814079</td>\n",
       "      <td>5.123309</td>\n",
       "      <td>43.569166</td>\n",
       "      <td>6.419772e-07</td>\n",
       "      <td>0.004890</td>\n",
       "      <td>0.001222</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DYNLRB1</th>\n",
       "      <td>VCT_fusing.Tg</td>\n",
       "      <td>DYNLRB1</td>\n",
       "      <td>0.418769</td>\n",
       "      <td>5.728626</td>\n",
       "      <td>39.369841</td>\n",
       "      <td>1.439906e-06</td>\n",
       "      <td>0.010968</td>\n",
       "      <td>0.002194</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               cluster     gene     logFC     logCPM          F        PValue  \\\n",
       "TIMP1    VCT_fusing.Tg    TIMP1  0.883189   6.423208  88.019460  1.126633e-09   \n",
       "IFI6     VCT_fusing.Tg     IFI6  0.448185  10.421668  55.318923  8.556514e-08   \n",
       "SPP1     VCT_fusing.Tg     SPP1  0.717596   6.877959  47.980641  2.897721e-07   \n",
       "COL3A1   VCT_fusing.Tg   COL3A1  0.814079   5.123309  43.569166  6.419772e-07   \n",
       "DYNLRB1  VCT_fusing.Tg  DYNLRB1  0.418769   5.728626  39.369841  1.439906e-06   \n",
       "\n",
       "         bonferroni  p_val_adj   cell_type  \n",
       "TIMP1      0.000009   0.000009  VCT_fusing  \n",
       "IFI6       0.000652   0.000326  VCT_fusing  \n",
       "SPP1       0.002207   0.000736  VCT_fusing  \n",
       "COL3A1     0.004890   0.001222  VCT_fusing  \n",
       "DYNLRB1    0.010968   0.002194  VCT_fusing  "
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTfDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "selected-recycling",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SPP1</th>\n",
       "      <td>VCT_fusing.Tg</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>0.717596</td>\n",
       "      <td>6.877959</td>\n",
       "      <td>47.980641</td>\n",
       "      <td>2.897721e-07</td>\n",
       "      <td>0.002207</td>\n",
       "      <td>0.000736</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            cluster  gene     logFC    logCPM          F        PValue  \\\n",
       "SPP1  VCT_fusing.Tg  SPP1  0.717596  6.877959  47.980641  2.897721e-07   \n",
       "\n",
       "      bonferroni  p_val_adj   cell_type  \n",
       "SPP1    0.002207   0.000736  VCT_fusing  "
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTfDEG_up.loc[['SPP1']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "convertible-beginning",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "soviet-inquiry",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "lined-pickup",
   "metadata": {},
   "source": [
    "### VCT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "romantic-renaissance",
   "metadata": {},
   "outputs": [],
   "source": [
    "VCTDEG = pd.read_csv(path+'/2_DEGs_SC_TOX_24h_VCT.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "numerical-entity",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CXCL8</th>\n",
       "      <td>0.974553</td>\n",
       "      <td>5.553529</td>\n",
       "      <td>118.972231</td>\n",
       "      <td>1.140086e-11</td>\n",
       "      <td>1.082967e-07</td>\n",
       "      <td>1.082967e-07</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Tg</td>\n",
       "      <td>CXCL8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TIMP1</th>\n",
       "      <td>0.808608</td>\n",
       "      <td>6.236301</td>\n",
       "      <td>80.224796</td>\n",
       "      <td>9.055197e-10</td>\n",
       "      <td>8.601531e-06</td>\n",
       "      <td>4.300766e-06</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Tg</td>\n",
       "      <td>TIMP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF17</th>\n",
       "      <td>1.837875</td>\n",
       "      <td>3.048054</td>\n",
       "      <td>60.373804</td>\n",
       "      <td>1.653236e-08</td>\n",
       "      <td>1.570409e-04</td>\n",
       "      <td>5.234696e-05</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Tg</td>\n",
       "      <td>RNF17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCL4L2</th>\n",
       "      <td>1.032796</td>\n",
       "      <td>2.540356</td>\n",
       "      <td>57.375081</td>\n",
       "      <td>2.714022e-08</td>\n",
       "      <td>2.578049e-04</td>\n",
       "      <td>6.445124e-05</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Tg</td>\n",
       "      <td>CCL4L2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IFI6</th>\n",
       "      <td>0.321170</td>\n",
       "      <td>10.736599</td>\n",
       "      <td>55.654252</td>\n",
       "      <td>3.636252e-08</td>\n",
       "      <td>3.454076e-04</td>\n",
       "      <td>6.908152e-05</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Tg</td>\n",
       "      <td>IFI6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           logFC     logCPM           F        PValue    bonferroni  \\\n",
       "CXCL8   0.974553   5.553529  118.972231  1.140086e-11  1.082967e-07   \n",
       "TIMP1   0.808608   6.236301   80.224796  9.055197e-10  8.601531e-06   \n",
       "RNF17   1.837875   3.048054   60.373804  1.653236e-08  1.570409e-04   \n",
       "CCL4L2  1.032796   2.540356   57.375081  2.714022e-08  2.578049e-04   \n",
       "IFI6    0.321170  10.736599   55.654252  3.636252e-08  3.454076e-04   \n",
       "\n",
       "           p_val_adj cell_type cluster    gene  \n",
       "CXCL8   1.082967e-07       VCT  VCT.Tg   CXCL8  \n",
       "TIMP1   4.300766e-06       VCT  VCT.Tg   TIMP1  \n",
       "RNF17   5.234696e-05       VCT  VCT.Tg   RNF17  \n",
       "CCL4L2  6.445124e-05       VCT  VCT.Tg  CCL4L2  \n",
       "IFI6    6.908152e-05       VCT  VCT.Tg    IFI6  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTDEG_up= recoverDEGs(VCTDEG, infection='Tg', celltype='VCT', direction='up')\n",
    "VCTDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "excessive-finding",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in VCTDEG_up.cluster])\n",
    "VCTDEG_up = VCTDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "right-hospital",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = VCTDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "VCTDEG_up= VCTDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(VCTDEG_up).to_csv('../data/toxoplasma_inf_VCT_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "flexible-montgomery",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CXCL8</th>\n",
       "      <td>VCT.Tg</td>\n",
       "      <td>CXCL8</td>\n",
       "      <td>0.974553</td>\n",
       "      <td>5.553529</td>\n",
       "      <td>118.972231</td>\n",
       "      <td>1.140086e-11</td>\n",
       "      <td>1.082967e-07</td>\n",
       "      <td>1.082967e-07</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TIMP1</th>\n",
       "      <td>VCT.Tg</td>\n",
       "      <td>TIMP1</td>\n",
       "      <td>0.808608</td>\n",
       "      <td>6.236301</td>\n",
       "      <td>80.224796</td>\n",
       "      <td>9.055197e-10</td>\n",
       "      <td>8.601531e-06</td>\n",
       "      <td>4.300766e-06</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF17</th>\n",
       "      <td>VCT.Tg</td>\n",
       "      <td>RNF17</td>\n",
       "      <td>1.837875</td>\n",
       "      <td>3.048054</td>\n",
       "      <td>60.373804</td>\n",
       "      <td>1.653236e-08</td>\n",
       "      <td>1.570409e-04</td>\n",
       "      <td>5.234696e-05</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCL4L2</th>\n",
       "      <td>VCT.Tg</td>\n",
       "      <td>CCL4L2</td>\n",
       "      <td>1.032796</td>\n",
       "      <td>2.540356</td>\n",
       "      <td>57.375081</td>\n",
       "      <td>2.714022e-08</td>\n",
       "      <td>2.578049e-04</td>\n",
       "      <td>6.445124e-05</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IFI6</th>\n",
       "      <td>VCT.Tg</td>\n",
       "      <td>IFI6</td>\n",
       "      <td>0.321170</td>\n",
       "      <td>10.736599</td>\n",
       "      <td>55.654252</td>\n",
       "      <td>3.636252e-08</td>\n",
       "      <td>3.454076e-04</td>\n",
       "      <td>6.908152e-05</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster    gene     logFC     logCPM           F        PValue  \\\n",
       "CXCL8   VCT.Tg   CXCL8  0.974553   5.553529  118.972231  1.140086e-11   \n",
       "TIMP1   VCT.Tg   TIMP1  0.808608   6.236301   80.224796  9.055197e-10   \n",
       "RNF17   VCT.Tg   RNF17  1.837875   3.048054   60.373804  1.653236e-08   \n",
       "CCL4L2  VCT.Tg  CCL4L2  1.032796   2.540356   57.375081  2.714022e-08   \n",
       "IFI6    VCT.Tg    IFI6  0.321170  10.736599   55.654252  3.636252e-08   \n",
       "\n",
       "          bonferroni     p_val_adj cell_type  \n",
       "CXCL8   1.082967e-07  1.082967e-07       VCT  \n",
       "TIMP1   8.601531e-06  4.300766e-06       VCT  \n",
       "RNF17   1.570409e-04  5.234696e-05       VCT  \n",
       "CCL4L2  2.578049e-04  6.445124e-05       VCT  \n",
       "IFI6    3.454076e-04  6.908152e-05       VCT  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "blank-union",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SPP1</th>\n",
       "      <td>VCT.Tg</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>0.399504</td>\n",
       "      <td>6.847701</td>\n",
       "      <td>15.368978</td>\n",
       "      <td>0.00051</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.024099</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     cluster  gene     logFC    logCPM          F   PValue  bonferroni  \\\n",
       "SPP1  VCT.Tg  SPP1  0.399504  6.847701  15.368978  0.00051         1.0   \n",
       "\n",
       "      p_val_adj cell_type  \n",
       "SPP1   0.024099       VCT  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTDEG_up.loc[['SPP1']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "pacific-packet",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "awful-exclusive",
   "metadata": {},
   "source": [
    "### (3) Run cellphoneDB"
   ]
  },
  {
   "cell_type": "raw",
   "id": "pointed-projection",
   "metadata": {},
   "source": [
    "source activate cpdb\n",
    "\n",
    "bash run_cellphoneDB.sh "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "complicated-fifth",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "turkish-carry",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
