{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "statutory-hepatitis",
   "metadata": {},
   "source": [
    "**Author:** Elias Rafael Ruiz-Morales\n",
    "\n",
    "**Institution:** Wellcome Sanger institute\n",
    "\n",
    "**July, 2023**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "necessary-suffering",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "remarkable-carry",
   "metadata": {},
   "source": [
    "## Generate cellphoneDB input files (1) _meta.tsv and (2) _degs.tsv"
   ]
  },
  {
   "cell_type": "raw",
   "id": "driven-international",
   "metadata": {},
   "source": [
    "reference: https://github.com/ventolab/CellphoneDB/blob/master/notebooks/0_prepare_your_data_from_anndata.ipynb\n",
    "reference: https://github.com/ventolab/HGDA/blob/main/human/scRNAseq/cellphoneDB/cellphoneDB1_granulosa_preparefiles.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "reported-bookmark",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/opt/conda/bin/python'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import os\n",
    "import sys\n",
    "sys.executable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "eight-tomato",
   "metadata": {},
   "outputs": [],
   "source": [
    "def recoverDEGs(DEGdf, infection, celltype, direction):\n",
    "\n",
    "    \n",
    "    ### Up reg\n",
    "    ###\n",
    "    if direction=='up':\n",
    "\n",
    "        #selecting only up regulated genes\n",
    "        tempUP = DEGdf[DEGdf['logFC'] > 0].copy()\n",
    "        \n",
    "        #making new columns\n",
    "        tempUP['cell_type'] = celltype\n",
    "        tempUP['cluster'] = celltype + '.' + infection\n",
    "        tempUP['gene']=list(tempUP.index.values)\n",
    "\n",
    "        # select significant less restrictive\n",
    "        tempUP =tempUP[ tempUP.p_val_adj < 0.05]\n",
    "        tempUP = tempUP[ tempUP.logFC > 0.1]\n",
    "\n",
    "        \n",
    "        return(tempUP)\n",
    "\n",
    "\n",
    "    ### Down reg\n",
    "    ###\n",
    "    if direction=='down':\n",
    "        #selecting only down regulated genes\n",
    "        tempUP = DEGdf[DEGdf['logFC'] < 0].copy()\n",
    "        \n",
    "        #making new columns\n",
    "        tempUP['cell_type'] = celltype\n",
    "        tempUP['cluster'] = celltype + '.' + infection\n",
    "        tempUP['gene']=list(tempUP.index.values)\n",
    "\n",
    "\n",
    "        # select significant less restrictive\n",
    "        tempUP =tempUP[ tempUP.p_val_adj < 0.05]\n",
    "        tempUP = tempUP[ tempUP.logFC < -0.1]\n",
    "        \n",
    "        return(tempDwn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deadly-madison",
   "metadata": {},
   "source": [
    "## Prepare INPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "lesser-being",
   "metadata": {},
   "source": [
    "### (1) Load annadata_ single-cell explants from listeria infection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "lonely-senate",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read('../../data_integration/results/scVI/listeria_singleCell_24h.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "graduate-garage",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>freqs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>categories</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EVT_1</th>\n",
       "      <td>106</td>\n",
       "      <td>0.003438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EVT_2</th>\n",
       "      <td>602</td>\n",
       "      <td>0.019523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Endo_f</th>\n",
       "      <td>5714</td>\n",
       "      <td>0.185309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>5017</td>\n",
       "      <td>0.162705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F_p</th>\n",
       "      <td>334</td>\n",
       "      <td>0.010832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F_sm</th>\n",
       "      <td>71</td>\n",
       "      <td>0.002303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HBC</th>\n",
       "      <td>7948</td>\n",
       "      <td>0.257759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HBC_p</th>\n",
       "      <td>18</td>\n",
       "      <td>0.000584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PAMM1</th>\n",
       "      <td>2055</td>\n",
       "      <td>0.066645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PV</th>\n",
       "      <td>3441</td>\n",
       "      <td>0.111594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT</th>\n",
       "      <td>1575</td>\n",
       "      <td>0.051078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_CCC</th>\n",
       "      <td>536</td>\n",
       "      <td>0.017383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_fusing</th>\n",
       "      <td>2682</td>\n",
       "      <td>0.086979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCT_p</th>\n",
       "      <td>403</td>\n",
       "      <td>0.013070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iEVT</th>\n",
       "      <td>333</td>\n",
       "      <td>0.010799</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            counts     freqs\n",
       "categories                  \n",
       "EVT_1          106  0.003438\n",
       "EVT_2          602  0.019523\n",
       "Endo_f        5714  0.185309\n",
       "F             5017  0.162705\n",
       "F_p            334  0.010832\n",
       "F_sm            71  0.002303\n",
       "HBC           7948  0.257759\n",
       "HBC_p           18  0.000584\n",
       "PAMM1         2055  0.066645\n",
       "PV            3441  0.111594\n",
       "VCT           1575  0.051078\n",
       "VCT_CCC        536  0.017383\n",
       "VCT_fusing    2682  0.086979\n",
       "VCT_p          403  0.013070\n",
       "iEVT           333  0.010799"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['cell_type'].values.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "balanced-transformation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>freqs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>categories</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lm_24h</th>\n",
       "      <td>17778</td>\n",
       "      <td>0.576553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UI_Lm_24h</th>\n",
       "      <td>13057</td>\n",
       "      <td>0.423447</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            counts     freqs\n",
       "categories                  \n",
       "Lm_24h       17778  0.576553\n",
       "UI_Lm_24h    13057  0.423447"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['stage_perInfection'].values.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "intermediate-congress",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 30835 × 36601\n",
       "    obs: 'sample', 'stage', 'hpi', 'infection', 'percent_mito', 'n_counts', 'sample_barcode', 'assignment_SoC', 'donor_id', 'scrublet_score', 'scrublet_cluster_score', 'zscore', 'bh_pval', 'bonf_pval', 'S_score', 'G2M_score', 'phase', 'n_genes_by_counts', 'total_counts', 'total_counts_hs', 'pct_counts_hs', 'total_counts_tg', 'pct_counts_tg', 'Tg_infected', 'n_genes', '_scvi_batch', '_scvi_labels', '_scvi_local_l_mean', '_scvi_local_l_var', 'leiden_scvi', 'celltype_predictions', 'probabilities', 'scrublet_doublet', 'cell_type_2022', 'cell_type', 'souporcell_MFgenotype', 'MFgenotype', 'cell_type_broad', 'umap_density_Tg_infected', 'stage_perInfection', 'celltype-Stage', 'Tg_intracellular', 'celltype-Intracellular', 'Dev_Stage'\n",
       "    var: 'gene_ids', 'feature_types', 'mean-0', 'std-0', 'mean-1', 'std-1', 'mean-2', 'std-2', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm'\n",
       "    uns: 'Dev_Stage_colors', 'MFgenotype_colors', '_scvi', 'cell_type_2022_colors', 'cell_type_broad_colors', 'cell_type_colors', 'celltype_predictions_colors', 'donor_id_colors', 'hpi_colors', 'hvg', 'infection_colors', 'leiden', 'neighbors', 'sample_colors', 'scrublet_doublet_colors', 'umap', 'umap_density_Tg_infected_params'\n",
       "    obsm: 'X_scVI_n_latent_20', 'X_scVI_n_latent_30', 'X_scVI_n_latent_40', 'X_umap', '_scvi_extra_categoricals'\n",
       "    layers: 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "french-rhythm",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "incident-polyester",
   "metadata": {},
   "source": [
    "### DO NOT Subset dataset, use full matrix counts from infected manifod!!\n",
    "This is because we want to predict the cell-cell interactions during infection using the DEG in infection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "unavailable-hundred",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "#Take only the INFECTED counts matrix\n",
    "cellphoneDB_adata = adata[adata.obs['stage_perInfection']=='Lm_24h'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "bound-norfolk",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Take only the UNINFECTED counts matrix\n",
    "#cellphoneDB_adataUI = adata[adata.obs['stage_perInfection']=='UI_Lm_24h'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lasting-neighbor",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "bored-participant",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Pla_HDBR12330713_AAACCCAAGTAGAGTT       HBC\n",
       "Pla_HDBR12330713_AAACCCACAATTCTCT         F\n",
       "Pla_HDBR12330713_AAACCCACAGGCTACC       HBC\n",
       "Pla_HDBR12330713_AAACCCACATCGTCCT         F\n",
       "Pla_HDBR12330713_AAACCCAGTACCGCGT       HBC\n",
       "                                      ...  \n",
       "Pla_HDBR13661572_TTTGTTGAGGTTCATC       HBC\n",
       "Pla_HDBR13661572_TTTGTTGCAAATTAGG    Endo_f\n",
       "Pla_HDBR13661572_TTTGTTGCAATAGGGC         F\n",
       "Pla_HDBR13661572_TTTGTTGGTGCTTATG    Endo_f\n",
       "Pla_HDBR13661572_TTTGTTGGTGGTAACG     PAMM1\n",
       "Name: cell_type, Length: 17778, dtype: category\n",
       "Categories (15, object): ['EVT_1', 'EVT_2', 'Endo_f', 'F', ..., 'VCT_CCC', 'VCT_fusing', 'VCT_p', 'iEVT']"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['cell_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "beautiful-request",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>freqs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>categories</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Hrv135</th>\n",
       "      <td>3964</td>\n",
       "      <td>0.222972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hrv136</th>\n",
       "      <td>2956</td>\n",
       "      <td>0.166273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hrv236</th>\n",
       "      <td>10858</td>\n",
       "      <td>0.610755</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            counts     freqs\n",
       "categories                  \n",
       "Hrv135        3964  0.222972\n",
       "Hrv136        2956  0.166273\n",
       "Hrv236       10858  0.610755"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['donor_id'].values.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "angry-columbus",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 17778 × 36601\n",
       "    obs: 'sample', 'stage', 'hpi', 'infection', 'percent_mito', 'n_counts', 'sample_barcode', 'assignment_SoC', 'donor_id', 'scrublet_score', 'scrublet_cluster_score', 'zscore', 'bh_pval', 'bonf_pval', 'S_score', 'G2M_score', 'phase', 'n_genes_by_counts', 'total_counts', 'total_counts_hs', 'pct_counts_hs', 'total_counts_tg', 'pct_counts_tg', 'Tg_infected', 'n_genes', '_scvi_batch', '_scvi_labels', '_scvi_local_l_mean', '_scvi_local_l_var', 'leiden_scvi', 'celltype_predictions', 'probabilities', 'scrublet_doublet', 'cell_type_2022', 'cell_type', 'souporcell_MFgenotype', 'MFgenotype', 'cell_type_broad', 'umap_density_Tg_infected', 'stage_perInfection', 'celltype-Stage', 'Tg_intracellular', 'celltype-Intracellular', 'Dev_Stage'\n",
       "    var: 'gene_ids', 'feature_types', 'mean-0', 'std-0', 'mean-1', 'std-1', 'mean-2', 'std-2', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm'\n",
       "    uns: 'Dev_Stage_colors', 'MFgenotype_colors', '_scvi', 'cell_type_2022_colors', 'cell_type_broad_colors', 'cell_type_colors', 'celltype_predictions_colors', 'donor_id_colors', 'hpi_colors', 'hvg', 'infection_colors', 'leiden', 'neighbors', 'sample_colors', 'scrublet_doublet_colors', 'umap', 'umap_density_Tg_infected_params'\n",
       "    obsm: 'X_scVI_n_latent_20', 'X_scVI_n_latent_30', 'X_scVI_n_latent_40', 'X_umap', '_scvi_extra_categoricals'\n",
       "    layers: 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "impressed-solid",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>freqs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>categories</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lm_24h</th>\n",
       "      <td>17778</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            counts  freqs\n",
       "categories               \n",
       "Lm_24h       17778    1.0"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['stage_perInfection'].values.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "improving-horizon",
   "metadata": {},
   "source": [
    "### Load *new* cell clusters annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "packed-warner",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add cluster subname such that this new name will match with DEG cluster name\n",
    "\n",
    "#INFECTED\n",
    "cellphoneDB_adata.obs['cell_type']= cellphoneDB_adata.obs.cell_type.astype('string')+'.'+cellphoneDB_adata.obs.infection.astype('string')\n",
    "\n",
    "#UI\n",
    "#cellphoneDB_adataUI.obs['cell_type']= cellphoneDB_adataUI.obs.cell_type.astype('string')+'.'+cellphoneDB_adataUI.obs.stage_perInfection.astype('string')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "incoming-charge",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Pla_HDBR12330713_AAACCCAAGTAGAGTT       HBC.Lm\n",
       "Pla_HDBR12330713_AAACCCACAATTCTCT         F.Lm\n",
       "Pla_HDBR12330713_AAACCCACAGGCTACC       HBC.Lm\n",
       "Pla_HDBR12330713_AAACCCACATCGTCCT         F.Lm\n",
       "Pla_HDBR12330713_AAACCCAGTACCGCGT       HBC.Lm\n",
       "                                       ...    \n",
       "Pla_HDBR13661572_TTTGTTGAGGTTCATC       HBC.Lm\n",
       "Pla_HDBR13661572_TTTGTTGCAAATTAGG    Endo_f.Lm\n",
       "Pla_HDBR13661572_TTTGTTGCAATAGGGC         F.Lm\n",
       "Pla_HDBR13661572_TTTGTTGGTGCTTATG    Endo_f.Lm\n",
       "Pla_HDBR13661572_TTTGTTGGTGGTAACG     PAMM1.Lm\n",
       "Name: cell_type, Length: 17778, dtype: string"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata.obs['cell_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "stopped-aaron",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "sized-worse",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 17778 × 36601\n",
       "    obs: 'sample', 'stage', 'hpi', 'infection', 'percent_mito', 'n_counts', 'sample_barcode', 'assignment_SoC', 'donor_id', 'scrublet_score', 'scrublet_cluster_score', 'zscore', 'bh_pval', 'bonf_pval', 'S_score', 'G2M_score', 'phase', 'n_genes_by_counts', 'total_counts', 'total_counts_hs', 'pct_counts_hs', 'total_counts_tg', 'pct_counts_tg', 'Tg_infected', 'n_genes', '_scvi_batch', '_scvi_labels', '_scvi_local_l_mean', '_scvi_local_l_var', 'leiden_scvi', 'celltype_predictions', 'probabilities', 'scrublet_doublet', 'cell_type_2022', 'cell_type', 'souporcell_MFgenotype', 'MFgenotype', 'cell_type_broad', 'umap_density_Tg_infected', 'stage_perInfection', 'celltype-Stage', 'Tg_intracellular', 'celltype-Intracellular', 'Dev_Stage'\n",
       "    var: 'gene_ids', 'feature_types', 'mean-0', 'std-0', 'mean-1', 'std-1', 'mean-2', 'std-2', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm'\n",
       "    uns: 'Dev_Stage_colors', 'MFgenotype_colors', '_scvi', 'cell_type_2022_colors', 'cell_type_broad_colors', 'cell_type_colors', 'celltype_predictions_colors', 'donor_id_colors', 'hpi_colors', 'hvg', 'infection_colors', 'leiden', 'neighbors', 'sample_colors', 'scrublet_doublet_colors', 'umap', 'umap_density_Tg_infected_params'\n",
       "    obsm: 'X_scVI_n_latent_20', 'X_scVI_n_latent_30', 'X_scVI_n_latent_40', 'X_umap', '_scvi_extra_categoricals'\n",
       "    layers: 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellphoneDB_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "industrial-hammer",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "successful-preserve",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "sc.pp.normalize_per_cell(cellphoneDB_adata)\n",
    "sc.pp.log1p(cellphoneDB_adata)                                                                                                            \n",
    "# sc.pp.scale(adata, max_value=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "impressed-phrase",
   "metadata": {},
   "source": [
    "### The anndata object contains counts that have been normalized (per cell) and log-transformed, therefore save this as below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "infinite-packaging",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'cell_type' as categorical\n"
     ]
    }
   ],
   "source": [
    "# counts input files \n",
    "cellphoneDB_adata.write('../data/listeria_inf_normlogTransformed_24h.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "psychological-mountain",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cellphoneDB_adata=sc.read('../data/listeria_inf_normlogTransformed.h5ad')\n",
    "#df_meta= pd.read_csv('../data/listeria_inf_meta.tsv', sep = '\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "specialized-eclipse",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "alone-diversity",
   "metadata": {},
   "outputs": [],
   "source": [
    "# meta file\n",
    "df_meta = pd.DataFrame(data={'Cell':list(cellphoneDB_adata.obs.index),\n",
    "                             'cell_type':[ str(i) for i in cellphoneDB_adata.obs['cell_type']] })\n",
    "df_meta.set_index('Cell', inplace=True)\n",
    "df_meta.to_csv('../data/listeria_inf_meta_24h.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "bridal-neutral",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR12330713_AAACCCAAGTAGAGTT</th>\n",
       "      <td>HBC.Lm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR12330713_AAACCCACAATTCTCT</th>\n",
       "      <td>F.Lm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR12330713_AAACCCACAGGCTACC</th>\n",
       "      <td>HBC.Lm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR12330713_AAACCCACATCGTCCT</th>\n",
       "      <td>F.Lm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR12330713_AAACCCAGTACCGCGT</th>\n",
       "      <td>HBC.Lm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13661572_TTTGTTGAGGTTCATC</th>\n",
       "      <td>HBC.Lm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13661572_TTTGTTGCAAATTAGG</th>\n",
       "      <td>Endo_f.Lm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13661572_TTTGTTGCAATAGGGC</th>\n",
       "      <td>F.Lm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13661572_TTTGTTGGTGCTTATG</th>\n",
       "      <td>Endo_f.Lm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pla_HDBR13661572_TTTGTTGGTGGTAACG</th>\n",
       "      <td>PAMM1.Lm</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17778 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   cell_type\n",
       "Cell                                        \n",
       "Pla_HDBR12330713_AAACCCAAGTAGAGTT     HBC.Lm\n",
       "Pla_HDBR12330713_AAACCCACAATTCTCT       F.Lm\n",
       "Pla_HDBR12330713_AAACCCACAGGCTACC     HBC.Lm\n",
       "Pla_HDBR12330713_AAACCCACATCGTCCT       F.Lm\n",
       "Pla_HDBR12330713_AAACCCAGTACCGCGT     HBC.Lm\n",
       "...                                      ...\n",
       "Pla_HDBR13661572_TTTGTTGAGGTTCATC     HBC.Lm\n",
       "Pla_HDBR13661572_TTTGTTGCAAATTAGG  Endo_f.Lm\n",
       "Pla_HDBR13661572_TTTGTTGCAATAGGGC       F.Lm\n",
       "Pla_HDBR13661572_TTTGTTGGTGCTTATG  Endo_f.Lm\n",
       "Pla_HDBR13661572_TTTGTTGGTGGTAACG   PAMM1.Lm\n",
       "\n",
       "[17778 rows x 1 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "familiar-state",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'EVT_1.Lm',\n",
       " 'EVT_2.Lm',\n",
       " 'Endo_f.Lm',\n",
       " 'F.Lm',\n",
       " 'F_p.Lm',\n",
       " 'F_sm.Lm',\n",
       " 'HBC.Lm',\n",
       " 'HBC_p.Lm',\n",
       " 'PAMM1.Lm',\n",
       " 'PV.Lm',\n",
       " 'VCT.Lm',\n",
       " 'VCT_CCC.Lm',\n",
       " 'VCT_fusing.Lm',\n",
       " 'VCT_p.Lm',\n",
       " 'iEVT.Lm'}"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(list(df_meta['cell_type']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "figured-lemon",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "compatible-incidence",
   "metadata": {},
   "source": [
    "### (2) Do DEGs individually from precalculated DEGs files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "formed-energy",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PATH to immune cells DGE tables\n",
    "path='../../diffGeneExpression/macrophages/results/SC_limma_Immune_LIS_24h/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sound-great",
   "metadata": {},
   "source": [
    "### HBC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "under-belle",
   "metadata": {},
   "outputs": [],
   "source": [
    "HBDEG = pd.read_csv(path+'/1_DEGs_SC_LIS_24h_HBC.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "sorted-parliament",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>STC1</th>\n",
       "      <td>1.288401</td>\n",
       "      <td>6.197037</td>\n",
       "      <td>242.503878</td>\n",
       "      <td>3.226768e-13</td>\n",
       "      <td>2.358122e-09</td>\n",
       "      <td>2.358122e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL1RN</th>\n",
       "      <td>1.022193</td>\n",
       "      <td>7.355690</td>\n",
       "      <td>222.829592</td>\n",
       "      <td>7.465613e-13</td>\n",
       "      <td>5.455870e-09</td>\n",
       "      <td>2.727935e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IGFBP4</th>\n",
       "      <td>-0.627356</td>\n",
       "      <td>6.728119</td>\n",
       "      <td>177.960289</td>\n",
       "      <td>6.730971e-12</td>\n",
       "      <td>4.918994e-08</td>\n",
       "      <td>1.639665e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP13A3</th>\n",
       "      <td>0.468337</td>\n",
       "      <td>7.538326</td>\n",
       "      <td>159.872988</td>\n",
       "      <td>1.887261e-11</td>\n",
       "      <td>1.379211e-07</td>\n",
       "      <td>3.448026e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCGRT</th>\n",
       "      <td>-0.369558</td>\n",
       "      <td>9.378415</td>\n",
       "      <td>151.271960</td>\n",
       "      <td>3.197164e-11</td>\n",
       "      <td>2.336487e-07</td>\n",
       "      <td>4.672975e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STK17B</th>\n",
       "      <td>-0.000118</td>\n",
       "      <td>6.717451</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>9.982881e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.986069e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FBXW4</th>\n",
       "      <td>0.000130</td>\n",
       "      <td>5.069345</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>9.982220e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.986069e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THUMPD3</th>\n",
       "      <td>-0.000173</td>\n",
       "      <td>4.767530</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>9.983336e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.986069e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATXN1</th>\n",
       "      <td>0.000060</td>\n",
       "      <td>8.172113</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>9.986977e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.987206e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSEN34</th>\n",
       "      <td>0.000099</td>\n",
       "      <td>5.062587</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>9.987206e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.987206e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7308 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            logFC    logCPM           F        PValue    bonferroni  \\\n",
       "STC1     1.288401  6.197037  242.503878  3.226768e-13  2.358122e-09   \n",
       "IL1RN    1.022193  7.355690  222.829592  7.465613e-13  5.455870e-09   \n",
       "IGFBP4  -0.627356  6.728119  177.960289  6.730971e-12  4.918994e-08   \n",
       "ATP13A3  0.468337  7.538326  159.872988  1.887261e-11  1.379211e-07   \n",
       "FCGRT   -0.369558  9.378415  151.271960  3.197164e-11  2.336487e-07   \n",
       "...           ...       ...         ...           ...           ...   \n",
       "STK17B  -0.000118  6.717451    0.000005  9.982881e-01  1.000000e+00   \n",
       "FBXW4    0.000130  5.069345    0.000005  9.982220e-01  1.000000e+00   \n",
       "THUMPD3 -0.000173  4.767530    0.000004  9.983336e-01  1.000000e+00   \n",
       "ATXN1    0.000060  8.172113    0.000003  9.986977e-01  1.000000e+00   \n",
       "TSEN34   0.000099  5.062587    0.000003  9.987206e-01  1.000000e+00   \n",
       "\n",
       "            p_val_adj  \n",
       "STC1     2.358122e-09  \n",
       "IL1RN    2.727935e-09  \n",
       "IGFBP4   1.639665e-08  \n",
       "ATP13A3  3.448026e-08  \n",
       "FCGRT    4.672975e-08  \n",
       "...               ...  \n",
       "STK17B   9.986069e-01  \n",
       "FBXW4    9.986069e-01  \n",
       "THUMPD3  9.986069e-01  \n",
       "ATXN1    9.987206e-01  \n",
       "TSEN34   9.987206e-01  \n",
       "\n",
       "[7308 rows x 6 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HBDEG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "exposed-substance",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>STC1</th>\n",
       "      <td>1.288401</td>\n",
       "      <td>6.197037</td>\n",
       "      <td>242.503878</td>\n",
       "      <td>3.226768e-13</td>\n",
       "      <td>2.358122e-09</td>\n",
       "      <td>2.358122e-09</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Lm</td>\n",
       "      <td>STC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL1RN</th>\n",
       "      <td>1.022193</td>\n",
       "      <td>7.355690</td>\n",
       "      <td>222.829592</td>\n",
       "      <td>7.465613e-13</td>\n",
       "      <td>5.455870e-09</td>\n",
       "      <td>2.727935e-09</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Lm</td>\n",
       "      <td>IL1RN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP13A3</th>\n",
       "      <td>0.468337</td>\n",
       "      <td>7.538326</td>\n",
       "      <td>159.872988</td>\n",
       "      <td>1.887261e-11</td>\n",
       "      <td>1.379211e-07</td>\n",
       "      <td>3.448026e-08</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Lm</td>\n",
       "      <td>ATP13A3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMSN1</th>\n",
       "      <td>0.451065</td>\n",
       "      <td>8.824758</td>\n",
       "      <td>130.151077</td>\n",
       "      <td>1.316877e-10</td>\n",
       "      <td>9.623734e-07</td>\n",
       "      <td>1.374819e-07</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Lm</td>\n",
       "      <td>SAMSN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL360227.1</th>\n",
       "      <td>0.779019</td>\n",
       "      <td>4.882505</td>\n",
       "      <td>126.551758</td>\n",
       "      <td>1.709469e-10</td>\n",
       "      <td>1.249280e-06</td>\n",
       "      <td>1.388089e-07</td>\n",
       "      <td>HBC</td>\n",
       "      <td>HBC.Lm</td>\n",
       "      <td>AL360227.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               logFC    logCPM           F        PValue    bonferroni  \\\n",
       "STC1        1.288401  6.197037  242.503878  3.226768e-13  2.358122e-09   \n",
       "IL1RN       1.022193  7.355690  222.829592  7.465613e-13  5.455870e-09   \n",
       "ATP13A3     0.468337  7.538326  159.872988  1.887261e-11  1.379211e-07   \n",
       "SAMSN1      0.451065  8.824758  130.151077  1.316877e-10  9.623734e-07   \n",
       "AL360227.1  0.779019  4.882505  126.551758  1.709469e-10  1.249280e-06   \n",
       "\n",
       "               p_val_adj cell_type cluster        gene  \n",
       "STC1        2.358122e-09       HBC  HBC.Lm        STC1  \n",
       "IL1RN       2.727935e-09       HBC  HBC.Lm       IL1RN  \n",
       "ATP13A3     3.448026e-08       HBC  HBC.Lm     ATP13A3  \n",
       "SAMSN1      1.374819e-07       HBC  HBC.Lm      SAMSN1  \n",
       "AL360227.1  1.388089e-07       HBC  HBC.Lm  AL360227.1  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HBDEG_up= recoverDEGs(HBDEG, infection='Lm', celltype='HBC', direction='up')\n",
    "HBDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "excited-impossible",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in HBDEG_up.cluster])\n",
    "HBDEG_up = HBDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "drawn-bikini",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = HBDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "HBDEG_up= HBDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(HBDEG_up).to_csv('../data/listeria_inf_HBC_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "regular-marble",
   "metadata": {},
   "outputs": [],
   "source": [
    "#list(HBDEG_up.gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "rough-wright",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VEGFA</th>\n",
       "      <td>HBC.Lm</td>\n",
       "      <td>VEGFA</td>\n",
       "      <td>0.77482</td>\n",
       "      <td>6.253933</td>\n",
       "      <td>90.455904</td>\n",
       "      <td>3.572045e-09</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>7.458431e-07</td>\n",
       "      <td>HBC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cluster   gene    logFC    logCPM          F        PValue  bonferroni  \\\n",
       "VEGFA  HBC.Lm  VEGFA  0.77482  6.253933  90.455904  3.572045e-09    0.000026   \n",
       "\n",
       "          p_val_adj cell_type  \n",
       "VEGFA  7.458431e-07       HBC  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#HBDEG_up=pd.read_table('../data/listeria_inf_HBC_DEGsup.tsv')\n",
    "#HBDEG_up.loc[['VEGFA']]\n",
    "HBDEG_up.loc[HBDEG_up['gene'] == 'VEGFA']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "opposite-montana",
   "metadata": {},
   "source": [
    "### PAMM1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "dirty-politics",
   "metadata": {},
   "outputs": [],
   "source": [
    "PAMM1DEG = pd.read_csv(path+'/2_DEGs_SC_LIS_24h_PAMM1.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "organic-depression",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C15orf48</th>\n",
       "      <td>1.144871</td>\n",
       "      <td>9.545963</td>\n",
       "      <td>88.965254</td>\n",
       "      <td>7.574555e-10</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Lm</td>\n",
       "      <td>C15orf48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STC1</th>\n",
       "      <td>1.372473</td>\n",
       "      <td>5.604870</td>\n",
       "      <td>80.251079</td>\n",
       "      <td>2.121684e-09</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Lm</td>\n",
       "      <td>STC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf56</th>\n",
       "      <td>1.185322</td>\n",
       "      <td>5.357605</td>\n",
       "      <td>62.887467</td>\n",
       "      <td>2.195464e-08</td>\n",
       "      <td>0.000134</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Lm</td>\n",
       "      <td>C1orf56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TIMP1</th>\n",
       "      <td>0.744604</td>\n",
       "      <td>10.769514</td>\n",
       "      <td>57.788443</td>\n",
       "      <td>4.768015e-08</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Lm</td>\n",
       "      <td>TIMP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MARCKS</th>\n",
       "      <td>0.597890</td>\n",
       "      <td>9.892981</td>\n",
       "      <td>58.744905</td>\n",
       "      <td>4.107608e-08</td>\n",
       "      <td>0.000250</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>PAMM1</td>\n",
       "      <td>PAMM1.Lm</td>\n",
       "      <td>MARCKS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             logFC     logCPM          F        PValue  bonferroni  p_val_adj  \\\n",
       "C15orf48  1.144871   9.545963  88.965254  7.574555e-10    0.000005   0.000005   \n",
       "STC1      1.372473   5.604870  80.251079  2.121684e-09    0.000013   0.000006   \n",
       "C1orf56   1.185322   5.357605  62.887467  2.195464e-08    0.000134   0.000045   \n",
       "TIMP1     0.744604  10.769514  57.788443  4.768015e-08    0.000290   0.000058   \n",
       "MARCKS    0.597890   9.892981  58.744905  4.107608e-08    0.000250   0.000058   \n",
       "\n",
       "         cell_type   cluster      gene  \n",
       "C15orf48     PAMM1  PAMM1.Lm  C15orf48  \n",
       "STC1         PAMM1  PAMM1.Lm      STC1  \n",
       "C1orf56      PAMM1  PAMM1.Lm   C1orf56  \n",
       "TIMP1        PAMM1  PAMM1.Lm     TIMP1  \n",
       "MARCKS       PAMM1  PAMM1.Lm    MARCKS  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PAMM1DEG_up= recoverDEGs(PAMM1DEG, infection='Lm', celltype='PAMM1', direction='up')\n",
    "PAMM1DEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "abstract-highlight",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in PAMM1DEG_up.cluster])\n",
    "PAMM1DEG_up = PAMM1DEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "sorted-andrew",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = PAMM1DEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "PAMM1DEG_up= PAMM1DEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(PAMM1DEG_up).to_csv('../data/listeria_inf_PAMM1_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "employed-aside",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C15orf48</th>\n",
       "      <td>PAMM1.Lm</td>\n",
       "      <td>C15orf48</td>\n",
       "      <td>1.144871</td>\n",
       "      <td>9.545963</td>\n",
       "      <td>88.965254</td>\n",
       "      <td>7.574555e-10</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STC1</th>\n",
       "      <td>PAMM1.Lm</td>\n",
       "      <td>STC1</td>\n",
       "      <td>1.372473</td>\n",
       "      <td>5.604870</td>\n",
       "      <td>80.251079</td>\n",
       "      <td>2.121684e-09</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf56</th>\n",
       "      <td>PAMM1.Lm</td>\n",
       "      <td>C1orf56</td>\n",
       "      <td>1.185322</td>\n",
       "      <td>5.357605</td>\n",
       "      <td>62.887467</td>\n",
       "      <td>2.195464e-08</td>\n",
       "      <td>0.000134</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TIMP1</th>\n",
       "      <td>PAMM1.Lm</td>\n",
       "      <td>TIMP1</td>\n",
       "      <td>0.744604</td>\n",
       "      <td>10.769514</td>\n",
       "      <td>57.788443</td>\n",
       "      <td>4.768015e-08</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MARCKS</th>\n",
       "      <td>PAMM1.Lm</td>\n",
       "      <td>MARCKS</td>\n",
       "      <td>0.597890</td>\n",
       "      <td>9.892981</td>\n",
       "      <td>58.744905</td>\n",
       "      <td>4.107608e-08</td>\n",
       "      <td>0.000250</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>PAMM1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           cluster      gene     logFC     logCPM          F        PValue  \\\n",
       "C15orf48  PAMM1.Lm  C15orf48  1.144871   9.545963  88.965254  7.574555e-10   \n",
       "STC1      PAMM1.Lm      STC1  1.372473   5.604870  80.251079  2.121684e-09   \n",
       "C1orf56   PAMM1.Lm   C1orf56  1.185322   5.357605  62.887467  2.195464e-08   \n",
       "TIMP1     PAMM1.Lm     TIMP1  0.744604  10.769514  57.788443  4.768015e-08   \n",
       "MARCKS    PAMM1.Lm    MARCKS  0.597890   9.892981  58.744905  4.107608e-08   \n",
       "\n",
       "          bonferroni  p_val_adj cell_type  \n",
       "C15orf48    0.000005   0.000005     PAMM1  \n",
       "STC1        0.000013   0.000006     PAMM1  \n",
       "C1orf56     0.000134   0.000045     PAMM1  \n",
       "TIMP1       0.000290   0.000058     PAMM1  \n",
       "MARCKS      0.000250   0.000058     PAMM1  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PAMM1DEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "inside-favor",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "divine-excerpt",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "iraqi-spouse",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "julian-jaguar",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "significant-program",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PATH to FibroEndo cells DGE tables\n",
    "path='../../diffGeneExpression/fibroEndo/results/SC_limma_fibroEndo_LIS_24h/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "wireless-pricing",
   "metadata": {},
   "source": [
    "### F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "respective-understanding",
   "metadata": {},
   "outputs": [],
   "source": [
    "FDEG = pd.read_csv(path+'/1_DEGs_SC_LIS_24h_F.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "quality-sullivan",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SLC30A4</th>\n",
       "      <td>0.773521</td>\n",
       "      <td>5.872970</td>\n",
       "      <td>121.331122</td>\n",
       "      <td>5.458952e-10</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Lm</td>\n",
       "      <td>SLC30A4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP10B</th>\n",
       "      <td>1.159572</td>\n",
       "      <td>4.096377</td>\n",
       "      <td>112.813973</td>\n",
       "      <td>1.026411e-09</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Lm</td>\n",
       "      <td>ATP10B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WDR45B</th>\n",
       "      <td>0.466109</td>\n",
       "      <td>6.572808</td>\n",
       "      <td>99.179367</td>\n",
       "      <td>3.086850e-09</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Lm</td>\n",
       "      <td>WDR45B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MARCKS</th>\n",
       "      <td>0.681913</td>\n",
       "      <td>9.354571</td>\n",
       "      <td>99.287132</td>\n",
       "      <td>3.058607e-09</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Lm</td>\n",
       "      <td>MARCKS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC02326</th>\n",
       "      <td>0.698494</td>\n",
       "      <td>6.232710</td>\n",
       "      <td>96.452308</td>\n",
       "      <td>3.906280e-09</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>F</td>\n",
       "      <td>F.Lm</td>\n",
       "      <td>LINC02326</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              logFC    logCPM           F        PValue  bonferroni  \\\n",
       "SLC30A4    0.773521  5.872970  121.331122  5.458952e-10    0.000004   \n",
       "ATP10B     1.159572  4.096377  112.813973  1.026411e-09    0.000008   \n",
       "WDR45B     0.466109  6.572808   99.179367  3.086850e-09    0.000023   \n",
       "MARCKS     0.681913  9.354571   99.287132  3.058607e-09    0.000023   \n",
       "LINC02326  0.698494  6.232710   96.452308  3.906280e-09    0.000030   \n",
       "\n",
       "           p_val_adj cell_type cluster       gene  \n",
       "SLC30A4     0.000001         F    F.Lm    SLC30A4  \n",
       "ATP10B      0.000002         F    F.Lm     ATP10B  \n",
       "WDR45B      0.000002         F    F.Lm     WDR45B  \n",
       "MARCKS      0.000002         F    F.Lm     MARCKS  \n",
       "LINC02326   0.000002         F    F.Lm  LINC02326  "
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FDEG_up= recoverDEGs(FDEG, infection='Lm', celltype='F', direction='up')\n",
    "FDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "indie-technique",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in FDEG_up.cluster])\n",
    "FDEG_up = FDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "entertaining-dynamics",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = FDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "FDEG_up= FDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(FDEG_up).to_csv('../data/listeria_inf_F_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "opened-superintendent",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SLC30A4</th>\n",
       "      <td>F.Lm</td>\n",
       "      <td>SLC30A4</td>\n",
       "      <td>0.773521</td>\n",
       "      <td>5.872970</td>\n",
       "      <td>121.331122</td>\n",
       "      <td>5.458952e-10</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP10B</th>\n",
       "      <td>F.Lm</td>\n",
       "      <td>ATP10B</td>\n",
       "      <td>1.159572</td>\n",
       "      <td>4.096377</td>\n",
       "      <td>112.813973</td>\n",
       "      <td>1.026411e-09</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WDR45B</th>\n",
       "      <td>F.Lm</td>\n",
       "      <td>WDR45B</td>\n",
       "      <td>0.466109</td>\n",
       "      <td>6.572808</td>\n",
       "      <td>99.179367</td>\n",
       "      <td>3.086850e-09</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MARCKS</th>\n",
       "      <td>F.Lm</td>\n",
       "      <td>MARCKS</td>\n",
       "      <td>0.681913</td>\n",
       "      <td>9.354571</td>\n",
       "      <td>99.287132</td>\n",
       "      <td>3.058607e-09</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC02326</th>\n",
       "      <td>F.Lm</td>\n",
       "      <td>LINC02326</td>\n",
       "      <td>0.698494</td>\n",
       "      <td>6.232710</td>\n",
       "      <td>96.452308</td>\n",
       "      <td>3.906280e-09</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          cluster       gene     logFC    logCPM           F        PValue  \\\n",
       "SLC30A4      F.Lm    SLC30A4  0.773521  5.872970  121.331122  5.458952e-10   \n",
       "ATP10B       F.Lm     ATP10B  1.159572  4.096377  112.813973  1.026411e-09   \n",
       "WDR45B       F.Lm     WDR45B  0.466109  6.572808   99.179367  3.086850e-09   \n",
       "MARCKS       F.Lm     MARCKS  0.681913  9.354571   99.287132  3.058607e-09   \n",
       "LINC02326    F.Lm  LINC02326  0.698494  6.232710   96.452308  3.906280e-09   \n",
       "\n",
       "           bonferroni  p_val_adj cell_type  \n",
       "SLC30A4      0.000004   0.000001         F  \n",
       "ATP10B       0.000008   0.000002         F  \n",
       "WDR45B       0.000023   0.000002         F  \n",
       "MARCKS       0.000023   0.000002         F  \n",
       "LINC02326    0.000030   0.000002         F  "
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "hawaiian-paste",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>THBS2</th>\n",
       "      <td>F.Lm</td>\n",
       "      <td>THBS2</td>\n",
       "      <td>0.783961</td>\n",
       "      <td>5.117159</td>\n",
       "      <td>21.837357</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001318</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cluster   gene     logFC    logCPM          F    PValue  bonferroni  \\\n",
       "THBS2    F.Lm  THBS2  0.783961  5.117159  21.837357  0.000143         1.0   \n",
       "\n",
       "       p_val_adj cell_type  \n",
       "THBS2   0.001318         F  "
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FDEG_up.loc[['THBS2']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "satisfied-nebraska",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "spectacular-input",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "existing-negative",
   "metadata": {},
   "source": [
    "### PV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "fifth-trunk",
   "metadata": {},
   "outputs": [],
   "source": [
    "PVDEG = pd.read_csv(path+'/2_DEGs_SC_LIS_24h_PV.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "published-eligibility",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IL1B</th>\n",
       "      <td>1.441441</td>\n",
       "      <td>5.748745</td>\n",
       "      <td>83.425492</td>\n",
       "      <td>2.861042e-09</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>IL1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL133453.1</th>\n",
       "      <td>0.628851</td>\n",
       "      <td>6.338373</td>\n",
       "      <td>70.006608</td>\n",
       "      <td>1.441496e-08</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>AL133453.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC02326</th>\n",
       "      <td>0.805504</td>\n",
       "      <td>6.203114</td>\n",
       "      <td>71.988762</td>\n",
       "      <td>1.119157e-08</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>LINC02326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STC1</th>\n",
       "      <td>1.174349</td>\n",
       "      <td>9.604055</td>\n",
       "      <td>65.198570</td>\n",
       "      <td>2.726236e-08</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>STC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC009554.1</th>\n",
       "      <td>0.919839</td>\n",
       "      <td>5.432419</td>\n",
       "      <td>66.342901</td>\n",
       "      <td>2.335296e-08</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>AC009554.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               logFC    logCPM          F        PValue  bonferroni  \\\n",
       "IL1B        1.441441  5.748745  83.425492  2.861042e-09    0.000021   \n",
       "AL133453.1  0.628851  6.338373  70.006608  1.441496e-08    0.000108   \n",
       "LINC02326   0.805504  6.203114  71.988762  1.119157e-08    0.000084   \n",
       "STC1        1.174349  9.604055  65.198570  2.726236e-08    0.000205   \n",
       "AC009554.1  0.919839  5.432419  66.342901  2.335296e-08    0.000175   \n",
       "\n",
       "            p_val_adj cell_type cluster        gene  \n",
       "IL1B         0.000021        PV   PV.Lm        IL1B  \n",
       "AL133453.1   0.000036        PV   PV.Lm  AL133453.1  \n",
       "LINC02326    0.000036        PV   PV.Lm   LINC02326  \n",
       "STC1         0.000041        PV   PV.Lm        STC1  \n",
       "AC009554.1   0.000041        PV   PV.Lm  AC009554.1  "
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PVDEG_up= recoverDEGs(PVDEG, infection='Lm', celltype='PV', direction='up')\n",
    "PVDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "micro-matrix",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IL1B</th>\n",
       "      <td>1.441441</td>\n",
       "      <td>5.748745</td>\n",
       "      <td>83.425492</td>\n",
       "      <td>2.861042e-09</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>IL1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL133453.1</th>\n",
       "      <td>0.628851</td>\n",
       "      <td>6.338373</td>\n",
       "      <td>70.006608</td>\n",
       "      <td>1.441496e-08</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>AL133453.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC02326</th>\n",
       "      <td>0.805504</td>\n",
       "      <td>6.203114</td>\n",
       "      <td>71.988762</td>\n",
       "      <td>1.119157e-08</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>LINC02326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STC1</th>\n",
       "      <td>1.174349</td>\n",
       "      <td>9.604055</td>\n",
       "      <td>65.198570</td>\n",
       "      <td>2.726236e-08</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>STC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC009554.1</th>\n",
       "      <td>0.919839</td>\n",
       "      <td>5.432419</td>\n",
       "      <td>66.342901</td>\n",
       "      <td>2.335296e-08</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>AC009554.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYO9B</th>\n",
       "      <td>0.275462</td>\n",
       "      <td>6.544115</td>\n",
       "      <td>9.349547</td>\n",
       "      <td>5.418162e-03</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.047871</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>MYO9B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LGALS1</th>\n",
       "      <td>0.148322</td>\n",
       "      <td>11.244907</td>\n",
       "      <td>9.338860</td>\n",
       "      <td>5.440826e-03</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.048015</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>LGALS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKAP13</th>\n",
       "      <td>0.219309</td>\n",
       "      <td>8.574686</td>\n",
       "      <td>9.308130</td>\n",
       "      <td>5.506573e-03</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.048424</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>AKAP13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMM2</th>\n",
       "      <td>0.290057</td>\n",
       "      <td>6.346882</td>\n",
       "      <td>9.249752</td>\n",
       "      <td>5.633888e-03</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.049256</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>PMM2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FUT11</th>\n",
       "      <td>0.247233</td>\n",
       "      <td>6.037282</td>\n",
       "      <td>9.224587</td>\n",
       "      <td>5.689764e-03</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.049628</td>\n",
       "      <td>PV</td>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>FUT11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>486 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               logFC     logCPM          F        PValue  bonferroni  \\\n",
       "IL1B        1.441441   5.748745  83.425492  2.861042e-09    0.000021   \n",
       "AL133453.1  0.628851   6.338373  70.006608  1.441496e-08    0.000108   \n",
       "LINC02326   0.805504   6.203114  71.988762  1.119157e-08    0.000084   \n",
       "STC1        1.174349   9.604055  65.198570  2.726236e-08    0.000205   \n",
       "AC009554.1  0.919839   5.432419  66.342901  2.335296e-08    0.000175   \n",
       "...              ...        ...        ...           ...         ...   \n",
       "MYO9B       0.275462   6.544115   9.349547  5.418162e-03    1.000000   \n",
       "LGALS1      0.148322  11.244907   9.338860  5.440826e-03    1.000000   \n",
       "AKAP13      0.219309   8.574686   9.308130  5.506573e-03    1.000000   \n",
       "PMM2        0.290057   6.346882   9.249752  5.633888e-03    1.000000   \n",
       "FUT11       0.247233   6.037282   9.224587  5.689764e-03    1.000000   \n",
       "\n",
       "            p_val_adj cell_type cluster        gene  \n",
       "IL1B         0.000021        PV   PV.Lm        IL1B  \n",
       "AL133453.1   0.000036        PV   PV.Lm  AL133453.1  \n",
       "LINC02326    0.000036        PV   PV.Lm   LINC02326  \n",
       "STC1         0.000041        PV   PV.Lm        STC1  \n",
       "AC009554.1   0.000041        PV   PV.Lm  AC009554.1  \n",
       "...               ...       ...     ...         ...  \n",
       "MYO9B        0.047871        PV   PV.Lm       MYO9B  \n",
       "LGALS1       0.048015        PV   PV.Lm      LGALS1  \n",
       "AKAP13       0.048424        PV   PV.Lm      AKAP13  \n",
       "PMM2         0.049256        PV   PV.Lm        PMM2  \n",
       "FUT11        0.049628        PV   PV.Lm       FUT11  \n",
       "\n",
       "[486 rows x 9 columns]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PVDEG_up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "electrical-attempt",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in PVDEG_up.cluster])\n",
    "PVDEG_up = PVDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "powerful-peter",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = PVDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "PVDEG_up= PVDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "\n",
    "pd.DataFrame(PVDEG_up).to_csv('../data/listeria_inf_PV_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "republican-corps",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IL1B</th>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>IL1B</td>\n",
       "      <td>1.441441</td>\n",
       "      <td>5.748745</td>\n",
       "      <td>83.425492</td>\n",
       "      <td>2.861042e-09</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL133453.1</th>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>AL133453.1</td>\n",
       "      <td>0.628851</td>\n",
       "      <td>6.338373</td>\n",
       "      <td>70.006608</td>\n",
       "      <td>1.441496e-08</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC02326</th>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>LINC02326</td>\n",
       "      <td>0.805504</td>\n",
       "      <td>6.203114</td>\n",
       "      <td>71.988762</td>\n",
       "      <td>1.119157e-08</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STC1</th>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>STC1</td>\n",
       "      <td>1.174349</td>\n",
       "      <td>9.604055</td>\n",
       "      <td>65.198570</td>\n",
       "      <td>2.726236e-08</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC009554.1</th>\n",
       "      <td>PV.Lm</td>\n",
       "      <td>AC009554.1</td>\n",
       "      <td>0.919839</td>\n",
       "      <td>5.432419</td>\n",
       "      <td>66.342901</td>\n",
       "      <td>2.335296e-08</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>PV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           cluster        gene     logFC    logCPM          F        PValue  \\\n",
       "IL1B         PV.Lm        IL1B  1.441441  5.748745  83.425492  2.861042e-09   \n",
       "AL133453.1   PV.Lm  AL133453.1  0.628851  6.338373  70.006608  1.441496e-08   \n",
       "LINC02326    PV.Lm   LINC02326  0.805504  6.203114  71.988762  1.119157e-08   \n",
       "STC1         PV.Lm        STC1  1.174349  9.604055  65.198570  2.726236e-08   \n",
       "AC009554.1   PV.Lm  AC009554.1  0.919839  5.432419  66.342901  2.335296e-08   \n",
       "\n",
       "            bonferroni  p_val_adj cell_type  \n",
       "IL1B          0.000021   0.000021        PV  \n",
       "AL133453.1    0.000108   0.000036        PV  \n",
       "LINC02326     0.000084   0.000036        PV  \n",
       "STC1          0.000205   0.000041        PV  \n",
       "AC009554.1    0.000175   0.000041        PV  "
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PVDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bronze-cutting",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "copyrighted-acrylic",
   "metadata": {},
   "source": [
    "### Endo_f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "complimentary-preserve",
   "metadata": {},
   "outputs": [],
   "source": [
    "EndoDEG = pd.read_csv(path+'/3_DEGs_SC_LIS_24h_Endof.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "private-federation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C15orf48</th>\n",
       "      <td>0.954024</td>\n",
       "      <td>4.448099</td>\n",
       "      <td>126.433440</td>\n",
       "      <td>1.752048e-11</td>\n",
       "      <td>1.298793e-07</td>\n",
       "      <td>1.290975e-07</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Lm</td>\n",
       "      <td>C15orf48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL1B</th>\n",
       "      <td>1.031443</td>\n",
       "      <td>5.523735</td>\n",
       "      <td>98.954441</td>\n",
       "      <td>2.370360e-10</td>\n",
       "      <td>1.757148e-06</td>\n",
       "      <td>4.392869e-07</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Lm</td>\n",
       "      <td>IL1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf56</th>\n",
       "      <td>0.969427</td>\n",
       "      <td>5.265645</td>\n",
       "      <td>101.557146</td>\n",
       "      <td>1.808770e-10</td>\n",
       "      <td>1.340841e-06</td>\n",
       "      <td>4.392869e-07</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Lm</td>\n",
       "      <td>C1orf56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DTWD2</th>\n",
       "      <td>1.169978</td>\n",
       "      <td>4.599164</td>\n",
       "      <td>95.351859</td>\n",
       "      <td>3.479421e-10</td>\n",
       "      <td>2.579295e-06</td>\n",
       "      <td>5.158590e-07</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Lm</td>\n",
       "      <td>DTWD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CXCL5</th>\n",
       "      <td>0.943957</td>\n",
       "      <td>5.881886</td>\n",
       "      <td>89.663507</td>\n",
       "      <td>6.534115e-10</td>\n",
       "      <td>4.843740e-06</td>\n",
       "      <td>8.072899e-07</td>\n",
       "      <td>Endo_f</td>\n",
       "      <td>Endo_f.Lm</td>\n",
       "      <td>CXCL5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             logFC    logCPM           F        PValue    bonferroni  \\\n",
       "C15orf48  0.954024  4.448099  126.433440  1.752048e-11  1.298793e-07   \n",
       "IL1B      1.031443  5.523735   98.954441  2.370360e-10  1.757148e-06   \n",
       "C1orf56   0.969427  5.265645  101.557146  1.808770e-10  1.340841e-06   \n",
       "DTWD2     1.169978  4.599164   95.351859  3.479421e-10  2.579295e-06   \n",
       "CXCL5     0.943957  5.881886   89.663507  6.534115e-10  4.843740e-06   \n",
       "\n",
       "             p_val_adj cell_type    cluster      gene  \n",
       "C15orf48  1.290975e-07    Endo_f  Endo_f.Lm  C15orf48  \n",
       "IL1B      4.392869e-07    Endo_f  Endo_f.Lm      IL1B  \n",
       "C1orf56   4.392869e-07    Endo_f  Endo_f.Lm   C1orf56  \n",
       "DTWD2     5.158590e-07    Endo_f  Endo_f.Lm     DTWD2  \n",
       "CXCL5     8.072899e-07    Endo_f  Endo_f.Lm     CXCL5  "
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EndoDEG_up= recoverDEGs(EndoDEG, infection='Lm', celltype='Endo_f', direction='up')\n",
    "EndoDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "applicable-sample",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in EndoDEG_up.cluster])\n",
    "EndoDEG_up = EndoDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "floating-increase",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = EndoDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "EndoDEG_up= EndoDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(EndoDEG_up).to_csv('../data/listeria_inf_Endof_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "august-baseline",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C15orf48</th>\n",
       "      <td>Endo_f.Lm</td>\n",
       "      <td>C15orf48</td>\n",
       "      <td>0.954024</td>\n",
       "      <td>4.448099</td>\n",
       "      <td>126.433440</td>\n",
       "      <td>1.752048e-11</td>\n",
       "      <td>1.298793e-07</td>\n",
       "      <td>1.290975e-07</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL1B</th>\n",
       "      <td>Endo_f.Lm</td>\n",
       "      <td>IL1B</td>\n",
       "      <td>1.031443</td>\n",
       "      <td>5.523735</td>\n",
       "      <td>98.954441</td>\n",
       "      <td>2.370360e-10</td>\n",
       "      <td>1.757148e-06</td>\n",
       "      <td>4.392869e-07</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf56</th>\n",
       "      <td>Endo_f.Lm</td>\n",
       "      <td>C1orf56</td>\n",
       "      <td>0.969427</td>\n",
       "      <td>5.265645</td>\n",
       "      <td>101.557146</td>\n",
       "      <td>1.808770e-10</td>\n",
       "      <td>1.340841e-06</td>\n",
       "      <td>4.392869e-07</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DTWD2</th>\n",
       "      <td>Endo_f.Lm</td>\n",
       "      <td>DTWD2</td>\n",
       "      <td>1.169978</td>\n",
       "      <td>4.599164</td>\n",
       "      <td>95.351859</td>\n",
       "      <td>3.479421e-10</td>\n",
       "      <td>2.579295e-06</td>\n",
       "      <td>5.158590e-07</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CXCL5</th>\n",
       "      <td>Endo_f.Lm</td>\n",
       "      <td>CXCL5</td>\n",
       "      <td>0.943957</td>\n",
       "      <td>5.881886</td>\n",
       "      <td>89.663507</td>\n",
       "      <td>6.534115e-10</td>\n",
       "      <td>4.843740e-06</td>\n",
       "      <td>8.072899e-07</td>\n",
       "      <td>Endo_f</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            cluster      gene     logFC    logCPM           F        PValue  \\\n",
       "C15orf48  Endo_f.Lm  C15orf48  0.954024  4.448099  126.433440  1.752048e-11   \n",
       "IL1B      Endo_f.Lm      IL1B  1.031443  5.523735   98.954441  2.370360e-10   \n",
       "C1orf56   Endo_f.Lm   C1orf56  0.969427  5.265645  101.557146  1.808770e-10   \n",
       "DTWD2     Endo_f.Lm     DTWD2  1.169978  4.599164   95.351859  3.479421e-10   \n",
       "CXCL5     Endo_f.Lm     CXCL5  0.943957  5.881886   89.663507  6.534115e-10   \n",
       "\n",
       "            bonferroni     p_val_adj cell_type  \n",
       "C15orf48  1.298793e-07  1.290975e-07    Endo_f  \n",
       "IL1B      1.757148e-06  4.392869e-07    Endo_f  \n",
       "C1orf56   1.340841e-06  4.392869e-07    Endo_f  \n",
       "DTWD2     2.579295e-06  5.158590e-07    Endo_f  \n",
       "CXCL5     4.843740e-06  8.072899e-07    Endo_f  "
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EndoDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "million-generator",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "liquid-bubble",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fatal-atlanta",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "central-angola",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PATH to Trophoblast cells DGE tables\n",
    "path='../../diffGeneExpression/trophoblast/results/SC_limma_Trophoblast_LIS_24h'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nervous-robin",
   "metadata": {},
   "source": [
    "### VCT_fusing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "aging-stamp",
   "metadata": {},
   "outputs": [],
   "source": [
    "VCTfDEG = pd.read_csv(path+'/1_DEGs_SC_LIS_24h_VCT_fusing.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "crazy-tower",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TIMP1</th>\n",
       "      <td>0.680611</td>\n",
       "      <td>8.260483</td>\n",
       "      <td>79.159917</td>\n",
       "      <td>1.465612e-08</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Lm</td>\n",
       "      <td>TIMP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL1B</th>\n",
       "      <td>1.054424</td>\n",
       "      <td>4.226661</td>\n",
       "      <td>75.946173</td>\n",
       "      <td>2.071042e-08</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Lm</td>\n",
       "      <td>IL1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCL20</th>\n",
       "      <td>1.289987</td>\n",
       "      <td>4.382292</td>\n",
       "      <td>75.347699</td>\n",
       "      <td>2.211625e-08</td>\n",
       "      <td>0.000168</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Lm</td>\n",
       "      <td>CCL20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTGS2</th>\n",
       "      <td>1.212487</td>\n",
       "      <td>3.317895</td>\n",
       "      <td>72.663077</td>\n",
       "      <td>2.984766e-08</td>\n",
       "      <td>0.000227</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Lm</td>\n",
       "      <td>PTGS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC02109</th>\n",
       "      <td>1.920732</td>\n",
       "      <td>5.358849</td>\n",
       "      <td>58.137602</td>\n",
       "      <td>1.789002e-07</td>\n",
       "      <td>0.001358</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>VCT_fusing</td>\n",
       "      <td>VCT_fusing.Lm</td>\n",
       "      <td>LINC02109</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              logFC    logCPM          F        PValue  bonferroni  p_val_adj  \\\n",
       "TIMP1      0.680611  8.260483  79.159917  1.465612e-08    0.000111   0.000034   \n",
       "IL1B       1.054424  4.226661  75.946173  2.071042e-08    0.000157   0.000034   \n",
       "CCL20      1.289987  4.382292  75.347699  2.211625e-08    0.000168   0.000034   \n",
       "PTGS2      1.212487  3.317895  72.663077  2.984766e-08    0.000227   0.000038   \n",
       "LINC02109  1.920732  5.358849  58.137602  1.789002e-07    0.001358   0.000194   \n",
       "\n",
       "            cell_type        cluster       gene  \n",
       "TIMP1      VCT_fusing  VCT_fusing.Lm      TIMP1  \n",
       "IL1B       VCT_fusing  VCT_fusing.Lm       IL1B  \n",
       "CCL20      VCT_fusing  VCT_fusing.Lm      CCL20  \n",
       "PTGS2      VCT_fusing  VCT_fusing.Lm      PTGS2  \n",
       "LINC02109  VCT_fusing  VCT_fusing.Lm  LINC02109  "
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTfDEG_up= recoverDEGs(VCTfDEG, infection='Lm', celltype='VCT_fusing', direction='up')\n",
    "VCTfDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "korean-given",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in VCTfDEG_up.cluster])\n",
    "VCTfVCTfDEG_upDEG = VCTfDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "other-assets",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = VCTfDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "VCTfDEG_up= VCTfDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(VCTfDEG_up).to_csv('../data/listeria_inf_VCTfusing_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "uniform-waters",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TIMP1</th>\n",
       "      <td>VCT_fusing.Lm</td>\n",
       "      <td>TIMP1</td>\n",
       "      <td>0.680611</td>\n",
       "      <td>8.260483</td>\n",
       "      <td>79.159917</td>\n",
       "      <td>1.465612e-08</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL1B</th>\n",
       "      <td>VCT_fusing.Lm</td>\n",
       "      <td>IL1B</td>\n",
       "      <td>1.054424</td>\n",
       "      <td>4.226661</td>\n",
       "      <td>75.946173</td>\n",
       "      <td>2.071042e-08</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCL20</th>\n",
       "      <td>VCT_fusing.Lm</td>\n",
       "      <td>CCL20</td>\n",
       "      <td>1.289987</td>\n",
       "      <td>4.382292</td>\n",
       "      <td>75.347699</td>\n",
       "      <td>2.211625e-08</td>\n",
       "      <td>0.000168</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTGS2</th>\n",
       "      <td>VCT_fusing.Lm</td>\n",
       "      <td>PTGS2</td>\n",
       "      <td>1.212487</td>\n",
       "      <td>3.317895</td>\n",
       "      <td>72.663077</td>\n",
       "      <td>2.984766e-08</td>\n",
       "      <td>0.000227</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC02109</th>\n",
       "      <td>VCT_fusing.Lm</td>\n",
       "      <td>LINC02109</td>\n",
       "      <td>1.920732</td>\n",
       "      <td>5.358849</td>\n",
       "      <td>58.137602</td>\n",
       "      <td>1.789002e-07</td>\n",
       "      <td>0.001358</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 cluster       gene     logFC    logCPM          F  \\\n",
       "TIMP1      VCT_fusing.Lm      TIMP1  0.680611  8.260483  79.159917   \n",
       "IL1B       VCT_fusing.Lm       IL1B  1.054424  4.226661  75.946173   \n",
       "CCL20      VCT_fusing.Lm      CCL20  1.289987  4.382292  75.347699   \n",
       "PTGS2      VCT_fusing.Lm      PTGS2  1.212487  3.317895  72.663077   \n",
       "LINC02109  VCT_fusing.Lm  LINC02109  1.920732  5.358849  58.137602   \n",
       "\n",
       "                 PValue  bonferroni  p_val_adj   cell_type  \n",
       "TIMP1      1.465612e-08    0.000111   0.000034  VCT_fusing  \n",
       "IL1B       2.071042e-08    0.000157   0.000034  VCT_fusing  \n",
       "CCL20      2.211625e-08    0.000168   0.000034  VCT_fusing  \n",
       "PTGS2      2.984766e-08    0.000227   0.000038  VCT_fusing  \n",
       "LINC02109  1.789002e-07    0.001358   0.000194  VCT_fusing  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTfDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "fifteen-breathing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SPP1</th>\n",
       "      <td>VCT_fusing.Lm</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>0.662044</td>\n",
       "      <td>7.817243</td>\n",
       "      <td>56.992413</td>\n",
       "      <td>2.089290e-07</td>\n",
       "      <td>0.001586</td>\n",
       "      <td>0.000198</td>\n",
       "      <td>VCT_fusing</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            cluster  gene     logFC    logCPM          F        PValue  \\\n",
       "SPP1  VCT_fusing.Lm  SPP1  0.662044  7.817243  56.992413  2.089290e-07   \n",
       "\n",
       "      bonferroni  p_val_adj   cell_type  \n",
       "SPP1    0.001586   0.000198  VCT_fusing  "
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTfDEG_up.loc[['SPP1']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "consecutive-intranet",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "close-paint",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "baking-bangladesh",
   "metadata": {},
   "source": [
    "### VCT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "indian-teach",
   "metadata": {},
   "outputs": [],
   "source": [
    "VCTDEG = pd.read_csv(path+'/2_DEGs_SC_LIS_24h_VCT.tsv', header=0, index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "animal-nigeria",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TIMP1</th>\n",
       "      <td>0.790908</td>\n",
       "      <td>8.111007</td>\n",
       "      <td>161.926302</td>\n",
       "      <td>1.839387e-12</td>\n",
       "      <td>1.749257e-08</td>\n",
       "      <td>1.749257e-08</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Lm</td>\n",
       "      <td>TIMP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CXCL8</th>\n",
       "      <td>0.717749</td>\n",
       "      <td>7.067350</td>\n",
       "      <td>132.519030</td>\n",
       "      <td>1.602635e-11</td>\n",
       "      <td>1.524106e-07</td>\n",
       "      <td>7.620529e-08</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Lm</td>\n",
       "      <td>CXCL8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLIP4</th>\n",
       "      <td>0.551446</td>\n",
       "      <td>6.074261</td>\n",
       "      <td>96.669251</td>\n",
       "      <td>4.221855e-10</td>\n",
       "      <td>4.014984e-06</td>\n",
       "      <td>5.735692e-07</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Lm</td>\n",
       "      <td>CLIP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL1B</th>\n",
       "      <td>1.245488</td>\n",
       "      <td>4.258540</td>\n",
       "      <td>99.110268</td>\n",
       "      <td>3.281763e-10</td>\n",
       "      <td>3.120956e-06</td>\n",
       "      <td>5.735692e-07</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Lm</td>\n",
       "      <td>IL1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM12</th>\n",
       "      <td>1.024389</td>\n",
       "      <td>7.483833</td>\n",
       "      <td>92.547554</td>\n",
       "      <td>6.536630e-10</td>\n",
       "      <td>6.216335e-06</td>\n",
       "      <td>6.907039e-07</td>\n",
       "      <td>VCT</td>\n",
       "      <td>VCT.Lm</td>\n",
       "      <td>ADAM12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           logFC    logCPM           F        PValue    bonferroni  \\\n",
       "TIMP1   0.790908  8.111007  161.926302  1.839387e-12  1.749257e-08   \n",
       "CXCL8   0.717749  7.067350  132.519030  1.602635e-11  1.524106e-07   \n",
       "CLIP4   0.551446  6.074261   96.669251  4.221855e-10  4.014984e-06   \n",
       "IL1B    1.245488  4.258540   99.110268  3.281763e-10  3.120956e-06   \n",
       "ADAM12  1.024389  7.483833   92.547554  6.536630e-10  6.216335e-06   \n",
       "\n",
       "           p_val_adj cell_type cluster    gene  \n",
       "TIMP1   1.749257e-08       VCT  VCT.Lm   TIMP1  \n",
       "CXCL8   7.620529e-08       VCT  VCT.Lm   CXCL8  \n",
       "CLIP4   5.735692e-07       VCT  VCT.Lm   CLIP4  \n",
       "IL1B    5.735692e-07       VCT  VCT.Lm    IL1B  \n",
       "ADAM12  6.907039e-07       VCT  VCT.Lm  ADAM12  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTDEG_up= recoverDEGs(VCTDEG, infection='Lm', celltype='VCT', direction='up')\n",
    "VCTDEG_up.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "naked-carnival",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure DEGs only includes samples in meta\n",
    "cl2include = set(cellphoneDB_adata.obs.cell_type.tolist())\n",
    "idx = np.array([i in cl2include for i in VCTDEG_up.cluster])\n",
    "VCTDEG_up = VCTDEG_up[ idx ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "color-bowling",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1st column = cluster; 2nd column = gene \n",
    "cncol = VCTDEG_up.columns.tolist()\n",
    "cncol = [item for item in cncol if item not in ['cluster', 'gene'] ]\n",
    "VCTDEG_up= VCTDEG_up[ ['cluster', 'gene']+cncol ]\n",
    "pd.DataFrame(VCTDEG_up).to_csv('../data/listeria_inf_VCT_DEGsup.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "induced-desire",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TIMP1</th>\n",
       "      <td>VCT.Lm</td>\n",
       "      <td>TIMP1</td>\n",
       "      <td>0.790908</td>\n",
       "      <td>8.111007</td>\n",
       "      <td>161.926302</td>\n",
       "      <td>1.839387e-12</td>\n",
       "      <td>1.749257e-08</td>\n",
       "      <td>1.749257e-08</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CXCL8</th>\n",
       "      <td>VCT.Lm</td>\n",
       "      <td>CXCL8</td>\n",
       "      <td>0.717749</td>\n",
       "      <td>7.067350</td>\n",
       "      <td>132.519030</td>\n",
       "      <td>1.602635e-11</td>\n",
       "      <td>1.524106e-07</td>\n",
       "      <td>7.620529e-08</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLIP4</th>\n",
       "      <td>VCT.Lm</td>\n",
       "      <td>CLIP4</td>\n",
       "      <td>0.551446</td>\n",
       "      <td>6.074261</td>\n",
       "      <td>96.669251</td>\n",
       "      <td>4.221855e-10</td>\n",
       "      <td>4.014984e-06</td>\n",
       "      <td>5.735692e-07</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL1B</th>\n",
       "      <td>VCT.Lm</td>\n",
       "      <td>IL1B</td>\n",
       "      <td>1.245488</td>\n",
       "      <td>4.258540</td>\n",
       "      <td>99.110268</td>\n",
       "      <td>3.281763e-10</td>\n",
       "      <td>3.120956e-06</td>\n",
       "      <td>5.735692e-07</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM12</th>\n",
       "      <td>VCT.Lm</td>\n",
       "      <td>ADAM12</td>\n",
       "      <td>1.024389</td>\n",
       "      <td>7.483833</td>\n",
       "      <td>92.547554</td>\n",
       "      <td>6.536630e-10</td>\n",
       "      <td>6.216335e-06</td>\n",
       "      <td>6.907039e-07</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster    gene     logFC    logCPM           F        PValue  \\\n",
       "TIMP1   VCT.Lm   TIMP1  0.790908  8.111007  161.926302  1.839387e-12   \n",
       "CXCL8   VCT.Lm   CXCL8  0.717749  7.067350  132.519030  1.602635e-11   \n",
       "CLIP4   VCT.Lm   CLIP4  0.551446  6.074261   96.669251  4.221855e-10   \n",
       "IL1B    VCT.Lm    IL1B  1.245488  4.258540   99.110268  3.281763e-10   \n",
       "ADAM12  VCT.Lm  ADAM12  1.024389  7.483833   92.547554  6.536630e-10   \n",
       "\n",
       "          bonferroni     p_val_adj cell_type  \n",
       "TIMP1   1.749257e-08  1.749257e-08       VCT  \n",
       "CXCL8   1.524106e-07  7.620529e-08       VCT  \n",
       "CLIP4   4.014984e-06  5.735692e-07       VCT  \n",
       "IL1B    3.120956e-06  5.735692e-07       VCT  \n",
       "ADAM12  6.216335e-06  6.907039e-07       VCT  "
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTDEG_up.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "emerging-method",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>bonferroni</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SPP1</th>\n",
       "      <td>VCT.Lm</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>0.683136</td>\n",
       "      <td>7.794127</td>\n",
       "      <td>65.946544</td>\n",
       "      <td>1.703181e-08</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>VCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     cluster  gene     logFC    logCPM          F        PValue  bonferroni  \\\n",
       "SPP1  VCT.Lm  SPP1  0.683136  7.794127  65.946544  1.703181e-08    0.000162   \n",
       "\n",
       "      p_val_adj cell_type  \n",
       "SPP1   0.000004       VCT  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCTDEG_up.loc[['SPP1']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "million-merchant",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "associate-break",
   "metadata": {},
   "source": [
    "### (3) Run cellphoneDB"
   ]
  },
  {
   "cell_type": "raw",
   "id": "auburn-chain",
   "metadata": {},
   "source": [
    "source activate cpdb\n",
    "\n",
    "bash run_cellphoneDB.sh "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
